<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>flotilla.data_model.base module &mdash; flotilla 0.2.8dev documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.2.0/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.8dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.8dev documentation" href="index.html" />
    <link rel="up" title="flotilla.data_model package" href="flotilla.data_model.html" />
    <link rel="prev" title="flotilla.data_model package" href="flotilla.data_model.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.8dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="tutorial.html">Tutorial</a></li>
                <li><a href="gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">flotilla</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Flotilla tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#create-a-datapackage">Create a datapackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#computing-on-splicing-data">Computing on splicing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#day-to-day-helpful-tips">Day to day helpful tips</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">flotilla.data_model.base module</a><ul>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/flotilla.data_model.base.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="flotilla-data-model-base-module">
<h1>flotilla.data_model.base module<a class="headerlink" href="#flotilla-data-model-base-module" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-flotilla.data_model.base"></span><p>Base data class for all data types. All data types in flotilla inherit from
this, or a child object (like ExpressionData).</p>
<dl class="class">
<dt id="flotilla.data_model.base.BaseData">
<em class="property">class </em><tt class="descclassname">flotilla.data_model.base.</tt><tt class="descname">BaseData</tt><big>(</big><em>data</em>, <em>thresh=-inf</em>, <em>minimum_samples=0</em>, <em>feature_data=None</em>, <em>feature_rename_col=None</em>, <em>feature_ignore_subset_cols=None</em>, <em>technical_outliers=None</em>, <em>outliers=None</em>, <em>pooled=None</em>, <em>predictor_config_manager=None</em>, <em>data_type=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="http://docs.python.org/library/functions.html#object" title="(in Python v2.7)"><tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></a></p>
<p>Base class for biological data measurements.</p>
<p>All data types in flotilla inherit from this, and have all functionality
described here</p>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#flotilla.data_model.base.BaseData.feature_subsets" title="flotilla.data_model.base.BaseData.feature_subsets"><tt class="xref py py-obj docutils literal"><span class="pre">feature_subsets</span></tt></a></td>
<td>Dict of feature subset names to their list of feature ids</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#flotilla.data_model.base.BaseData.variant" title="flotilla.data_model.base.BaseData.variant"><tt class="xref py py-obj docutils literal"><span class="pre">variant</span></tt></a></td>
<td>Features whose variance is 2 std devs away from mean variance</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="93%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>data</td>
<td>(pandas.DataFrame) A (n_samples, m_features) sized DataFrame of filtered input data, with features with too few samples (<tt class="docutils literal"><span class="pre">minimum_samples</span></tt>) detected at <tt class="docutils literal"><span class="pre">thresh</span></tt> removed. Compared to <tt class="xref py py-attr docutils literal"><span class="pre">data_original</span></tt>, <a href="#id1"><span class="problematic" id="id2">``</span></a>m_features &lt;= n_features`</td>
</tr>
<tr class="row-even"><td>data_type</td>
<td>(str) String indicating what kind of data this is, e.g. &#8220;splicing&#8221; or &#8220;expression&#8221;</td>
</tr>
<tr class="row-odd"><td>data_original</td>
<td>(pandas.DataFrame) A (n_samples, n_features) sized DataFrame of all input data, before removing features for having too few samples</td>
</tr>
<tr class="row-even"><td>feature_data</td>
<td>(pandas.DataFrame) A (k_features, n_features_about_features) sized DataFrame of features about the feature data. Notice that this DataFrame does not need to be the same size as the data, but must at least include all the features from <tt class="xref py py-attr docutils literal"><span class="pre">data</span></tt>. Compared to <tt class="xref py py-attr docutils literal"><span class="pre">data</span></tt>, <tt class="docutils literal"><span class="pre">k_features</span> <span class="pre">&gt;=</span> <span class="pre">m_features</span></tt></td>
</tr>
<tr class="row-odd"><td>predictor_config_manager</td>
<td>(PredictorConfigManager) Manage different combinations of predictor on different data subtypes</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#flotilla.data_model.base.BaseData.maybe_renamed_to_feature_id" title="flotilla.data_model.base.BaseData.maybe_renamed_to_feature_id"><tt class="xref py py-obj docutils literal"><span class="pre">maybe_renamed_to_feature_id</span></tt></a>(feature_id)</td>
<td>To be able to give a simple gene name, e.g.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="93%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>feature_renamer</td>
<td>If <tt class="docutils literal"><span class="pre">feature_rename_col</span></tt> is specified in <tt class="xref py py-meth docutils literal"><span class="pre">BaseData.__init__()</span></tt>, this will rename the feature ID to a new name. If <tt class="docutils literal"><span class="pre">feature_rename_col</span></tt> is not specified, then this will return the original id</td>
</tr>
</tbody>
</table>
<p>Abstract base class for biological measurements</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data</strong> : pandas.DataFrame</p>
<blockquote>
<div><p>A samples x features (samples on rows, features on columns)
dataframe with some kind of measurements of cells,
e.g. gene expression values such as TPM, RPKM or FPKM, alternative
splicing &#8220;Percent-spliced-in&#8221; (PSI) values, or RNA editing scores.
Note: If the columns are a multi-index, the &#8220;level 0&#8221; is assumed to
be the unique, crazy ID like &#8216;ENSG00000100320&#8217;, and &#8220;level 1&#8221; is
assumed to be the convenient gene name like &#8220;RBFOX2&#8221;</p>
</div></blockquote>
<p><strong>thresh</strong> : float, optional (default=-np.inf)</p>
<blockquote>
<div><p>Minimum value to accept for this data.</p>
</div></blockquote>
<p><strong>minimum_samples</strong> : int, optional (default=0)</p>
<blockquote>
<div><p>Minimum number of samples with values greater than <tt class="docutils literal"><span class="pre">thresh</span></tt>.
E.g., for use with &#8220;at least 3 single cells expressing the gene at
greater than 1 TPM.&#8221;</p>
</div></blockquote>
<p><strong>feature_data</strong> : pandas.DataFrame, optional (default=None)</p>
<blockquote>
<div><p>A features x attributes dataframe of metadata about the features,
e.g. annotating whether the gene is a housekeeping gene</p>
</div></blockquote>
<p><strong>feature_rename_col</strong> : str, optional (default=None)</p>
<blockquote>
<div><p>Which column in the feature_data to use to rename feature IDs
from a crazy ID to a common gene symbol, e.g. to transform
&#8216;ENSG00000100320&#8217; into &#8216;RBFOX2&#8217;</p>
</div></blockquote>
<p><strong>feature_ignore_subset_cols</strong> : list-like (default=None)</p>
<blockquote>
<div><p>Columns in the feature data to ignore when making subsets,
e.g. &#8220;gene_name&#8221; shouldn&#8217;t be used to create subsets, since it&#8217;s
just a small number of them.</p>
</div></blockquote>
<p><strong>technical_outliers</strong> : list-like, optional (default=None)</p>
<blockquote>
<div><p>List of sample IDs which should be completely ignored because
they didn&#8217;t pass the technical quality control</p>
</div></blockquote>
<p><strong>outliers</strong> : list-like, optional (default=None)</p>
<blockquote>
<div><p>List of sample IDs which should be marked as outliers for
plotting and interpretation purposes</p>
</div></blockquote>
<p><strong>pooled</strong> : list-like, optional (default=None)</p>
<blockquote>
<div><p>List of sample IDs which should be marked as pooled for plotting
and interpretation purposes.</p>
</div></blockquote>
<p><strong>predictor_config_manager</strong> : PredictorConfigManager, optional</p>
<blockquote>
<div><p>(default=None)
Object used to organize inputs to
<tt class="xref py py-class docutils literal"><span class="pre">compute.predict.Regressor</span></tt> and
<tt class="xref py py-class docutils literal"><span class="pre">compute.predict.Classifier</span></tt>. If None, one is initialized
for this instance.</p>
</div></blockquote>
<p><strong>data_type</strong> : str, optional (default=None)</p>
<blockquote class="last">
<div><p>A string indicating what kind of data this is, e.g. &#8220;expression&#8221; or
&#8220;splicing&#8221;</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>Any cells not marked as &#8220;technical_outliers&#8221;, &#8220;outliers&#8221; or &#8220;pooled&#8221;
are considered as single-cell samples.</p>
<dl class="method">
<dt id="flotilla.data_model.base.BaseData.big_nmf_space_transitions">
<tt class="descname">big_nmf_space_transitions</tt><big>(</big><em>groupby</em>, <em>phenotype_transitions</em>, <em>n=0.5</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.big_nmf_space_transitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.big_nmf_space_transitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get features whose change in NMF space between phenotypes is large</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>groupby</strong> : mappable</p>
<blockquote>
<div><p>A sample id to phenotype group mapping</p>
</div></blockquote>
<p><strong>phenotype_transitions</strong> : list of length-2 tuples of str</p>
<blockquote>
<div><p>List of (&#8216;phenotype1&#8217;, &#8216;phenotype2&#8217;) transitions whose change in
distribution you are interested in</p>
</div></blockquote>
<p><strong>n</strong> : int</p>
<blockquote>
<div><p>Minimum number of samples per phenotype, per event</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>big_transitions</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_events, n_transitions) dataframe of the NMF distances between
splicing events</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.binify">
<tt class="descname">binify</tt><big>(</big><em>data</em>, <em>bins=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.binify"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.binify" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.binned_nmf_reduced">
<tt class="descname">binned_nmf_reduced</tt><big>(</big><em>*args</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.binned_nmf_reduced"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.binned_nmf_reduced" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.classify">
<tt class="descname">classify</tt><big>(</big><em>trait</em>, <em>sample_ids</em>, <em>feature_ids</em>, <em>standardize=True</em>, <em>data_name='expression'</em>, <em>predictor_name='ExtraTreesClassifier'</em>, <em>predictor_obj=None</em>, <em>predictor_scoring_fun=None</em>, <em>score_cutoff_fun=None</em>, <em>n_features_dependent_kwargs=None</em>, <em>constant_kwargs=None</em>, <em>plotting_kwargs=None</em>, <em>color=None</em>, <em>groupby=None</em>, <em>label_to_color=None</em>, <em>label_to_marker=None</em>, <em>order=None</em>, <em>bins=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.classify"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.classify" title="Permalink to this definition">¶</a></dt>
<dd><p>Make and memoize a predictor on a categorical trait (associated
with samples) subset of genes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>trait</strong> : pandas.Series</p>
<blockquote>
<div><p>samples x categorical feature</p>
</div></blockquote>
<p><strong>sample_ids</strong> : None or list of strings</p>
<blockquote>
<div><p>If None, all sample ids will be used, else only the sample ids
specified</p>
</div></blockquote>
<p><strong>feature_ids</strong> : None or list of strings</p>
<blockquote>
<div><p>If None, all features will be used, else only the features
specified</p>
</div></blockquote>
<p><strong>standardize</strong> : bool</p>
<blockquote>
<div><p>Whether or not to &#8220;whiten&#8221; (make all variables uncorrelated) and
mean-center and make unit-variance all the data via sklearn
.preprocessing.StandardScaler</p>
</div></blockquote>
<p><strong>predictor</strong> : flotilla.visualize.predict classifier</p>
<blockquote>
<div><p>Must inherit from flotilla.visualize.PredictorBaseViz. Default is
flotilla.visualize.predict.ClassifierViz</p>
</div></blockquote>
<p><strong>predictor_kwargs</strong> : dict or None</p>
<blockquote>
<div><p>Additional &#8216;keyword arguments&#8217; to supply to the predictor class</p>
</div></blockquote>
<p><strong>predictor_scoring_fun</strong> : function</p>
<blockquote>
<div><p>Function to get the feature scores for a scikit-learn classifier.
This can be different for different classifiers, e.g. for a
classifier named &#8220;x&#8221; it could be <a href="#id5"><span class="problematic" id="id6">x.scores_</span></a>, for other it&#8217;s
<a href="#id7"><span class="problematic" id="id8">x.feature_importances_</span></a>. Default: lambda x: <a href="#id9"><span class="problematic" id="id10">x.feature_importances_</span></a></p>
</div></blockquote>
<p><strong>score_cutoff_fun</strong> : function</p>
<blockquote>
<div><p>Function to cut off insignificant scores
Default: lambda scores: np.mean(x) + 2 * np.std(x)</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>predictor</strong> : flotilla.compute.predict.PredictorBaseViz</p>
<blockquote class="last">
<div><p>A ready-to-plot object containing the predictions</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.detect_outliers">
<tt class="descname">detect_outliers</tt><big>(</big><em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>featurewise=False</em>, <em>reducer=None</em>, <em>standardize=True</em>, <em>reducer_kwargs=None</em>, <em>bins=None</em>, <em>outlier_detection_method=None</em>, <em>outlier_detection_method_kwargs=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.detect_outliers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.detect_outliers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.feature_renamer_series">
<tt class="descname">feature_renamer_series</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.feature_renamer_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.feature_renamer_series" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas Series of the original feature ids to the renamed ids</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.feature_subset_to_feature_ids">
<tt class="descname">feature_subset_to_feature_ids</tt><big>(</big><em>feature_subset</em>, <em>rename=True</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.feature_subset_to_feature_ids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.feature_subset_to_feature_ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a feature subset name to a list of feature ids</p>
</dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.feature_subsets">
<tt class="descname">feature_subsets</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.feature_subsets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.feature_subsets" title="Permalink to this definition">¶</a></dt>
<dd><p>Dict of feature subset names to their list of feature ids</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.jsd_2d">
<tt class="descname">jsd_2d</tt><big>(</big><em>groupby=None</em>, <em>n_iter=100</em>, <em>n_bins=10</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.jsd_2d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.jsd_2d" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean Jensen-Shannon divergence of features across phenotypes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>groupby</strong> : mappable</p>
<blockquote>
<div><p>A samples to phenotypes mapping</p>
</div></blockquote>
<p><strong>n_iter</strong> : int</p>
<blockquote>
<div><p>Number of bootstrap resampling iterations to perform for the
within-group comparisons</p>
</div></blockquote>
<p><strong>n_bins</strong> : int</p>
<blockquote>
<div><p>Number of bins to binify the singles data on</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>jsd_2d</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_phenotypes, n_phenotypes) symmetric dataframe of the mean JSD
between and within phenotypes</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.jsd_df">
<tt class="descname">jsd_df</tt><big>(</big><em>groupby=None</em>, <em>n_iter=100</em>, <em>n_bins=10</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.jsd_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.jsd_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Jensen-Shannon divergence of features across phenotypes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>groupby</strong> : mappable</p>
<blockquote>
<div><p>A samples to phenotypes mapping</p>
</div></blockquote>
<p><strong>n_iter</strong> : int</p>
<blockquote>
<div><p>Number of bootstrap resampling iterations to perform for the
within-group comparisons</p>
</div></blockquote>
<p><strong>n_bins</strong> : int</p>
<blockquote>
<div><p>Number of bins to binify the singles data on</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>jsd_df</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_features, n_phenotypes^2) dataframe of the JSD between each
feature between and within phenotypes</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.maybe_renamed_to_feature_id">
<tt class="descname">maybe_renamed_to_feature_id</tt><big>(</big><em>feature_id</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.maybe_renamed_to_feature_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.maybe_renamed_to_feature_id" title="Permalink to this definition">¶</a></dt>
<dd><p>To be able to give a simple gene name, e.g. &#8220;RBFOX2&#8221; and get the
official ENSG ids or MISO ids</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>feature_id</strong> : str</p>
<blockquote>
<div><p>The name of a feature ID. Could be either a common gene name, as in
what the crazy IDs are <tt class="xref py py-meth docutils literal"><span class="pre">feature_renamer()</span></tt> to, or</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>feature_id</strong> : str or list-like</p>
<blockquote class="last">
<div><p>Valid Feature ID(s) that can be used to subset self.data</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.nmf_space_positions">
<tt class="descname">nmf_space_positions</tt><big>(</big><em>groupby</em>, <em>n=0.5</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.nmf_space_positions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.nmf_space_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate NMF-space position of splicing events in phenotype groups</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>groupby</strong> : mappable</p>
<blockquote>
<div><p>A sample id to phenotype mapping</p>
</div></blockquote>
<p><strong>n</strong> : int or float</p>
<blockquote>
<div><p>If int, then this is the absolute number of cells that are minimum
required to calculate modalities. If a float, then require this
fraction of samples to calculate modalities, e.g. if 0.6, then at
least 60% of samples must have an event detected for modality
detection</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>df</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_events, n_groups) dataframe of NMF positions</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.nmf_space_transitions">
<tt class="descname">nmf_space_transitions</tt><big>(</big><em>groupby</em>, <em>phenotype_transitions</em>, <em>n=0.5</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.nmf_space_transitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.nmf_space_transitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get distance in NMF space of different splicing events</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>groupby</strong> : mappable</p>
<blockquote>
<div><p>A sample id to phenotype mapping</p>
</div></blockquote>
<p><strong>phenotype_transitions</strong> : list of str pairs</p>
<blockquote>
<div><p>Which phenotype follows from one to the next, for calculating
distances between</p>
</div></blockquote>
<p><strong>n</strong> : int or float</p>
<blockquote>
<div><p>If int, then this is the absolute number of cells that are minimum
required to calculate modalities. If a float, then require this
fraction of samples to calculate modalities, e.g. if 0.6, then at
least 60% of samples must have an event detected for modality
detection</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>nmf_space_transitions</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_events, n_phenotype_transitions) sized DataFrame of the
distances of these events in NMF space</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.outliers">
<tt class="descname">outliers</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.outliers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.outliers" title="Permalink to this definition">¶</a></dt>
<dd><p>Data from only the outlier samples</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_big_nmf_space_transitions">
<tt class="descname">plot_big_nmf_space_transitions</tt><big>(</big><em>phenotype_groupby</em>, <em>phenotype_transitions</em>, <em>phenotype_order</em>, <em>color</em>, <em>phenotype_to_color</em>, <em>phenotype_to_marker</em>, <em>n=0.5</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_big_nmf_space_transitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_big_nmf_space_transitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Violinplots and NMF transitions of features different in phenotypes</p>
<p>Plot violinplots and NMF-space transitions of features that have large
NMF-space transitions between different phenotypes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>n</strong> : int</p>
<blockquote class="last">
<div><p>Minimum number of samples per phenotype, per event</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_classifier">
<tt class="descname">plot_classifier</tt><big>(</big><em>trait</em>, <em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>predictor_name=None</em>, <em>standardize=True</em>, <em>score_coefficient=None</em>, <em>data_name=None</em>, <em>groupby=None</em>, <em>label_to_color=None</em>, <em>label_to_marker=None</em>, <em>order=None</em>, <em>color=None</em>, <em>**plotting_kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_classifier"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_classifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify samples on boolean or categorical traits</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>trait</strong> : pandas.Series</p>
<blockquote>
<div><p>A (n_samples,) series of categorical features. Must have the same
index as <tt class="xref py py-attr docutils literal"><span class="pre">data</span></tt></p>
</div></blockquote>
<p><strong>sample_ids</strong> : list-like, optional (default=None)</p>
<blockquote>
<div><p>Which samples to use to classify</p>
</div></blockquote>
<p><strong>feature_ids</strong> : list-like, optional (default=None)</p>
<blockquote>
<div><p>Which features to use</p>
</div></blockquote>
<p><strong>predictor_name</strong> : str</p>
<blockquote>
<div><p>Name of the predictor to use, in
<tt class="xref py py-attr docutils literal"><span class="pre">predictor_config_manager</span></tt></p>
</div></blockquote>
<p><strong>standardize</strong> : bool, optional (default=True)</p>
<blockquote>
<div><p>If True, mean-center the data so the mean of all features is 0,
and divide by the standard deviation so the standard deviation of
all features is 1. This allows us to compare lowly expressed
features and highly expressed features on the same playing field</p>
</div></blockquote>
<p><strong>data_name</strong> : str, optional (default=None)</p>
<blockquote>
<div><p>Name for this subset of the data</p>
</div></blockquote>
<p><strong>groupby</strong> : mappable, optional (default=None)</p>
<blockquote>
<div><p>Map each sample id to a group, such as a phenotype label</p>
</div></blockquote>
<p><strong>label_to_color</strong> : dict, optional (default=None)</p>
<blockquote>
<div><p>For each phenotype label, assign a color</p>
</div></blockquote>
<p><strong>label_to_marker</strong> : dict, optional (default=None)</p>
<blockquote>
<div><p>For each phenotype label, assign a plotting marker symbol/shape</p>
</div></blockquote>
<p><strong>order</strong> : list, optional (default=None)</p>
<blockquote>
<div><p>For violinplots, the order of the phenotype groups</p>
</div></blockquote>
<p><strong>color</strong> : list, optional (default=None)</p>
<blockquote>
<div><p>For violinplots, the colors of the phenotypes in their order</p>
</div></blockquote>
<p><strong>plotting_kwargs</strong> : other keyword arguments</p>
<blockquote>
<div><p>All other keyword arguments are passed to
<tt class="xref py py-meth docutils literal"><span class="pre">Classifier.__call__()</span></tt>, which passes them to
<tt class="xref py py-meth docutils literal"><span class="pre">DecomopsitionViz.__call__()</span></tt></p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>cv</strong> : ClassifierViz</p>
<blockquote class="last">
<div><p>Visualziation of the classifier</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_clustermap">
<tt class="descname">plot_clustermap</tt><big>(</big><em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>data=None</em>, <em>feature_colors=None</em>, <em>sample_id_to_color=None</em>, <em>metric='euclidean'</em>, <em>method='average'</em>, <em>norm_features=True</em>, <em>scale_fig_by_data=True</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_clustermap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_clustermap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_correlations">
<tt class="descname">plot_correlations</tt><big>(</big><em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>data=None</em>, <em>featurewise=False</em>, <em>sample_id_to_color=None</em>, <em>metric='euclidean'</em>, <em>method='average'</em>, <em>scale_fig_by_data=True</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_correlations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_correlations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_dimensionality_reduction">
<tt class="descname">plot_dimensionality_reduction</tt><big>(</big><em>x_pc=1</em>, <em>y_pc=2</em>, <em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>featurewise=False</em>, <em>reducer=None</em>, <em>plot_violins=False</em>, <em>groupby=None</em>, <em>label_to_color=None</em>, <em>label_to_marker=None</em>, <em>order=None</em>, <em>reduce_kwargs=None</em>, <em>title=''</em>, <em>most_variant_features=False</em>, <em>std_multiplier=2</em>, <em>scale_by_variance=True</em>, <em>**plotting_kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_dimensionality_reduction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_dimensionality_reduction" title="Permalink to this definition">¶</a></dt>
<dd><p>Principal component-like analysis of measurements</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>x_pc</strong> : int, optional</p>
<blockquote>
<div><p>Which principal component to plot on the x-axis (default 1)</p>
</div></blockquote>
<p><strong>y_pc</strong> : int, optional</p>
<blockquote>
<div><p>Which principal component to plot on the y-axis (default 2)</p>
</div></blockquote>
<p><strong>sample_ids</strong> : list, optional</p>
<blockquote>
<div><p>If None, plot all the samples. If a list of strings, must be
valid sample ids of the data. (default None)</p>
</div></blockquote>
<p><strong>feature_ids</strong> : list, optional</p>
<blockquote>
<div><p>If None, plot all the features. If a list of strings, perform and
plot dimensionality reduction on only these feature ids</p>
</div></blockquote>
<p><strong>featurewise</strong> : bool, optional</p>
<blockquote>
<div><p>If True, the features are reduced on the samples, and the plotted
points are features, not samples. (default False)</p>
</div></blockquote>
<p><strong>reducer</strong> : <a class="reference internal" href="flotilla.compute.decomposition.html#flotilla.compute.decomposition.DataFrameReducerBase" title="flotilla.compute.decomposition.DataFrameReducerBase"><tt class="xref py py-class docutils literal"><span class="pre">DataFrameReducerBase</span></tt></a>, optional</p>
<blockquote>
<div><p>Which decomposition object to use. Must be a child of
<a class="reference internal" href="flotilla.compute.decomposition.html#flotilla.compute.decomposition.DataFrameReducerBase" title="flotilla.compute.decomposition.DataFrameReducerBase"><tt class="xref py py-class docutils literal"><span class="pre">DataFrameReducerBase</span></tt></a> as this has built-in
compatibility with pandas.DataFrames.
(default=:py:class:<cite>.DataFramePCA</cite>)</p>
</div></blockquote>
<p><strong>plot_violins</strong> : bool, optional</p>
<blockquote>
<div><p>If True, plot the violinplots of the top features. This
can take a long time, so to save time you can turn it off if you
just want a quick look at the PCA. (default False)</p>
</div></blockquote>
<p><strong>groupby</strong> : mappable, optional</p>
<blockquote>
<div><p>Map each sample id to a group, such as a phenotype label
(default None)</p>
</div></blockquote>
<p><strong>label_to_color</strong> : dict, optional</p>
<blockquote>
<div><p>For each phenotype label, assign a color
(default None)</p>
</div></blockquote>
<p><strong>label_to_marker</strong> : dict, optional</p>
<blockquote>
<div><p>For each phenotype label, assign a plotting marker symbol/shape
(default None)</p>
</div></blockquote>
<p><strong>order</strong> : list, optional</p>
<blockquote>
<div><p>For violinplots, the order of the phenotype groups
(default None)</p>
</div></blockquote>
<p><strong>reduce_kwargs</strong> : dict, optional</p>
<blockquote>
<div><p>Keyword arguments to the reducer (default None)</p>
</div></blockquote>
<p><strong>title</strong> : str, optional</p>
<blockquote>
<div><p>Title of the reduced space plot (default &#8216;&#8217;)</p>
</div></blockquote>
<p><strong>most_variant_features</strong> : bool, optional</p>
<blockquote>
<div><p>If True, then only take the most variant of the provided features.
The most variant are determined by taking the features whose
variance is <a href="#id3"><span class="problematic" id="id4">``</span></a>std_multiplier``standard deviations away from the
mean feature variance (default False)</p>
</div></blockquote>
<p><strong>std_multiplier</strong> : float, optional</p>
<blockquote>
<div><p>If <tt class="docutils literal"><span class="pre">most_variant_features</span></tt> is True, then use this as a cutoff
for the minimum variance of a feature to be included (default 2)</p>
</div></blockquote>
<p><strong>scale_by_variance</strong> : bool, optional</p>
<blockquote>
<div><p>If True, then scale the x- and y-axes by the explained variance
ratio of the principal component dimensions. Only valid for PCA
and its variations, not for NMF or tSNE. (default True)</p>
</div></blockquote>
<p><strong>plotting_kwargs</strong> : other keyword arguments</p>
<blockquote>
<div><p>All other keyword arguments are passed to
<tt class="xref py py-meth docutils literal"><span class="pre">DecomopsitionViz.plot()</span></tt></p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>viz</strong> : <a class="reference internal" href="flotilla.visualize.decomposition.html#flotilla.visualize.decomposition.DecompositionViz" title="flotilla.visualize.decomposition.DecompositionViz"><tt class="xref py py-class docutils literal"><span class="pre">DecompositionViz</span></tt></a></p>
<blockquote class="last">
<div><p>Object with plotted dimensionality reduction</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_feature">
<tt class="descname">plot_feature</tt><big>(</big><em>feature_id</em>, <em>sample_ids=None</em>, <em>phenotype_groupby=None</em>, <em>phenotype_order=None</em>, <em>color=None</em>, <em>phenotype_to_color=None</em>, <em>phenotype_to_marker=None</em>, <em>nmf_xlabel=None</em>, <em>nmf_ylabel=None</em>, <em>nmf_space=False</em>, <em>fig=None</em>, <em>axesgrid=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the violinplot of a feature. Have the option to show NMF movement</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_nmf_space_transitions">
<tt class="descname">plot_nmf_space_transitions</tt><big>(</big><em>feature_id</em>, <em>groupby</em>, <em>phenotype_to_color</em>, <em>phenotype_to_marker</em>, <em>order</em>, <em>ax=None</em>, <em>xlabel=None</em>, <em>ylabel=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_nmf_space_transitions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_nmf_space_transitions" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_outliers">
<tt class="descname">plot_outliers</tt><big>(</big><em>reducer</em>, <em>outlier_detector</em>, <em>**pca_args</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_outliers"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_outliers" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_pca">
<tt class="descname">plot_pca</tt><big>(</big><em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Call <tt class="docutils literal"><span class="pre">plot_dimensionality_reduction</span></tt> with PCA specifically</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_two_features">
<tt class="descname">plot_two_features</tt><big>(</big><em>feature1</em>, <em>feature2</em>, <em>groupby=None</em>, <em>label_to_color=None</em>, <em>fillna=None</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_two_features"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_two_features" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the values of two features</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.plot_two_samples">
<tt class="descname">plot_two_samples</tt><big>(</big><em>sample1</em>, <em>sample2</em>, <em>fillna=None</em>, <em>**kwargs</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.plot_two_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.plot_two_samples" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>sample1</strong> : str</p>
<blockquote>
<div><p>Name of the sample to plot on the x-axis</p>
</div></blockquote>
<p><strong>sample2</strong> : str</p>
<blockquote>
<div><p>Name of the sample to plot on the y-axis</p>
</div></blockquote>
<p><strong>fillna</strong> : float</p>
<blockquote>
<div><p>Value to replace NAs with</p>
</div></blockquote>
<p><strong>Any other keyword arguments valid for seaborn.jointplot</strong></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>jointgrid</strong> : seaborn.axisgrid.JointGrid</p>
<blockquote>
<div><p>Returns a JointGrid instance</p>
</div></blockquote>
<p>See Also</p>
<p class="last">seaborn.jointplot</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.pooled">
<tt class="descname">pooled</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.pooled"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.pooled" title="Permalink to this definition">¶</a></dt>
<dd><p>Data from only the pooled samples</p>
</dd></dl>

<dl class="method">
<dt id="flotilla.data_model.base.BaseData.reduce">
<tt class="descname">reduce</tt><big>(</big><em>sample_ids=None</em>, <em>feature_ids=None</em>, <em>featurewise=False</em>, <em>reducer=&lt;class 'flotilla.compute.decomposition.DataFramePCA'&gt;</em>, <em>standardize=True</em>, <em>reducer_kwargs=None</em>, <em>bins=None</em>, <em>most_variant_features=False</em>, <em>std_multiplier=2</em>, <em>cosine_transform=False</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.reduce"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.reduce" title="Permalink to this definition">¶</a></dt>
<dd><p>Make and memoize a reduced dimensionality representation of data</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>data</strong> : pandas.DataFrame</p>
<blockquote>
<div><p>samples x features data to reduce</p>
</div></blockquote>
<p><strong>sample_ids</strong> : None or list of strings</p>
<blockquote>
<div><p>If None, all sample ids will be used, else only the sample ids
specified</p>
</div></blockquote>
<p><strong>feature_ids</strong> : None or list of strings</p>
<blockquote>
<div><p>If None, all features will be used, else only the features
specified</p>
</div></blockquote>
<p><strong>featurewise</strong> : bool</p>
<blockquote>
<div><p>Whether or not to use the features as the &#8220;samples&#8221;, e.g. if you
want to reduce the features in to &#8220;sample-space&#8221; instead of
reducing the samples into &#8220;feature-space&#8221;</p>
</div></blockquote>
<p><strong>standardize</strong> : bool</p>
<blockquote>
<div><p>Whether or not to &#8220;whiten&#8221; (make all variables uncorrelated) and
mean-center via sklearn.preprocessing.StandardScaler</p>
</div></blockquote>
<p><strong>title</strong> : str</p>
<blockquote>
<div><p>Title of the plot</p>
</div></blockquote>
<p><strong>reducer_kwargs</strong> : dict</p>
<blockquote>
<div><p>Any additional arguments to send to the reducer</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>reducer_object</strong> : flotilla.compute.reduce.ReducerViz</p>
<blockquote class="last">
<div><p>A ready-to-plot object containing the reduced space</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.singles">
<tt class="descname">singles</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.singles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.singles" title="Permalink to this definition">¶</a></dt>
<dd><p>Data from only the single cells</p>
</dd></dl>

<dl class="staticmethod">
<dt id="flotilla.data_model.base.BaseData.transition_distances">
<em class="property">static </em><tt class="descname">transition_distances</tt><big>(</big><em>positions</em>, <em>transitions</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.transition_distances"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.transition_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Get NMF distance of features between phenotype transitions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>positions</strong> : pandas.DataFrame</p>
<blockquote>
<div><p>A ((n_features, phenotypes), 2) MultiIndex dataframe of the NMF
positions of splicing events for different phenotypes</p>
</div></blockquote>
<p><strong>transitions</strong> : list of 2-string tuples</p>
<blockquote>
<div><p>List of (phenotype1, phenotype2) transitions</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>transitions</strong> : pandas.DataFrame</p>
<blockquote class="last">
<div><p>A (n_features, n_transitions) DataFrame of the NMF distances
of features between different phenotypes</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="flotilla.data_model.base.BaseData.variant">
<tt class="descname">variant</tt><a class="reference internal" href="_modules/flotilla/data_model/base.html#BaseData.variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.BaseData.variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Features whose variance is 2 std devs away from mean variance</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="flotilla.data_model.base.subsets_from_metadata">
<tt class="descclassname">flotilla.data_model.base.</tt><tt class="descname">subsets_from_metadata</tt><big>(</big><em>metadata</em>, <em>minimum</em>, <em>subset_type</em>, <em>ignore=None</em><big>)</big><a class="reference internal" href="_modules/flotilla/data_model/base.html#subsets_from_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#flotilla.data_model.base.subsets_from_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Get subsets from metadata, including boolean and categorical columns</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>metadata</strong> : pandas.DataFrame</p>
<blockquote>
<div><p>The dataframe whose columns to use to create subsets of the rows</p>
</div></blockquote>
<p><strong>minimum</strong> : int</p>
<blockquote>
<div><p>Minimum number of rows required for a column or group in the column
to be included</p>
</div></blockquote>
<p><strong>subset_type</strong> : str</p>
<blockquote>
<div><p>The name of the kind of subset. e.g. &#8220;samples&#8221; or &#8220;features&#8221;</p>
</div></blockquote>
<p><strong>ignore</strong> : list-like</p>
<blockquote>
<div><p>List of columns to ignore</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>subsets</strong> : dict</p>
<blockquote class="last">
<div><p>A name: row_ids mapping of which samples correspond to which group</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

    
  </body>
</html>