<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Create a barebones datapackage &mdash; flotilla 0.2.7 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.7',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.7 documentation" href="../index.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.7</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Flotilla tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#getting-started-downloading-cleaning-and-importing-data">Getting Started: Downloading, cleaning, and importing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#exploratory-analysis">Exploratory Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Create a barebones datapackage</a><ul>
<li><a class="reference internal" href="#shalek-and-satija-et-al-2013">Shalek and Satija, <em>et al</em> (2013)</a></li>
<li><a class="reference internal" href="#expression-data">Expression data</a></li>
<li><a class="reference internal" href="#expression-feature-data">Expression feature data</a></li>
<li><a class="reference internal" href="#splicing-data">Splicing Data</a></li>
<li><a class="reference internal" href="#metadata">Metadata</a></li>
<li><a class="reference internal" href="#mapping-stats-data">Mapping stats data</a></li>
<li><a class="reference internal" href="#create-a-flotilla-study">Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="../_sources/tutorial/barebones_study_making.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="create-a-barebones-datapackage">
<h1>Create a barebones datapackage<a class="headerlink" href="#create-a-barebones-datapackage" title="Permalink to this headline">¶</a></h1>
<p>Before we begin, let&#8217;s import everything we need.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># Import the flotilla package for biological data analysis</span>
<span class="kn">import</span> <span class="nn">flotilla</span>

<span class="c"># Import &quot;numerical python&quot; library for number crunching</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="c"># Import &quot;panel data analysis&quot; library for tabular data</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
</pre></div>
</div>
<div class="section" id="shalek-and-satija-et-al-2013">
<h2>Shalek and Satija, <em>et al</em> (2013)<a class="headerlink" href="#shalek-and-satija-et-al-2013" title="Permalink to this headline">¶</a></h2>
<p>In the 2013 paper, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Single-cell transcriptomics reveals bimodality in
expression and splicing in immune
cells</a> (Shalek and
Satija, <em>et al</em>. <em>Nature</em> (2013)), Regev and colleagues performed
single-cell sequencing 18 bone marrow-derived dendritic cells (BMDCs),
in addition to 3 pooled samples.</p>
</div>
<div class="section" id="expression-data">
<h2>Expression data<a class="headerlink" href="#expression-data" title="Permalink to this headline">¶</a></h2>
<p>First, we will read in the expression data. These data were obtained
using,</p>
<div class="highlight-python"><div class="highlight"><pre>wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz
</pre></div>
</div>
<p>We will also compare to the supplementary table 2 data, obtained using</p>
<div class="highlight-python"><div class="highlight"><pre>wget http://www.nature.com/nature/journal/v498/n7453/extref/nature12172-s1.zip
unzip nature12172-s1.zip
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s">&quot;GSE41265_allGenesTPM.txt.gz&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s">&quot;gzip&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>---------------------------------------------------------------------------
IOError                                   Traceback (most recent call last)

&lt;ipython-input-2-b45e9120e279&gt; in &lt;module&gt;()
----&gt; 1 expression = pd.read_table(&quot;GSE41265_allGenesTPM.txt.gz&quot;, compression=&quot;gzip&quot;, index_col=0)
      2 expression.head()


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/parsers.pyc in parser_f(filepath_or_buffer, sep, dialect, compression, doublequote, escapechar, quotechar, quoting, skipinitialspace, lineterminator, header, index_col, names, prefix, skiprows, skipfooter, skip_footer, na_values, na_fvalues, true_values, false_values, delimiter, converters, dtype, usecols, engine, delim_whitespace, as_recarray, na_filter, compact_ints, use_unsigned, low_memory, buffer_lines, warn_bad_lines, error_bad_lines, keep_default_na, thousands, comment, decimal, parse_dates, keep_date_col, dayfirst, date_parser, memory_map, float_precision, nrows, iterator, chunksize, verbose, encoding, squeeze, mangle_dupe_cols, tupleize_cols, infer_datetime_format, skip_blank_lines)
    468                     skip_blank_lines=skip_blank_lines)
    469
--&gt; 470         return _read(filepath_or_buffer, kwds)
    471
    472     parser_f.__name__ = name


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/parsers.pyc in _read(filepath_or_buffer, kwds)
    244
    245     # Create the parser.
--&gt; 246     parser = TextFileReader(filepath_or_buffer, **kwds)
    247
    248     if (nrows is not None) and (chunksize is not None):


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/parsers.pyc in __init__(self, f, engine, **kwds)
    560             self.options[&#39;has_index_names&#39;] = kwds[&#39;has_index_names&#39;]
    561
--&gt; 562         self._make_engine(self.engine)
    563
    564     def _get_options_with_defaults(self, engine):


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/parsers.pyc in _make_engine(self, engine)
    697     def _make_engine(self, engine=&#39;c&#39;):
    698         if engine == &#39;c&#39;:
--&gt; 699             self._engine = CParserWrapper(self.f, **self.options)
    700         else:
    701             if engine == &#39;python&#39;:


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/parsers.pyc in __init__(self, src, **kwds)
   1064         kwds[&#39;allow_leading_cols&#39;] = self.index_col is not False
   1065
-&gt; 1066         self._reader = _parser.TextReader(src, **kwds)
   1067
   1068         # XXX


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/parser.so in pandas.parser.TextReader.__cinit__ (pandas/parser.c:3163)()


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/parser.so in pandas.parser.TextReader._setup_parser_source (pandas/parser.c:5335)()


/home/travis/miniconda/envs/testenv/lib/python2.7/gzip.pyc in __init__(self, filename, mode, compresslevel, fileobj, mtime)
     92             mode += &#39;b&#39;
     93         if fileobj is None:
---&gt; 94             fileobj = self.myfileobj = __builtin__.open(filename, mode or &#39;rb&#39;)
     95         if filename is None:
     96             # Issue #13781: os.fdopen() creates a fileobj with a bogus name


IOError: [Errno 2] No such file or directory: &#39;GSE41265_allGenesTPM.txt.gz&#39;
</pre></div>
</div>
<p>These data are in the &#8220;transcripts per million,&#8221; aka TPM unit. See
<a class="reference external" href="http://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">this</a>
blog post if that sounds weird to you.</p>
<p>These data are formatted with samples on the columns, and genes on the
rows. But we want the opposite, with samples on the rows and genes on
the columns. This follows
<tt class="docutils literal"><span class="pre">`scikit-learn</span></tt> &lt;<a class="reference external" href="http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset">http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset</a>&gt;`_&#8217;s
standard of data matrices with size (<tt class="docutils literal"><span class="pre">n_samples</span></tt>, <tt class="docutils literal"><span class="pre">n_features</span></tt>) as
each gene is a feature. So we will simply transpose this.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">T</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>GENE</th>
      <th>XKR4</th>
      <th>AB338584</th>
      <th>B3GAT2</th>
      <th>NPL</th>
      <th>T2</th>
      <th>T</th>
      <th>PDE10A</th>
      <th>1700010I14RIK</th>
      <th>6530411M01RIK</th>
      <th>PABPC6</th>
      <th>...</th>
      <th>AK085062</th>
      <th>DHX9</th>
      <th>RNASET2B</th>
      <th>FGFR1OP</th>
      <th>CCR6</th>
      <th>BRP44L</th>
      <th>AK014435</th>
      <th>AK015714</th>
      <th>SFT2D1</th>
      <th>PRR18</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>72.008590</td>
      <td>0.109249</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.774638</td>
      <td>23.520936</td>
      <td>0.000000</td>
      <td>0</td>
      <td>460.316773</td>
      <td>0</td>
      <td>0.000000</td>
      <td>39.442566</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.172009</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.367391</td>
      <td>1.887873</td>
      <td>0.000000</td>
      <td>0</td>
      <td>823.890290</td>
      <td>0</td>
      <td>0.000000</td>
      <td>4.967412</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>0</td>
      <td>0</td>
      <td>0.023441</td>
      <td>128.062012</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.249858</td>
      <td>0.313510</td>
      <td>0.166772</td>
      <td>0</td>
      <td>1002.354241</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.095082</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.354157</td>
      <td>0.000000</td>
      <td>0.887003</td>
      <td>0</td>
      <td>1230.766795</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.131215</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.182703</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.039263</td>
      <td>0.000000</td>
      <td>131.077131</td>
      <td>0</td>
      <td>1614.749122</td>
      <td>0</td>
      <td>0.242179</td>
      <td>95.485743</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27723 columns</p>
</div><p>The authors filtered the expression data based on having at least 3
single cells express genes with at TPM (transcripts per million, ) &gt; 1.
We can express this in using the
<tt class="docutils literal"><span class="pre">`pandas</span></tt> &lt;<a class="reference external" href="http://pandas.pydata.org">http://pandas.pydata.org</a>&gt;`_ DataFrames easily.</p>
<p>First, from reading the paper and looking at the data, I know there are
18 single cells, and there are 18 samples that start with the letter
&#8220;S.&#8221; So I will extract the single samples from the <tt class="docutils literal"><span class="pre">index</span></tt> (row names)
using a <tt class="docutils literal"><span class="pre">lambda</span></tt>, a tiny function which in this case, tells me whether
or not that sample id begins with the letter &#8220;S&#8221;.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">singles_ids</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;S&#39;</span><span class="p">))]</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">singles_ids</span><span class="p">))</span>
<span class="n">singles</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">]</span>

<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">singles</span><span class="p">[</span><span class="n">singles</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">expression_filtered</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">expression_filtered</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">6312</span><span class="p">)</span>
</pre></div>
</div>
<p>Hmm, that&#8217;s strange. The paper states that they had 6313 genes after
filtering, but I get 6312. Even using &#8220;<tt class="docutils literal"><span class="pre">singles</span> <span class="pre">&gt;=</span> <span class="pre">1</span></tt>&#8221; doesn&#8217;t help.</p>
<p>(I also tried this with the expression table provided in the
supplementary data as &#8220;<tt class="docutils literal"><span class="pre">SupplementaryTable2.xlsx</span></tt>,&#8221; and got the same
results.)</p>
<p>Now that we&#8217;ve taken care of importing and filtering the expression
data, let&#8217;s do the feature data of the expression data.</p>
</div>
<div class="section" id="expression-feature-data">
<h2>Expression feature data<a class="headerlink" href="#expression-feature-data" title="Permalink to this headline">¶</a></h2>
<p>This is similar to the <tt class="docutils literal"><span class="pre">fData</span></tt> from <tt class="docutils literal"><span class="pre">BioconductoR</span></tt>, where there&#8217;s
some additional data on your features that you want to look at. They
uploaded information about the features in their OTHER expression
matrix, uploaded as a supplementary file, <tt class="docutils literal"><span class="pre">Supplementary_Table2.xlsx</span></tt>.</p>
<p>Notice that this is a <tt class="docutils literal"><span class="pre">csv</span></tt> and not an <tt class="docutils literal"><span class="pre">xlsx</span></tt>. This is because Excel
mangled the gene IDS that started with <tt class="docutils literal"><span class="pre">201*</span></tt> and assumed they were
dates :(</p>
<p>The workaround I did was to add another column to the sheet with the
formula <tt class="docutils literal"><span class="pre">=&quot;'&quot;</span> <span class="pre">&amp;</span> <span class="pre">A1</span></tt>, press <tt class="docutils literal"><span class="pre">Command</span></tt>-<tt class="docutils literal"><span class="pre">Shift</span></tt>-<tt class="docutils literal"><span class="pre">End</span></tt> to select the
end of the rows, and then do <tt class="docutils literal"><span class="pre">Ctrl</span></tt>-<tt class="docutils literal"><span class="pre">D</span></tt> to &#8220;fill down&#8221; to the bottom
(thanks to
<a class="reference external" href="http://superuser.com/questions/298276/excel-keyboard-shortcut-to-copy-fill-down-for-all-cells-with-non-blank-adjacent">this</a>
stackoverflow post for teaching me how to Excel). Then, I saved the file
as a <tt class="docutils literal"><span class="pre">csv</span></tt> for maximum portability and compatibility.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table2.csv&#39;</span><span class="p">,</span>
                            <span class="c"># Need to specify the index column as both the first and the last columns,</span>
                            <span class="c"># Because the last column is the &quot;Gene Category&quot;</span>
                            <span class="n">index_col</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">parse_dates</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">infer_datetime_format</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c"># This was also in features x samples format, so we need to transpose</span>
<span class="n">expression2</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">T</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>---------------------------------------------------------------------------
IOError                                   Traceback (most recent call last)

&lt;ipython-input-5-623c304ac226&gt; in &lt;module&gt;()
      2                             # Need to specify the index column as both the first and the last columns,
      3                             # Because the last column is the &quot;Gene Category&quot;
----&gt; 4                             index_col=[0, -1], parse_dates=False, infer_datetime_format=False)
      5
      6 # This was also in features x samples format, so we need to transpose


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/io/parsers.pyc in parser_f(filepath_or_buffer, sep, dialect, compression, doublequote, escapechar, quotechar, quoting, skipinitialspace, lineterminator, header, index_col, names, prefix, skiprows, skipfooter, skip_footer, na_values, na_fvalues, true_values, false_values, delimiter, converters, dtype, usecols, engine, delim_whitespace, as_recarray, na_filter, compact_ints, use_unsigned, low_memory, buffer_lines, warn_bad_lines, error_bad_lines, keep_default_na, thousands, comment, decimal, parse_dates, keep_date_col, dayfirst, date_parser, memory_map, float_precision, nrows, iterator, chunksize, verbose, encoding, squeeze, mangle_dupe_cols, tupleize_cols, infer_datetime_format, skip_blank_lines)
    468                     skip_blank_lines=skip_blank_lines)
    469
--&gt; 470         return _read(filepath_or_buffer, kwds)
    471
    472     parser_f.__name__ = name


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/io/parsers.pyc in _read(filepath_or_buffer, kwds)
    244
    245     # Create the parser.
--&gt; 246     parser = TextFileReader(filepath_or_buffer, **kwds)
    247
    248     if (nrows is not None) and (chunksize is not None):


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/io/parsers.pyc in __init__(self, f, engine, **kwds)
    560             self.options[&#39;has_index_names&#39;] = kwds[&#39;has_index_names&#39;]
    561
--&gt; 562         self._make_engine(self.engine)
    563
    564     def _get_options_with_defaults(self, engine):


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/io/parsers.pyc in _make_engine(self, engine)
    697     def _make_engine(self, engine=&#39;c&#39;):
    698         if engine == &#39;c&#39;:
--&gt; 699             self._engine = CParserWrapper(self.f, **self.options)
    700         else:
    701             if engine == &#39;python&#39;:


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/io/parsers.pyc in __init__(self, src, **kwds)
   1064         kwds[&#39;allow_leading_cols&#39;] = self.index_col is not False
   1065
-&gt; 1066         self._reader = _parser.TextReader(src, **kwds)
   1067
   1068         # XXX


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/parser.so in pandas.parser.TextReader.__cinit__ (pandas/parser.c:3163)()


/Users/olga/anaconda/envs/flotilla_docs/lib/python2.7/site-packages/pandas/parser.so in pandas.parser.TextReader._setup_parser_source (pandas/parser.c:5779)()


IOError: File nature12172-s1/Supplementary_Table2.csv does not exist
</pre></div>
</div>
<p>Now we need to strip the single-quote I added to all the gene names:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">new_index</span><span class="p">,</span> <span class="n">indexer</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&quot;&#39;&quot;</span><span class="p">),</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">new_index</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<p>We want to create a <tt class="docutils literal"><span class="pre">pandas.DataFrame</span></tt> from the &#8220;Gene Category&#8221; row
for our <tt class="docutils literal"><span class="pre">expression_feature_data</span></tt>, which we will do via:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_ids</span><span class="p">,</span> <span class="n">gene_category</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">gene_categories</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">gene_category</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s">&#39;gene_category&#39;</span><span class="p">)</span>
<span class="n">gene_categories</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression_feature_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_categories</span><span class="p">)</span>
<span class="n">expression_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="splicing-data">
<h2>Splicing Data<a class="headerlink" href="#splicing-data" title="Permalink to this headline">¶</a></h2>
<p>We obtain the splicing data from this study from the supplementary
information, specifically the <tt class="docutils literal"><span class="pre">Supplementary_Table4.xls</span></tt></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table4.xls&#39;</span><span class="p">,</span> <span class="s">&#39;splicingTable.txt&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">T</span>
<span class="n">splicing</span>
</pre></div>
</div>
<p>The three pooled samples aren&#8217;t named consistently with the expression
data, so we have to fix that.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">&#39;P&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">)]</span>
</pre></div>
</div>
<p>Since the pooled sample IDs are inconsistent with the <tt class="docutils literal"><span class="pre">expression</span></tt>
data, we have to change them. We can get the &#8220;P&#8221; and the number after
that using regular expressions, called <tt class="docutils literal"><span class="pre">re</span></tt> in the Python standard
library, e.g.:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">re</span>
<span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="s">&#39;10,000 cell Rep1 (P1)&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">long_pooled_name_to_short</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="k">if</span> <span class="s">&#39;P&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">x</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>


<span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
</pre></div>
</div>
<p>And now we assign this new index as our index to the <tt class="docutils literal"><span class="pre">splicing</span></tt>
dataframe</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="metadata">
<h2>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">¶</a></h2>
<p>Now let&#8217;s get into creating a metadata dataframe. We&#8217;ll use the index
from the <tt class="docutils literal"><span class="pre">expression_filtered</span></tt> data to create the minimum required
column, <tt class="docutils literal"><span class="pre">'phenotype'</span></tt>, which has the name of the phenotype of that
cell. And we&#8217;ll also add the column <tt class="docutils literal"><span class="pre">'pooled'</span></tt> to indicate whether
this sample is pooled or not.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">expression_filtered</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;BDMC&#39;</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;pooled&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;P&#39;</span><span class="p">))</span>

<span class="n">metadata</span>
</pre></div>
</div>
</div>
<div class="section" id="mapping-stats-data">
<h2>Mapping stats data<a class="headerlink" href="#mapping-stats-data" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table1.xls&#39;</span><span class="p">,</span> <span class="n">sheetname</span><span class="o">=</span><span class="s">&#39;SuppTable1 2.txt&#39;</span><span class="p">)</span>
<span class="n">mapping_stats</span>
</pre></div>
</div>
</div>
<div class="section" id="create-a-flotilla-study">
<h2>Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!<a class="headerlink" href="#create-a-flotilla-study" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">Study</span><span class="p">(</span><span class="c"># The metadata describing phenotype and pooled samples</span>
                       <span class="n">metadata</span><span class="p">,</span>

                       <span class="c"># A version for this data</span>
                       <span class="n">version</span><span class="o">=</span><span class="s">&#39;0.1.0&#39;</span><span class="p">,</span>

                       <span class="c"># Dataframe of the filtered expression data</span>
                       <span class="n">expression_data</span><span class="o">=</span><span class="n">expression_filtered</span><span class="p">,</span>

                       <span class="c"># Dataframe of the feature data of the genes</span>
                       <span class="n">expression_feature_data</span><span class="o">=</span><span class="n">expression_feature_data</span><span class="p">,</span>

                       <span class="c"># Dataframe of the splicing data</span>
                       <span class="n">splicing_data</span><span class="o">=</span><span class="n">splicing</span><span class="p">,</span>

                       <span class="c"># Dataframe of the mapping stats data</span>
                       <span class="n">mapping_stats_data</span><span class="o">=</span><span class="n">mapping_stats</span><span class="p">,</span>

                       <span class="c"># Which column in &quot;mapping_stats&quot; has the number of reads</span>
                       <span class="n">mapping_stats_number_mapped_col</span><span class="o">=</span><span class="s">&#39;PF_READS&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>As a side note, you can save this study to disk now, so you can
&#8220;<tt class="docutils literal"><span class="pre">embark</span></tt>&#8221; later:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that this is saved to my home directory, in
<tt class="docutils literal"><span class="pre">~/flotilla_projects/&lt;study_name&gt;/</span></tt>. This will be saved in your home
directory, too.</p>
<p>The <tt class="docutils literal"><span class="pre">datapackage.json</span></tt> file is what holds all the information relative
to the study, and loosely follows the <a class="reference external" href="http://data.okfn.org/doc/data-package">datapackage
spec</a> created by the Open
Knowledge Foundation.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cat</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span><span class="o">/</span><span class="n">datapackage</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
<p>One thing to note is that when you save, the version number is bumped
up. <tt class="docutils literal"><span class="pre">study.version</span></tt> (the one we just made) is <tt class="docutils literal"><span class="pre">0.1.0</span></tt>, but the one
we saved is <tt class="docutils literal"><span class="pre">0.1.1</span></tt>, since we could have made some changes to the
data.</p>
<p>Let&#8217;s look at what else is in this folder:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ls</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span>
</pre></div>
</div>
<p>So this is where all the other files are. Good to know!</p>
<p>We can &#8220;embark&#8221; on this newly-saved study now very painlessly, without
having to open and process all those files again:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study2</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">embark</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

    
  </body>
</html>