<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Loading data from Shalek and Satija (2013) &mdash; flotilla 0.2.7 documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/journal/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.7',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.7 documentation" href="../index.html" />
    <link rel="up" title="Flotilla tutorial" href="../tutorial.html" />
    <link rel="next" title="&lt;no title&gt;" href="getting_started_brainspan.html" />
    <link rel="prev" title="OS X Installation instructions" href="../installation.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.7</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Flotilla tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#getting-started-downloading-cleaning-and-importing-data">Getting Started: Downloading, cleaning, and importing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#exploratory-analysis">Exploratory Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Loading data from Shalek and Satija (2013)</a><ul>
<li><a class="reference internal" href="#expression-data">Expression data</a></li>
<li><a class="reference internal" href="#expression-feature-data">Expression feature data</a></li>
<li><a class="reference internal" href="#splicing-data">Splicing Data</a><ul>
<li><a class="reference internal" href="#metadata">Metadata</a></li>
</ul>
</li>
<li><a class="reference internal" href="#mapping-stats-data">Mapping stats data</a></li>
<li><a class="reference internal" href="#create-a-flotilla-study">Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!</a></li>
<li><a class="reference internal" href="#save-the-study-so-you-can-load-it-later">Save the study so you can load it later</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="loading-data-from-shalek-and-satija-2013">
<span id="getting-started-shalek2013"></span><h1>Loading data from Shalek and Satija (2013)<a class="headerlink" href="#loading-data-from-shalek-and-satija-2013" title="Permalink to this headline">Â¶</a></h1>
<p>In the 2013 paper, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Single-cell transcriptomics reveals bimodality in expression and splicing in immune cells</a> (Shalek and Satija, <em>et al</em>. <em>Nature</em> (2013)), Regev and colleagues performed single-cell sequencing 18 bone marrow-derived dendritic cells (BMDCs), in addition to 3 pooled samples.</p>
<p>This will show you the nitty-gritty of what files you need and in what format to get started with using <tt class="docutils literal"><span class="pre">`flotilla</span></tt> &lt;<a class="reference external" href="http://yeolab.github.io/flotilla/docs">http://yeolab.github.io/flotilla/docs</a>&gt;`_. In addition to being a tutorial explaining <tt class="docutils literal"><span class="pre">flotilla</span></tt> inputs, the goal of this document is also to expose every individual step of reformatting data.</p>
<p>We will go over:</p>
<ul>
<li><p class="first">Using <tt class="docutils literal"><span class="pre">wget</span></tt> to download data over the internet from the command line</p>
</li>
<li><p class="first">Using <tt class="docutils literal"><span class="pre">pandas</span></tt> to read in Excel tables and comma-separated variable (<tt class="docutils literal"><span class="pre">csv</span></tt>) files</p>
</li>
<li><dl class="first docutils">
<dt>Basic orientation to working with <tt class="docutils literal"><span class="pre">pandas</span></tt></dt>
<dd><ul class="first last simple">
<li>Transposing matrix data</li>
<li>Subsetting data on rows and columns</li>
<li>Accessing a single column</li>
<li>Changing the values of a single column</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Cleaning data using <tt class="docutils literal"><span class="pre">pandas</span></tt> and Python standard library packages</dt>
<dd><ul class="first last simple">
<li>Removing genes with mangled IDs</li>
<li>Fixing inconsistent sample naming between data tables</li>
<li>Removing NAs</li>
</ul>
</dd>
</dl>
</li>
<li><p class="first">Importing the data into <tt class="docutils literal"><span class="pre">flotilla</span></tt></p>
</li>
</ul>
<p>Before we begin, let&#8217;s import everything we need.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># Import the flotilla package for biological data analysis</span>
<span class="kn">import</span> <span class="nn">flotilla</span>

<span class="c"># Import &quot;numerical python&quot; library for number crunching</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>

<span class="c"># Import &quot;panel data analysis&quot; library for tabular data</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>:0: FutureWarning: IPython widgets are experimental and may change in the future.
</pre></div>
</div>
<div class="section" id="expression-data">
<h2>Expression data<a class="headerlink" href="#expression-data" title="Permalink to this headline">Â¶</a></h2>
<p>First, we will read in the expression data. These data were obtained using the command <tt class="docutils literal"><span class="pre">wget</span></tt> to download files from the internet via the command line, like so:</p>
<div class="code python highlight-python"><div class="highlight"><pre>%%bash
wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>--2015-05-01 18:26:15--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz
           =&gt; `GSE41265_allGenesTPM.txt.gz&#39;
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 2607:f220:41e:250::12, 130.14.250.12
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|2607:f220:41e:250::12|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE41nnn/GSE41265/suppl ... done.
==&gt; SIZE GSE41265_allGenesTPM.txt.gz ... 1099290
==&gt; EPSV ... done.    ==&gt; RETR GSE41265_allGenesTPM.txt.gz ... done.
Length: 1099290 (1.0M) (unauthoritative)

     0K .......... .......... .......... .......... ..........  4% 1.20M 1s
    50K .......... .......... .......... .......... ..........  9% 6.39M 0s
   100K .......... .......... .......... .......... .......... 13% 6.52M 0s
   150K .......... .......... .......... .......... .......... 18% 54.5M 0s
   200K .......... .......... .......... .......... .......... 23% 6.68M 0s
   250K .......... .......... .......... .......... .......... 27% 84.5M 0s
   300K .......... .......... .......... .......... .......... 32% 7.03M 0s
   350K .......... .......... .......... .......... .......... 37% 66.1M 0s
   400K .......... .......... .......... .......... .......... 41%  117M 0s
   450K .......... .......... .......... .......... .......... 46% 8.21M 0s
   500K .......... .......... .......... .......... .......... 51% 31.2M 0s
   550K .......... .......... .......... .......... .......... 55%  116M 0s
   600K .......... .......... .......... .......... .......... 60% 93.2M 0s
   650K .......... .......... .......... .......... .......... 65%  118M 0s
   700K .......... .......... .......... .......... .......... 69% 7.39M 0s
   750K .......... .......... .......... .......... .......... 74% 92.6M 0s
   800K .......... .......... .......... .......... .......... 79%  163M 0s
   850K .......... .......... .......... .......... .......... 83% 90.1M 0s
   900K .......... .......... .......... .......... .......... 88%  119M 0s
   950K .......... .......... .......... .......... .......... 93% 12.1M 0s
  1000K .......... .......... .......... .......... .......... 97% 38.5M 0s
  1050K .......... .......... ...                             100%  205M=0.1s

2015-05-01 18:26:15 (11.0 MB/s) - `GSE41265_allGenesTPM.txt.gz&#39; saved [1099290]
</pre></div>
</div>
<p>We will also compare to the supplementary table 2 data, obtained also using <tt class="docutils literal"><span class="pre">wget</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre>%%bash
wget http://www.nature.com/nature/journal/v498/n7453/extref/nature12172-s1.zip
unzip nature12172-s1.zip
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Archive:  nature12172-s1.zip
   creating: nature12172-s1/
  inflating: nature12172-s1/Supplementary_Table1.xls
  inflating: nature12172-s1/Supplementary_Table2.xlsx
  inflating: nature12172-s1/Supplementary_Table3.xls
  inflating: nature12172-s1/Supplementary_Table4.xls
  inflating: nature12172-s1/Supplementary_Table5.xls
  inflating: nature12172-s1/Supplementary_Table6.xls
  inflating: nature12172-s1/Supplementary_Table7.xlsx
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>--2015-05-01 18:26:16--  http://www.nature.com/nature/journal/v498/n7453/extref/nature12172-s1.zip
Resolving www.nature.com (www.nature.com)... 107.14.45.48, 107.14.45.25
Connecting to www.nature.com (www.nature.com)|107.14.45.48|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4634226 (4.4M) [application/zip]
Saving to: `nature12172-s1.zip&#39;

     0K .......... .......... .......... .......... ..........  1%  122M 0s
    50K .......... .......... .......... .......... ..........  2% 13.9M 0s
   100K .......... .......... .......... .......... ..........  3% 19.2M 0s
   150K .......... .......... .......... .......... ..........  4% 22.4M 0s
   200K .......... .......... .......... .......... ..........  5% 61.2M 0s
   250K .......... .......... .......... .......... ..........  6% 24.1M 0s
   300K .......... .......... .......... .......... ..........  7% 63.8M 0s
   350K .......... .......... .......... .......... ..........  8% 62.0M 0s
   400K .......... .......... .......... .......... ..........  9% 30.7M 0s
   450K .......... .......... .......... .......... .......... 11% 62.2M 0s
   500K .......... .......... .......... .......... .......... 12% 99.8M 0s
   550K .......... .......... .......... .......... .......... 13% 37.4M 0s
   600K .......... .......... .......... .......... .......... 14% 63.3M 0s
   650K .......... .......... .......... .......... .......... 15% 81.8M 0s
   700K .......... .......... .......... .......... .......... 16% 80.7M 0s
   750K .......... .......... .......... .......... .......... 17% 91.6M 0s
   800K .......... .......... .......... .......... .......... 18% 38.3M 0s
   850K .......... .......... .......... .......... .......... 19% 94.9M 0s
   900K .......... .......... .......... .......... .......... 20% 76.5M 0s
   950K .......... .......... .......... .......... .......... 22% 92.1M 0s
  1000K .......... .......... .......... .......... .......... 23% 79.8M 0s
  1050K .......... .......... .......... .......... .......... 24%  116M 0s
  1100K .......... .......... .......... .......... .......... 25% 62.9M 0s
  1150K .......... .......... .......... .......... .......... 26% 91.7M 0s
  1200K .......... .......... .......... .......... .......... 27% 82.2M 0s
  1250K .......... .......... .......... .......... .......... 28% 51.7M 0s
  1300K .......... .......... .......... .......... .......... 29%  219M 0s
  1350K .......... .......... .......... .......... .......... 30% 90.0M 0s
  1400K .......... .......... .......... .......... .......... 32% 90.7M 0s
  1450K .......... .......... .......... .......... .......... 33% 84.8M 0s
  1500K .......... .......... .......... .......... .......... 34% 79.6M 0s
  1550K .......... .......... .......... .......... .......... 35% 83.3M 0s
  1600K .......... .......... .......... .......... .......... 36% 44.6M 0s
  1650K .......... .......... .......... .......... .......... 37%  174M 0s
  1700K .......... .......... .......... .......... .......... 38% 89.4M 0s
  1750K .......... .......... .......... .......... .......... 39%  126M 0s
  1800K .......... .......... .......... .......... .......... 40% 89.6M 0s
  1850K .......... .......... .......... .......... .......... 41%  119M 0s
  1900K .......... .......... .......... .......... .......... 43% 85.4M 0s
  1950K .......... .......... .......... .......... .......... 44% 88.0M 0s
  2000K .......... .......... .......... .......... .......... 45%  140M 0s
  2050K .......... .......... .......... .......... .......... 46% 49.3M 0s
  2100K .......... .......... .......... .......... .......... 47%  121M 0s
  2150K .......... .......... .......... .......... .......... 48%  136M 0s
  2200K .......... .......... .......... .......... .......... 49% 62.2M 0s
  2250K .......... .......... .......... .......... .......... 50% 78.7M 0s
  2300K .......... .......... .......... .......... .......... 51%  102M 0s
  2350K .......... .......... .......... .......... .......... 53% 63.2M 0s
  2400K .......... .......... .......... .......... .......... 54% 73.6M 0s
  2450K .......... .......... .......... .......... .......... 55% 88.0M 0s
  2500K .......... .......... .......... .......... .......... 56% 87.3M 0s
  2550K .......... .......... .......... .......... .......... 57% 66.1M 0s
  2600K .......... .......... .......... .......... .......... 58%  125M 0s
  2650K .......... .......... .......... .......... .......... 59% 79.4M 0s
  2700K .......... .......... .......... .......... .......... 60% 87.9M 0s
  2750K .......... .......... .......... .......... .......... 61% 29.0M 0s
  2800K .......... .......... .......... .......... .......... 62%  201M 0s
  2850K .......... .......... .......... .......... .......... 64%  120M 0s
  2900K .......... .......... .......... .......... .......... 65%  126M 0s
  2950K .......... .......... .......... .......... .......... 66% 91.5M 0s
  3000K .......... .......... .......... .......... .......... 67%  130M 0s
  3050K .......... .......... .......... .......... .......... 68%  133M 0s
  3100K .......... .......... .......... .......... .......... 69% 65.4M 0s
  3150K .......... .......... .......... .......... .......... 70% 73.6M 0s
  3200K .......... .......... .......... .......... .......... 71% 72.1M 0s
  3250K .......... .......... .......... .......... .......... 72%  185M 0s
  3300K .......... .......... .......... .......... .......... 74% 70.5M 0s
  3350K .......... .......... .......... .......... .......... 75% 61.7M 0s
  3400K .......... .......... .......... .......... .......... 76% 34.9M 0s
  3450K .......... .......... .......... .......... .......... 77% 99.4M 0s
  3500K .......... .......... .......... .......... .......... 78%  192M 0s
  3550K .......... .......... .......... .......... .......... 79%  126M 0s
  3600K .......... .......... .......... .......... .......... 80% 91.4M 0s
  3650K .......... .......... .......... .......... .......... 81% 36.5M 0s
  3700K .......... .......... .......... .......... .......... 82% 77.6M 0s
  3750K .......... .......... .......... .......... .......... 83% 80.2M 0s
  3800K .......... .......... .......... .......... .......... 85% 82.7M 0s
  3850K .......... .......... .......... .......... .......... 86% 92.0M 0s
  3900K .......... .......... .......... .......... .......... 87% 86.2M 0s
  3950K .......... .......... .......... .......... .......... 88% 83.7M 0s
  4000K .......... .......... .......... .......... .......... 89% 96.0M 0s
  4050K .......... .......... .......... .......... .......... 90% 29.8M 0s
  4100K .......... .......... .......... .......... .......... 91% 19.7M 0s
  4150K .......... .......... .......... .......... .......... 92%  176M 0s
  4200K .......... .......... .......... .......... .......... 93%  160M 0s
  4250K .......... .......... .......... .......... .......... 95%  182M 0s
  4300K .......... .......... .......... .......... .......... 96%  187M 0s
  4350K .......... .......... .......... .......... .......... 97%  193M 0s
  4400K .......... .......... .......... .......... .......... 98%  155M 0s
  4450K .......... .......... .......... .......... .......... 99%  190M 0s
  4500K .......... .......... .....                           100% 95.8M=0.07s

2015-05-01 18:26:16 (67.6 MB/s) - `nature12172-s1.zip&#39; saved [4634226/4634226]
</pre></div>
</div>
<p>Now we will read the data in using <tt class="docutils literal"><span class="pre">`pandas</span></tt> &lt;<a class="reference external" href="http://pandas.pydata.org">http://pandas.pydata.org</a>&gt;`_, indicating that the file is compressed via gzip, and that we want the first column to be the &#8220;index&#8221;, or row names of the table.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s">&quot;GSE41265_allGenesTPM.txt.gz&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s">&quot;gzip&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
      <th>S5</th>
      <th>S6</th>
      <th>S7</th>
      <th>S8</th>
      <th>S9</th>
      <th>S10</th>
      <th>...</th>
      <th>S12</th>
      <th>S13</th>
      <th>S14</th>
      <th>S15</th>
      <th>S16</th>
      <th>S17</th>
      <th>S18</th>
      <th>P1</th>
      <th>P2</th>
      <th>P3</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>XKR4</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.019906</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>AB338584</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>B3GAT2</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.023441</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.029378</td>
      <td>0.000000</td>
      <td>0.055452</td>
      <td>0.000000</td>
      <td>0.029448</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.031654</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>42.150208</td>
      <td>0.680327</td>
      <td>0.022996</td>
      <td>0.110236</td>
    </tr>
    <tr>
      <th>NPL</th>
      <td>72.008590</td>
      <td>0.000000</td>
      <td>128.062012</td>
      <td>0.095082</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>112.310234</td>
      <td>104.329122</td>
      <td>0.119230</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.116802</td>
      <td>0.104200</td>
      <td>0.106188</td>
      <td>0.229197</td>
      <td>0.110582</td>
      <td>0.000000</td>
      <td>7.109356</td>
      <td>6.727028</td>
      <td>14.525447</td>
    </tr>
    <tr>
      <th>T2</th>
      <td>0.109249</td>
      <td>0.172009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.182703</td>
      <td>0.076012</td>
      <td>0.078698</td>
      <td>0.000000</td>
      <td>0.093698</td>
      <td>0.076583</td>
      <td>...</td>
      <td>0.693459</td>
      <td>0.010137</td>
      <td>0.081936</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.086879</td>
      <td>0.068174</td>
      <td>0.062063</td>
      <td>0.000000</td>
      <td>0.050605</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã 21 columns</p>
</div><p>These data are in the &#8220;transcripts per million,&#8221; aka TPM unit. See <a class="reference external" href="http://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">this blog post</a> if that sounds weird to you.</p>
<p>These data are formatted with samples on the columns, and genes on the rows. But we want the opposite, with samples on the rows and genes on the columns. This follows <tt class="docutils literal"><span class="pre">`scikit-learn</span></tt> &lt;<a class="reference external" href="http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset">http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset</a>&gt;`_&#8217;s standard of data matrices with size (<tt class="docutils literal"><span class="pre">n_samples</span></tt>, <tt class="docutils literal"><span class="pre">n_features</span></tt>) as each gene is a feature. So we will simply transpose this.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">T</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>GENE</th>
      <th>XKR4</th>
      <th>AB338584</th>
      <th>B3GAT2</th>
      <th>NPL</th>
      <th>T2</th>
      <th>T</th>
      <th>PDE10A</th>
      <th>1700010I14RIK</th>
      <th>6530411M01RIK</th>
      <th>PABPC6</th>
      <th>...</th>
      <th>AK085062</th>
      <th>DHX9</th>
      <th>RNASET2B</th>
      <th>FGFR1OP</th>
      <th>CCR6</th>
      <th>BRP44L</th>
      <th>AK014435</th>
      <th>AK015714</th>
      <th>SFT2D1</th>
      <th>PRR18</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>72.008590</td>
      <td>0.109249</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.774638</td>
      <td>23.520936</td>
      <td>0.000000</td>
      <td>0</td>
      <td>460.316773</td>
      <td>0</td>
      <td>0.000000</td>
      <td>39.442566</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.172009</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.367391</td>
      <td>1.887873</td>
      <td>0.000000</td>
      <td>0</td>
      <td>823.890290</td>
      <td>0</td>
      <td>0.000000</td>
      <td>4.967412</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>0</td>
      <td>0</td>
      <td>0.023441</td>
      <td>128.062012</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.249858</td>
      <td>0.313510</td>
      <td>0.166772</td>
      <td>0</td>
      <td>1002.354241</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.095082</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.354157</td>
      <td>0.000000</td>
      <td>0.887003</td>
      <td>0</td>
      <td>1230.766795</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.131215</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.182703</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.039263</td>
      <td>0.000000</td>
      <td>131.077131</td>
      <td>0</td>
      <td>1614.749122</td>
      <td>0</td>
      <td>0.242179</td>
      <td>95.485743</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã 27723 columns</p>
</div><p>The authors filtered the expression data based on having at least 3 single cells express genes with at TPM (transcripts per million, ) &gt; 1. We can express this in using the <tt class="docutils literal"><span class="pre">pandas</span></tt> DataFrames easily.</p>
<p>First, from reading the paper and looking at the data, I know there are 18 single cells, and there are 18 samples that start with the letter &#8220;S.&#8221; So I will extract the single samples from the <tt class="docutils literal"><span class="pre">index</span></tt> (row names) using a <tt class="docutils literal"><span class="pre">lambda</span></tt>, a tiny function which in this case, tells me whether or not that sample id begins with the letter &#8220;S&#8221;.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">singles_ids</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;S&#39;</span><span class="p">))]</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">singles_ids</span><span class="p">))</span>
<span class="n">singles</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">]</span>

<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">singles</span><span class="p">[</span><span class="n">singles</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">expression_filtered</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">expression_filtered</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>
</pre></div>
</div>
<p>Hmm, that&#8217;s strange. The paper states that they had 6313 genes after filtering, but I get 6312. Even using &#8220;<cite>singles &gt;= 1</cite>&#8221; doesn&#8217;t help. I suppose it&#8217;s one of the usual bioinformatics 1-off errors :)</p>
<p>(I also tried this with the expression table provided in the supplementary data as &#8220;<cite>SupplementaryTable2.xlsx</cite>,&#8221; and got the same results.)</p>
<p>Now that we&#8217;ve taken care of importing and filtering the expression data, let&#8217;s do the feature data of the expression data.</p>
</div>
<div class="section" id="expression-feature-data">
<h2>Expression feature data<a class="headerlink" href="#expression-feature-data" title="Permalink to this headline">Â¶</a></h2>
<p>The &#8220;Expresion feature data&#8221; is similar to the <tt class="docutils literal"><span class="pre">fData</span></tt> from <tt class="docutils literal"><span class="pre">BioconductoR</span></tt>, where there&#8217;s some additional data on your features that you want to look at. They uploaded information about the features in their OTHER expression matrix, uploaded as a supplementary file, <tt class="docutils literal"><span class="pre">Supplementary_Table2.xlsx</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table2.xlsx&#39;</span><span class="p">)</span>

<span class="c"># # This was also in features x samples format, so we need to transpose</span>
<span class="c"># expression2 = expression2.T</span>
<span class="c"># expression2.head()</span>
</pre></div>
</div>
<p>Unfortunately, Excel mangled the gene names of the genes that started with a <tt class="docutils literal"><span class="pre">201</span></tt>, and converted those into <tt class="docutils literal"><span class="pre">datetime.datetime</span></tt> objects, or if the gene ID was all numbers, then it was converted to an integer. We could go in to excel save the file as a csv which would fix this, but for now we&#8217;ll just ignore those rows.</p>
<p>See, not all the items in the <tt class="docutils literal"><span class="pre">GENE</span></tt> column are <tt class="docutils literal"><span class="pre">unicode</span></tt> string objects:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="nb">set</span><span class="p">([</span><span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span><span class="p">])</span>
</pre></div>
</div>
<p>One is an integer:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>
</pre></div>
</div>
<p>Several more were converted to <tt class="docutils literal"><span class="pre">datetime</span></tt> objects.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">datetime</span>

<span class="nb">set</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="p">)])</span>
</pre></div>
</div>
<p>To ignore those columns, we&#8217;ll only use items in the <tt class="docutils literal"><span class="pre">GENE</span></tt> column if they are of the type <tt class="docutils literal"><span class="pre">unicode</span></tt>, as such:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span><span class="p">[</span><span class="n">expression2</span><span class="o">.</span><span class="n">GENE</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">unicode</span><span class="p">))]</span>
</pre></div>
</div>
<p>We can take a look at a subset and see that row 25&#8217;s gene was replaced with an NaN, indicating that the gene there wasn&#8217;t a string and didn&#8217;t pass our filter.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="mi">20</span><span class="p">:</span><span class="mi">25</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GENE</th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>20</th>
      <td>0610038B21RIK</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>21</th>
      <td>0610039K10RIK</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>22</th>
      <td>0610040B10RIK</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>23</th>
      <td>0610040J01RIK</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>24</th>
      <td>0910001L09RIK</td>
      <td>140.785306</td>
      <td>0.000000</td>
      <td>6.633041</td>
      <td>53.966101</td>
    </tr>
    <tr>
      <th>25</th>
      <td>NaN</td>
      <td>0.000000</td>
      <td>0.073505</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div><p>We can remove all the rows with an NaN using <tt class="docutils literal"><span class="pre">dropna</span></tt> and indicating we only want the function to pay attention to a <tt class="docutils literal"><span class="pre">subset</span></tt> of the columns. Let&#8217;s check the <tt class="docutils literal"><span class="pre">shape</span></tt> of the dataframe so we can see if there were tons of columns removed.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<p>Let&#8217;s look at the 20th to 25th rows again to see if the NaN gene is still there.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;GENE&#39;</span><span class="p">])</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="mi">20</span><span class="p">:</span><span class="mi">25</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GENE</th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>20</th>
      <td>0610038B21RIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>21</th>
      <td>0610039K10RIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>22</th>
      <td>0610040B10RIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>23</th>
      <td>0610040J01RIK</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>24</th>
      <td>0910001L09RIK</td>
      <td>140.785306</td>
      <td>0</td>
      <td>6.633041</td>
      <td>53.966101</td>
    </tr>
  </tbody>
</table>
</div><p>Nope, not there! How many columns did we end up removing?</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<p>Not many, only 27 total genes. Shouldn&#8217;t have a HUGE impact on the analysis. It&#8217;s not like we know what those genes do anyway...</p>
<p>Now let&#8217;s get down to doing what we came here to do: making a table of the feature data. We&#8217;ll use the column &#8220;Gene Category&#8221; in this dataframe, which we can see is in the last column by subsetting on the first five rows with <tt class="docutils literal"><span class="pre">:5</span></tt> and the last five columns with <tt class="docutils literal"><span class="pre">-5:</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span><span class="o">.</span><span class="n">ix</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">:]</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>S18</th>
      <th>10,000 cell Rep1 (P1)</th>
      <th>10,000 cell Rep2 (P2)</th>
      <th>10,000 cell Rep3 (P3)</th>
      <th>Gene Category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.031866</td>
      <td>11.067499</td>
      <td>12.466731</td>
      <td>7.190973</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>116.671522</td>
      <td>77.300963</td>
      <td>67.974030</td>
      <td>64.828913</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.000000</td>
      <td>4.893351</td>
      <td>6.623289</td>
      <td>6.600724</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.000000</td>
      <td>32.245673</td>
      <td>32.180601</td>
      <td>46.125662</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>209.268266</td>
      <td>121.227248</td>
      <td>128.939312</td>
      <td>113.595313</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div><p>We can subset <tt class="docutils literal"><span class="pre">expression2</span></tt> on the two columns we want: the gene ID in <tt class="docutils literal"><span class="pre">GENE</span></tt> and the gene category:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_category</span> <span class="o">=</span> <span class="n">expression2</span><span class="p">[[</span><span class="s">&#39;GENE&#39;</span><span class="p">,</span> <span class="s">&#39;Gene Category&#39;</span><span class="p">]]</span>
<span class="n">gene_category</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>GENE</th>
      <th>Gene Category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0610007L01RIK</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0610007P14RIK</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0610007P22RIK</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0610008F07RIK</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0610009B22RIK</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div><p>We want to create a <tt class="docutils literal"><span class="pre">pandas.DataFrame</span></tt> from the &#8220;Gene Category&#8221; row for our <tt class="docutils literal"><span class="pre">expression_feature_data</span></tt>, which we will do via:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression_feature_data</span> <span class="o">=</span> <span class="n">gene_category</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s">&#39;GENE&#39;</span><span class="p">)</span>
<span class="n">expression_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene Category</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0610007L01RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610009B22RIK</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
<div class="section" id="splicing-data">
<h2>Splicing Data<a class="headerlink" href="#splicing-data" title="Permalink to this headline">Â¶</a></h2>
<p>We obtain the splicing data from this study from the supplementary information, specifically the file, <tt class="docutils literal"><span class="pre">Supplementary_Table4.xls</span></tt></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table4.xls&#39;</span><span class="p">,</span> <span class="s">&#39;splicingTable.txt&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
      <th>S5</th>
      <th>S6</th>
      <th>S7</th>
      <th>S8</th>
      <th>S9</th>
      <th>S10</th>
      <th>S11</th>
      <th>S13</th>
      <th>S14</th>
      <th>S15</th>
      <th>S16</th>
      <th>S17</th>
      <th>S18</th>
      <th>10,000 cell Rep1 (P1)</th>
      <th>10,000 cell Rep2 (P2)</th>
      <th>10,000 cell Rep3 (P3)</th>
    </tr>
    <tr>
      <th>Event name</th>
      <th>gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>Os9</th>
      <td>0.84</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.03</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>NaN</td>
      <td>0.01</td>
      <td>NaN</td>
      <td>0.27</td>
      <td>0.37</td>
      <td>0.31</td>
    </tr>
    <tr>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>Vta1</th>
      <td>0.95</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.84</td>
      <td>0.95</td>
      <td>0.91</td>
      <td>0.87</td>
      <td>0.86</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.93</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.96</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.83</td>
      <td>0.85</td>
      <td>0.64</td>
    </tr>
    <tr>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>Bclaf1</th>
      <td>NaN</td>
      <td>0.04</td>
      <td>0.02</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.14</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.40</td>
      <td>0.49</td>
      <td>0.59</td>
    </tr>
    <tr>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>Bclaf1</th>
      <td>0.02</td>
      <td>0.98</td>
      <td>0.55</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.06</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.62</td>
      <td>0.63</td>
      <td>0.70</td>
    </tr>
    <tr>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>P4ha1</th>
      <td>0.42</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.94</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.03</td>
      <td>0.97</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.43</td>
      <td>0.36</td>
      <td>0.52</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">T</span>
<span class="n">splicing</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>Event name</th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
    <tr>
      <th>gene</th>
      <th>Os9</th>
      <th>Vta1</th>
      <th>Bclaf1</th>
      <th>Bclaf1</th>
      <th>P4ha1</th>
      <th>Bsg</th>
      <th>Ptbp1</th>
      <th>Igf1</th>
      <th>Elk3</th>
      <th>Nbr1</th>
      <th>...</th>
      <th>Afg3l1</th>
      <th>Tep1</th>
      <th>Fgd2</th>
      <th>Ttc17</th>
      <th>Tmed1</th>
      <th>Tmed1</th>
      <th>Sbno2</th>
      <th>Synj2</th>
      <th>Tbc1d13</th>
      <th>Usp47</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0.84</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.42</td>
      <td>NaN</td>
      <td>0.57</td>
      <td>0.31</td>
      <td>0.93</td>
      <td>0.57</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.55</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.20</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>NaN</td>
      <td>0.84</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.94</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.73</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S6</th>
      <td>0.01</td>
      <td>0.91</td>
      <td>0.14</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.61</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S7</th>
      <td>NaN</td>
      <td>0.87</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.62</td>
      <td>NaN</td>
      <td>0.85</td>
      <td>0.73</td>
      <td>0.55</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S8</th>
      <td>NaN</td>
      <td>0.86</td>
      <td>0.02</td>
      <td>0.98</td>
      <td>0.03</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.89</td>
      <td>0.82</td>
      <td>0.83</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S9</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.97</td>
      <td>NaN</td>
      <td>0.97</td>
      <td>NaN</td>
      <td>0.90</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S10</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.06</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S11</th>
      <td>0.03</td>
      <td>0.93</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.97</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S13</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S14</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.88</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S15</th>
      <td>0.02</td>
      <td>0.96</td>
      <td>0.01</td>
      <td>0.06</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.44</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>0.91</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S16</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>0.27</td>
      <td>0.99</td>
      <td>0.99</td>
      <td>0.98</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S17</th>
      <td>0.01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.96</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.99</td>
      <td>0.98</td>
      <td>0.67</td>
      <td>0.07</td>
    </tr>
    <tr>
      <th>S18</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10,000 cell Rep1 (P1)</th>
      <td>0.27</td>
      <td>0.83</td>
      <td>0.40</td>
      <td>0.62</td>
      <td>0.43</td>
      <td>0.78</td>
      <td>NaN</td>
      <td>0.60</td>
      <td>0.76</td>
      <td>0.52</td>
      <td>...</td>
      <td>0.92</td>
      <td>NaN</td>
      <td>0.81</td>
      <td>0.77</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.84</td>
      <td>0.50</td>
      <td>0.56</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>10,000 cell Rep2 (P2)</th>
      <td>0.37</td>
      <td>0.85</td>
      <td>0.49</td>
      <td>0.63</td>
      <td>0.36</td>
      <td>0.72</td>
      <td>0.47</td>
      <td>0.60</td>
      <td>0.73</td>
      <td>0.68</td>
      <td>...</td>
      <td>0.67</td>
      <td>0.15</td>
      <td>0.52</td>
      <td>0.67</td>
      <td>0.63</td>
      <td>0.73</td>
      <td>0.82</td>
      <td>0.90</td>
      <td>0.71</td>
      <td>0.55</td>
    </tr>
    <tr>
      <th>10,000 cell Rep3 (P3)</th>
      <td>0.31</td>
      <td>0.64</td>
      <td>0.59</td>
      <td>0.70</td>
      <td>0.52</td>
      <td>0.79</td>
      <td>NaN</td>
      <td>0.65</td>
      <td>0.42</td>
      <td>0.64</td>
      <td>...</td>
      <td>0.58</td>
      <td>0.79</td>
      <td>0.74</td>
      <td>0.85</td>
      <td>0.73</td>
      <td>0.39</td>
      <td>0.56</td>
      <td>NaN</td>
      <td>0.64</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>20 rows Ã 352 columns</p>
</div><p>The three pooled samples aren&#8217;t named consistently with the expression data, so we have to fix that.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">&#39;P&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">)]</span>
</pre></div>
</div>
<p>Since the pooled sample IDs are inconsistent with the <cite>expression</cite> data, we have to change them. We can get the &#8220;P&#8221; and the number after that using regular expressions, called <tt class="docutils literal"><span class="pre">re</span></tt> in the Python standard library, e.g.:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">re</span>
<span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="s">&#39;10,000 cell Rep1 (P1)&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">long_pooled_name_to_short</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="k">if</span> <span class="s">&#39;P&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">x</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>


<span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
</pre></div>
</div>
<p>And now we assign this new index as our index to the <tt class="docutils literal"><span class="pre">splicing</span></tt> dataframe.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>Event name</th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
    <tr>
      <th>gene</th>
      <th>Os9</th>
      <th>Vta1</th>
      <th>Bclaf1</th>
      <th>Bclaf1</th>
      <th>P4ha1</th>
      <th>Bsg</th>
      <th>Ptbp1</th>
      <th>Igf1</th>
      <th>Elk3</th>
      <th>Nbr1</th>
      <th>...</th>
      <th>Afg3l1</th>
      <th>Tep1</th>
      <th>Fgd2</th>
      <th>Ttc17</th>
      <th>Tmed1</th>
      <th>Tmed1</th>
      <th>Sbno2</th>
      <th>Synj2</th>
      <th>Tbc1d13</th>
      <th>Usp47</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0.84</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.42</td>
      <td>NaN</td>
      <td>0.57</td>
      <td>0.31</td>
      <td>0.93</td>
      <td>0.57</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.55</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.20</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>NaN</td>
      <td>0.84</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.94</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.73</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã 352 columns</p>
</div><p>Right now, <tt class="docutils literal"><span class="pre">flotilla</span></tt> doesn&#8217;t work with multi-indexed dataframes, so to deal with this, we&#8217;ll use <tt class="docutils literal"><span class="pre">droplevel</span></tt> from <tt class="docutils literal"><span class="pre">pandas</span></tt> which will remove the gene name &#8220;level&#8221; from the column names. This is because the <tt class="docutils literal"><span class="pre">chr10:.....</span></tt>-type thing is a unique identifier for splicing events, while the gene name is not (one gene may have many splicing events)</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">droplevel</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Event name</th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0.84</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.42</td>
      <td>NaN</td>
      <td>0.57</td>
      <td>0.31</td>
      <td>0.93</td>
      <td>0.57</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>0.98</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.02</td>
      <td>0.55</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.20</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>NaN</td>
      <td>0.84</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>0.04</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>NaN</td>
      <td>0.95</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.94</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>0.73</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã 352 columns</p>
</div><p>Great, now we have a perfectly formatted splicing dataframe!</p>
<div class="section" id="metadata">
<h3>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">Â¶</a></h3>
<p>Now let&#8217;s get into creating a metadata dataframe. We&#8217;ll use the index from the <tt class="docutils literal"><span class="pre">expression_filtered</span></tt> data to create the minimum required column, <tt class="docutils literal"><span class="pre">'phenotype'</span></tt>, which has the name of the phenotype of that cell. And we&#8217;ll also add the column <tt class="docutils literal"><span class="pre">'pooled'</span></tt> to indicate whether this sample is pooled or not.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">expression_filtered</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;BDMC&#39;</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;pooled&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;P&#39;</span><span class="p">))</span>

<span class="n">metadata</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
      <th>pooled</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S6</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S7</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S8</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S9</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S10</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S11</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S12</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S13</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S14</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S15</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S16</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S17</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>S18</th>
      <td>BDMC</td>
      <td>False</td>
    </tr>
    <tr>
      <th>P1</th>
      <td>BDMC</td>
      <td>True</td>
    </tr>
    <tr>
      <th>P2</th>
      <td>BDMC</td>
      <td>True</td>
    </tr>
    <tr>
      <th>P3</th>
      <td>BDMC</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="section" id="mapping-stats-data">
<h2>Mapping stats data<a class="headerlink" href="#mapping-stats-data" title="Permalink to this headline">Â¶</a></h2>
<p>We also want to include the information about how well each sample&#8217;s reads mapped to the genome, which the authors included as <tt class="docutils literal"><span class="pre">Supplementary_Table1.xls</span></tt></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table1.xls&#39;</span><span class="p">,</span> <span class="n">sheetname</span><span class="o">=</span><span class="s">&#39;SuppTable1 2.txt&#39;</span><span class="p">)</span>
<span class="n">mapping_stats</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>PF_READS</th>
      <th>PCT_MAPPED_GENOME</th>
      <th>PCT_RIBOSOMAL_BASES</th>
      <th>MEDIAN_CV_COVERAGE</th>
      <th>MEDIAN_5PRIME_BIAS</th>
      <th>MEDIAN_3PRIME_BIAS</th>
      <th>MEDIAN_5PRIME_TO_3PRIME_BIAS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>S1</td>
      <td>21326048</td>
      <td>0.706590</td>
      <td>0.006820</td>
      <td>0.509939</td>
      <td>0.092679</td>
      <td>0.477321</td>
      <td>0.247741</td>
    </tr>
    <tr>
      <th>1</th>
      <td>S2</td>
      <td>27434011</td>
      <td>0.745385</td>
      <td>0.004111</td>
      <td>0.565732</td>
      <td>0.056583</td>
      <td>0.321053</td>
      <td>0.244062</td>
    </tr>
    <tr>
      <th>2</th>
      <td>S3</td>
      <td>31142391</td>
      <td>0.722087</td>
      <td>0.006428</td>
      <td>0.540341</td>
      <td>0.079551</td>
      <td>0.382286</td>
      <td>0.267367</td>
    </tr>
    <tr>
      <th>3</th>
      <td>S4</td>
      <td>26231852</td>
      <td>0.737854</td>
      <td>0.004959</td>
      <td>0.530978</td>
      <td>0.067041</td>
      <td>0.351670</td>
      <td>0.279782</td>
    </tr>
    <tr>
      <th>4</th>
      <td>S5</td>
      <td>29977214</td>
      <td>0.746466</td>
      <td>0.006121</td>
      <td>0.525598</td>
      <td>0.066543</td>
      <td>0.353995</td>
      <td>0.274252</td>
    </tr>
    <tr>
      <th>5</th>
      <td>S6</td>
      <td>24148387</td>
      <td>0.730079</td>
      <td>0.008794</td>
      <td>0.529650</td>
      <td>0.072095</td>
      <td>0.413696</td>
      <td>0.225929</td>
    </tr>
    <tr>
      <th>6</th>
      <td>S7</td>
      <td>24078116</td>
      <td>0.730638</td>
      <td>0.007945</td>
      <td>0.540913</td>
      <td>0.051991</td>
      <td>0.358597</td>
      <td>0.201984</td>
    </tr>
    <tr>
      <th>7</th>
      <td>S8</td>
      <td>25032126</td>
      <td>0.739989</td>
      <td>0.004133</td>
      <td>0.512725</td>
      <td>0.058783</td>
      <td>0.373509</td>
      <td>0.212337</td>
    </tr>
    <tr>
      <th>8</th>
      <td>S9</td>
      <td>22257682</td>
      <td>0.747427</td>
      <td>0.004869</td>
      <td>0.521622</td>
      <td>0.063566</td>
      <td>0.334294</td>
      <td>0.240641</td>
    </tr>
    <tr>
      <th>9</th>
      <td>S10</td>
      <td>29436289</td>
      <td>0.748795</td>
      <td>0.005499</td>
      <td>0.560454</td>
      <td>0.036219</td>
      <td>0.306729</td>
      <td>0.187479</td>
    </tr>
    <tr>
      <th>10</th>
      <td>S11</td>
      <td>31130278</td>
      <td>0.741882</td>
      <td>0.002740</td>
      <td>0.558882</td>
      <td>0.049581</td>
      <td>0.349191</td>
      <td>0.211787</td>
    </tr>
    <tr>
      <th>11</th>
      <td>S12</td>
      <td>21161595</td>
      <td>0.750782</td>
      <td>0.006837</td>
      <td>0.756339</td>
      <td>0.013878</td>
      <td>0.324264</td>
      <td>0.195430</td>
    </tr>
    <tr>
      <th>12</th>
      <td>S13</td>
      <td>28612833</td>
      <td>0.733976</td>
      <td>0.011718</td>
      <td>0.598687</td>
      <td>0.035392</td>
      <td>0.357447</td>
      <td>0.198566</td>
    </tr>
    <tr>
      <th>13</th>
      <td>S14</td>
      <td>26351189</td>
      <td>0.748323</td>
      <td>0.004106</td>
      <td>0.517518</td>
      <td>0.070293</td>
      <td>0.381095</td>
      <td>0.259122</td>
    </tr>
    <tr>
      <th>14</th>
      <td>S15</td>
      <td>25739575</td>
      <td>0.748421</td>
      <td>0.003353</td>
      <td>0.526238</td>
      <td>0.050938</td>
      <td>0.324207</td>
      <td>0.212366</td>
    </tr>
    <tr>
      <th>15</th>
      <td>S16</td>
      <td>26802346</td>
      <td>0.739833</td>
      <td>0.009370</td>
      <td>0.520287</td>
      <td>0.071503</td>
      <td>0.358758</td>
      <td>0.240009</td>
    </tr>
    <tr>
      <th>16</th>
      <td>S17</td>
      <td>26343522</td>
      <td>0.749358</td>
      <td>0.003155</td>
      <td>0.673195</td>
      <td>0.024121</td>
      <td>0.301588</td>
      <td>0.245854</td>
    </tr>
    <tr>
      <th>17</th>
      <td>S18</td>
      <td>25290073</td>
      <td>0.749358</td>
      <td>0.007465</td>
      <td>0.562382</td>
      <td>0.048528</td>
      <td>0.314776</td>
      <td>0.215160</td>
    </tr>
    <tr>
      <th>18</th>
      <td>10k_rep1</td>
      <td>28247826</td>
      <td>0.688553</td>
      <td>0.018993</td>
      <td>0.547000</td>
      <td>0.056113</td>
      <td>0.484393</td>
      <td>0.140333</td>
    </tr>
    <tr>
      <th>19</th>
      <td>10k_rep2</td>
      <td>39303876</td>
      <td>0.690313</td>
      <td>0.017328</td>
      <td>0.547621</td>
      <td>0.055600</td>
      <td>0.474634</td>
      <td>0.142889</td>
    </tr>
    <tr>
      <th>20</th>
      <td>10k_rep3</td>
      <td>29831281</td>
      <td>0.710875</td>
      <td>0.010610</td>
      <td>0.518053</td>
      <td>0.066053</td>
      <td>0.488738</td>
      <td>0.168180</td>
    </tr>
    <tr>
      <th>21</th>
      <td>MB_SC1</td>
      <td>13848219</td>
      <td>0.545000</td>
      <td>0.007000</td>
      <td>0.531495</td>
      <td>0.127934</td>
      <td>0.207841</td>
      <td>0.728980</td>
    </tr>
    <tr>
      <th>22</th>
      <td>MB_SC2</td>
      <td>13550218</td>
      <td>0.458000</td>
      <td>0.010800</td>
      <td>0.569271</td>
      <td>0.102581</td>
      <td>0.179407</td>
      <td>0.694747</td>
    </tr>
    <tr>
      <th>23</th>
      <td>MB_SC3</td>
      <td>26765848</td>
      <td>0.496000</td>
      <td>0.007900</td>
      <td>0.535192</td>
      <td>0.141893</td>
      <td>0.231068</td>
      <td>0.722080</td>
    </tr>
  </tbody>
</table>
</div></div>
<div class="section" id="create-a-flotilla-study">
<h2>Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!<a class="headerlink" href="#create-a-flotilla-study" title="Permalink to this headline">Â¶</a></h2>
<p>We&#8217;ve now prepared five (5!) different dataframes of information that together represent this body of work. They are:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">metadata</span></tt>: A table describing each individual sample</li>
<li><tt class="docutils literal"><span class="pre">expression_filtered</span></tt>: A big matrix of gene expression values for each sample</li>
<li><tt class="docutils literal"><span class="pre">expression_feature_data</span></tt>: A table describing metadata on the genes, e.g. the category of the gene</li>
<li><tt class="docutils literal"><span class="pre">splicing_data</span></tt>: A big matrix of &#8220;percent spliced-in&#8221; splicing scores for each sample</li>
<li><tt class="docutils literal"><span class="pre">mapping_stats_data</span></tt>: A table describing the percentage of reads that mapped to the genome, plus other information about where the genes mapped, for each sample</li>
</ul>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">Study</span><span class="p">(</span><span class="c"># The metadata describing phenotype and pooled samples</span>
                       <span class="n">metadata</span><span class="p">,</span>

                       <span class="c"># A version for this data</span>
                       <span class="n">version</span><span class="o">=</span><span class="s">&#39;0.1.0&#39;</span><span class="p">,</span>

                       <span class="c"># Dataframe of the filtered expression data</span>
                       <span class="n">expression_data</span><span class="o">=</span><span class="n">expression_filtered</span><span class="p">,</span>

                       <span class="c"># Dataframe of the feature data of the genes</span>
                       <span class="n">expression_feature_data</span><span class="o">=</span><span class="n">expression_feature_data</span><span class="p">,</span>

                       <span class="c"># Dataframe of the splicing data</span>
                       <span class="n">splicing_data</span><span class="o">=</span><span class="n">splicing</span><span class="p">,</span>

                       <span class="c"># Dataframe of the mapping stats data</span>
                       <span class="n">mapping_stats_data</span><span class="o">=</span><span class="n">mapping_stats</span><span class="p">,</span>

                       <span class="c"># Which column in &quot;mapping_stats&quot; has the number of reads</span>
                       <span class="n">mapping_stats_number_mapped_col</span><span class="o">=</span><span class="s">&#39;PF_READS&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>2015-05-01 18:26:32 Initializing Study
2015-05-01 18:26:32 Initializing Predictor configuration manager for Study
2015-05-01 18:26:32 Predictor ExtraTreesClassifier is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesClassifier&#39;&gt;
2015-05-01 18:26:32 Added ExtraTreesClassifier to default predictors
2015-05-01 18:26:32 Predictor ExtraTreesRegressor is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesRegressor&#39;&gt;
2015-05-01 18:26:32 Added ExtraTreesRegressor to default predictors
2015-05-01 18:26:32 Predictor GradientBoostingClassifier is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingClassifier&#39;&gt;
2015-05-01 18:26:32 Added GradientBoostingClassifier to default predictors
2015-05-01 18:26:32 Predictor GradientBoostingRegressor is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingRegressor&#39;&gt;
2015-05-01 18:26:32 Added GradientBoostingRegressor to default predictors
2015-05-01 18:26:32 Loading metadata
2015-05-01 18:26:32 Loading expression data
2015-05-01 18:26:32 Initializing expression
2015-05-01 18:26:32 Done initializing expression
2015-05-01 18:26:32 Loading splicing data
2015-05-01 18:26:32 Initializing splicing
2015-05-01 18:26:32 Done initializing splicing
2015-05-01 18:26:33 Successfully initialized a Study object!
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>No phenotype to color mapping was provided, falling back on reasonable defaults.
No phenotype to marker (matplotlib plotting symbol) was provided, falling back on reasonable defaults.
</pre></div>
</div>
<p>As a side note, you can save this study to disk now, so you can &#8220;<tt class="docutils literal"><span class="pre">embark</span></tt>&#8221; (load it) later:</p>
</div>
<div class="section" id="save-the-study-so-you-can-load-it-later">
<h2>Save the study so you can load it later<a class="headerlink" href="#save-the-study-so-you-can-load-it-later" title="Permalink to this headline">Â¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Wrote datapackage to /home/travis/flotilla_projects/shalek2013/datapackage.json
</pre></div>
</div>
<p>Note that this is saved to my home directory, in <tt class="docutils literal"><span class="pre">~/flotilla_projects/&lt;study_name&gt;/</span></tt>. This will be saved in your home directory, too.</p>
<p>The <tt class="docutils literal"><span class="pre">datapackage.json</span></tt> file is what holds all the information relative to the study, and loosely follows the <a class="reference external" href="http://data.okfn.org/doc/data-package">datapackage spec</a> created by the Open Knowledge Foundation.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cat</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span><span class="o">/</span><span class="n">datapackage</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>cat: /Users/olga/flotilla_projects/shalek2013/datapackage.json: No such file or directory
</pre></div>
</div>
<p>One thing to note is that when you save, the version number is bumped up. <tt class="docutils literal"><span class="pre">study.version</span></tt> (the one we just made) is <tt class="docutils literal"><span class="pre">0.1.0</span></tt>, but the one we saved is <tt class="docutils literal"><span class="pre">0.1.1</span></tt>, since we could have made some changes to the data.</p>
<p>Let&#8217;s look at what else is in this folder:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ls</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>ls: cannot access /Users/olga/flotilla_projects/shalek2013: No such file or directory
</pre></div>
</div>
<p>So this is where all the other files are. Good to know!</p>
<p>We can &#8220;embark&#8221; on this newly-saved study now very painlessly, without having to open and process all those files again:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study2</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">embark</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>2015-05-01 18:26:34 Reading datapackage from /home/travis/flotilla_projects/shalek2013/datapackage.json
2015-05-01 18:26:34 Parsing datapackage to create a Study object
2015-05-01 18:26:35 Initializing Study
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>2015-05-01 18:26:35 Initializing Predictor configuration manager for Study
2015-05-01 18:26:35 Predictor ExtraTreesClassifier is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesClassifier&#39;&gt;
2015-05-01 18:26:35 Added ExtraTreesClassifier to default predictors
2015-05-01 18:26:35 Predictor ExtraTreesRegressor is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesRegressor&#39;&gt;
2015-05-01 18:26:35 Added ExtraTreesRegressor to default predictors
2015-05-01 18:26:35 Predictor GradientBoostingClassifier is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingClassifier&#39;&gt;
2015-05-01 18:26:35 Added GradientBoostingClassifier to default predictors
2015-05-01 18:26:35 Predictor GradientBoostingRegressor is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingRegressor&#39;&gt;
2015-05-01 18:26:35 Added GradientBoostingRegressor to default predictors
2015-05-01 18:26:35 Loading metadata
2015-05-01 18:26:35 Loading expression data
2015-05-01 18:26:35 Initializing expression
2015-05-01 18:26:35 Done initializing expression
2015-05-01 18:26:35 Loading splicing data
2015-05-01 18:26:35 Initializing splicing
2015-05-01 18:26:35 Done initializing splicing
2015-05-01 18:26:35 Successfully initialized a Study object!
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span> <span class="o">==</span> <span class="n">study2</span>
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../_sources/tutorial/getting_started_shalek2013.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <p>
    Stuck? Send questions, comments, and other feedback to:
    <ul>
        <li>The <a href="https://gitter.im/YeoLab/flotilla">Gitter</a> chat
            room for immediate questions</li>
        <li>The <a href="https://github.com/yeolab/flotilla/issues">Github
            issues page</a> for feature requests</li>
    </ul>
    Or send an email to obotvinn@ucsd.edu if you're oldskool like that.
    </p>

    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

  </body>
</html>