<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>&lt;no title&gt; &mdash; flotilla 0.2.8dev documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.8dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.8dev documentation" href="../index.html" />
    <link rel="up" title="Flotilla tutorial" href="../tutorial.html" />
    <link rel="prev" title="OS X Installation instructions" href="../installation.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.8dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Flotilla tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#getting-started-downloading-cleaning-and-importing-data">Getting Started: Downloading, cleaning, and importing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#exploratory-analysis">Exploratory Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul class="simple">
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <p>[BrainSpan](<a class="reference external" href="http://www.brainspan.org/">http://www.brainspan.org/</a>) is an interface to data from the developing human brain. They have samples extracted from many post-mortem brains, and the dataset I&#8217;m most interested in is the developmental transcriptome, which has tissues collected from 42 donors of ages ranging from post conception week 8 to 40 years old. This is an incredible resource for anyone interested in studying changes in gene expression within brain regions across developmental time.</p>
<p>One major issue is the &#8220;batch effects&#8221; that occur and skew the data because of factors unrelated to the biological phenomenon. For example, each donor was could have been collected at a different time, at different post-mortem times, from either fresh or frozen tissue, collected at different facilities, sectioned by a different scientist (who has their own unique style of dissecting tissue), and prepared tissue samples for sequencing libraries for RNA-sequencing differently.</p>
<p>For BrainSpan, they have a detailed [white paper](<a class="reference external" href="http://help.brain-map.org/download/attachments/3506181/Transcriptome_Profiling.pdf">http://help.brain-map.org/download/attachments/3506181/Transcriptome_Profiling.pdf</a>) describing every step of their process, and they&#8217;ve controlled as many variables as they could, which is everything downstream of sectioning the tissue. Different people could have sectioned the different donor&#8217;s tissues, and some tissues were frozen (which had to be thawed) and some were fresh (which had to be chilled), so there&#8217;s a lot of factors that could affect the results that have nothing to do with the actual biology of the human brain.</p>
<p>A commonly used software package for adjusting for batch effects is called ComBat, originally designed for microarray batches ([Johnson and Li, <em>Biostatistics</em> 2007]((<a class="reference external" href="http://biostatistics.oxfordjournals.org/content/8/1/118.abstract">http://biostatistics.oxfordjournals.org/content/8/1/118.abstract</a>)). Their software is available as part of an R package called [<cite>sva</cite>](<a class="reference external" href="http://www.bioconductor.org/packages/release/bioc/html/sva.html">http://www.bioconductor.org/packages/release/bioc/html/sva.html</a>) on [BioconductoR](<a class="reference external" href="http://www.bioconductor.org/">http://www.bioconductor.org/</a>). I wanted to use it for this BrainSpan data, because the donor-specific signals are much stronger than the tissue-specific signals, but I knew I&#8217;d have to use <cite>rpy2</cite> to call R and didn&#8217;t really want to do that. Thankfully, an [awesome human being](<a class="reference external" href="https://github.com/brentp">https://github.com/brentp</a>) ported the contents of <cite>combat</cite> to python to create [<cite>combat.py</cite>](<a class="reference external" href="https://github.com/brentp/combat.py">https://github.com/brentp/combat.py</a>)!</p>
<p>In this blog post, I will use the BrainSpan data as an example of how to use <cite>combat.py</cite>. As a bonus, this feature has been added to the large-scale RNA-seq analysis software [<cite>flotilla</cite>](<a class="reference external" href="https://github.com/yeolab/flotilla">https://github.com/yeolab/flotilla</a>) so while you can do all this work yourself, you could also just tell <cite>flotilla</cite> to do it for you!</p>
<p>Downloading the data is fairly straightforward, it&#8217;s all contained in a zip archive online. This is the <em>&#8220;RNA-Seq Gencode v10 summarized to genes&#8221;</em> download link from BrainSpan&#8217;s [Downloads](<a class="reference external" href="http://www.brainspan.org/static/download.html">http://www.brainspan.org/static/download.html</a>) page.</p>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../_sources/tutorial/getting_started_brainspan.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <p>
    Stuck? Send questions, comments, and other feedback to:
    <ul>
        <li>The <a href="https://gitter.im/YeoLab/flotilla">Gitter</a> chat
            room for immediate questions</li>
        <li>The <a href="https://github.com/yeolab/flotilla/issues">Github
            issues page</a> for feature requests</li>
    </ul>
    Or send an email to obotvinn@ucsd.edu if you're oldskool like that.
    </p>

    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

  </body>
</html>