<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Accurate estimation of alternative splicing modalities &mdash; flotilla 0.2.8dev documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.2.0/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.8dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.8dev documentation" href="../index.html" />
    <link rel="up" title="Flotilla tutorial" href="../tutorial.html" />
    <link rel="prev" title="Storing supplemental data on Study objects" href="supplementaldata.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.8dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Flotilla tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#getting-started-downloading-cleaning-and-importing-data">Getting Started: Downloading, cleaning, and importing data</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#exploratory-analysis">Exploratory Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Accurate estimation of alternative splicing modalities</a><ul>
<li><a class="reference internal" href="#parameterizations-of-bayesian-model">Parameterizations of Bayesian model</a></li>
<li><a class="reference internal" href="#prior-probability-on-parameterizations">Prior probability on parameterizations</a></li>
<li><a class="reference internal" href="#testing-modality-estimation">Testing modality estimation</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <div class="section" id="accurate-estimation-of-alternative-splicing-modalities">
<span id="reproduce-shalek2013"></span><h1>Accurate estimation of alternative splicing modalities<a class="headerlink" href="#accurate-estimation-of-alternative-splicing-modalities" title="Permalink to this headline">Â¶</a></h1>
<p>As shown in previous work (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Shalek+Satija et al 2013</a>), the majority of alternative splicing in single cells is all or nothing - either the included isoform or the excluded isoform. Rarely are there genes whose included/excluded isoforms are held at a 50/50 ratio. However, it is unclear what fraction of transcripts are maintained at specific proportions across cells, and we aim to categorize these splicing events into five bins: included, middle, excluded, bimodal, and uniform.</p>
<p>First, let&#8217;s import the modules we&#8217;ll need.</p>
<div class="code python highlight-python"><div class="highlight"><pre>%matplotlib inline

import os

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
sns.set(context=&#39;notebook&#39;, style=&#39;ticks&#39;)

from flotilla.compute.splicing import ModalityEstimator, ModalityModel
from flotilla.visualize.splicing import ModalitiesViz
</pre></div>
</div>
<pre class='nb-text-output'>
:0: FutureWarning: IPython widgets are experimental and may change in the future.

</pre><p>Next, we will initialize a <tt class="docutils literal"><span class="pre">ModalityEstimator</span></tt> for performing the
calculations, and a <tt class="docutils literal"><span class="pre">ModalityViz</span></tt> for visualizing the results.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">modality_estimator</span> <span class="o">=</span> <span class="n">ModalityEstimator</span><span class="p">(</span><span class="n">step</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">modality_visualizer</span> <span class="o">=</span> <span class="n">ModalitiesViz</span><span class="p">()</span>
</pre></div>
</div>
<p>Next, let&#8217;s plot the modalities we are modeling as <a class="reference external" href="http://en.wikipedia.org/wiki/Violin_plot">violin plots</a>. We model splicing events&#8217; &#8220;percent spliced-in&#8221; or Psi/$Psi$ as following a <a class="reference external" href="http://en.wikipedia.org/wiki/Beta_distribution">Beta distribution</a>:</p>
<p>$$
begin{align}
Psi &amp;sim text{Beta}left(alpha, betaright)\
&amp;= frac{1}{B(alpha, beta)} x^{alpha-1}(1-x)^{beta-1}
end{align}
$$</p>
<p>Where $B(alpha, beta)$ is the [Beta <em>function</em>](<a class="reference external" href="http://en.wikipedia.org/wiki/Beta_function">http://en.wikipedia.org/wiki/Beta_function</a>), from which the Beta <em>distribution</em> gets its name,</p>
<p>$$
Bleft(alpha, betaright) = int_0^1 t^{alpha-1}(1-t)^{beta-1}dt
$$</p>
<p>The Beta distribution is perfect for splicing scores because ...</p>
<ol class="arabic simple">
<li>It is continuous between 0 and 1, so any and all new splicing events we encounter are always valid by the model</li>
<li>Its parameters, $alpha$ and $beta$ can be tuned to create the characteristic modalities we are interested in</li>
</ol>
<p>Specifically, we can model the <em>included</em>, <em>middle</em>, <em>excluded</em>, or <em>bimodal</em> modalities. In the plot below, note that only the first four are actually in the model, and that the uniform distribution was separately created. This is because the uniform distribution cannot be parameterized, it is exactly when $alpha=1$ and $beta=1$, and as every splicing event has exactly the same probability under the uniform distribution, $Pr_{text{Uniform}}(text{event})=1$, we will use the uniform distribution as our null model for testing.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

<span class="n">n</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="c"># Use the last parameterization (we&#39;ll get to that next) in the model list of random variables &quot;rvs&quot;</span>
<span class="n">model_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="n">model_name</span><span class="p">:</span> <span class="n">model</span><span class="o">.</span><span class="n">rvs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">for</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()})</span>
<span class="n">model_params</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">({</span><span class="n">model_name</span><span class="p">:</span> <span class="n">model</span><span class="o">.</span><span class="n">rvs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">args</span> <span class="k">for</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()})</span>

<span class="c"># Add the uniform distribution</span>
<span class="n">model_data</span><span class="p">[</span><span class="s">&#39;uniform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>

<span class="c"># Reorder the columns</span>
<span class="n">columns_ordered</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;included&#39;</span><span class="p">,</span> <span class="s">&#39;middle&#39;</span><span class="p">,</span> <span class="s">&#39;bimodal&#39;</span><span class="p">,</span> <span class="s">&#39;excluded&#39;</span><span class="p">,</span> <span class="s">&#39;uniform&#39;</span><span class="p">]</span>
<span class="n">model_data</span> <span class="o">=</span> <span class="n">model_data</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns_ordered</span><span class="p">)</span>

<span class="c"># Get modality colors</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">modality_visualizer</span><span class="o">.</span><span class="n">modality_colors</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">model_data</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

<span class="c"># Rename the columns to include the alpha and beta parameters</span>
<span class="n">modality_with_params_renamer</span> <span class="o">=</span> <span class="p">{</span><span class="n">model_name</span><span class="p">:</span> <span class="s">&#39;{}</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">alpha={:.2f}$</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">beta={:.2f}$&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span>
 <span class="k">for</span> <span class="n">model_name</span><span class="p">,</span> <span class="p">(</span><span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span><span class="p">)</span> <span class="ow">in</span> <span class="n">model_params</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()}</span>
<span class="n">modality_with_params_renamer</span><span class="p">[</span><span class="s">&#39;uniform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;uniform</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">alpha=1$</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">beta=1$&#39;</span>
<span class="n">model_data</span> <span class="o">=</span> <span class="n">model_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">modality_with_params_renamer</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">model_data</span><span class="p">,</span> <span class="n">bw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;$\Psi$&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
<pre class='nb-text-output'>
/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/matplotlib/font_manager.py:1282: UserWarning: findfont: Font family [u'Helvetica', u'Arial'] not found. Falling back to Bitstream Vera Sans
  (prop.get_family(), self.defaultFamily[fontext]))

</pre><img alt="../_images/modalities_7_1.png" src="../_images/modalities_7_1.png" />
<div class="section" id="parameterizations-of-bayesian-model">
<h2>Parameterizations of Bayesian model<a class="headerlink" href="#parameterizations-of-bayesian-model" title="Permalink to this headline">Â¶</a></h2>
<p>We will employ a Bayesian model to do a (very non-exhaustive) &#8220;grid search&#8221; of parameter space for splicing events. Meaning, given a particular splicing event, does it match one of these modalities&#8217; parameterizations?</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ncols</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="p">[</span><span class="s">&#39;included&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">alphas</span><span class="p">)</span>
<span class="n">nrows</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">ncols</span><span class="p">,</span> <span class="mi">4</span><span class="o">*</span><span class="n">nrows</span><span class="p">))</span>

<span class="n">n</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">model</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">parameterizations</span> <span class="o">=</span> <span class="p">[</span><span class="n">rv</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">for</span> <span class="n">rv</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">rvs</span><span class="p">]</span>
    <span class="n">color</span> <span class="o">=</span> <span class="n">modality_visualizer</span><span class="o">.</span><span class="n">modality_colors</span><span class="p">[</span><span class="n">model_name</span><span class="p">]</span>
    <span class="n">xticklabels</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;$</span><span class="se">\\</span><span class="s">alpha={:.3f}$</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">beta={:.3f}$&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">alphas</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">betas</span><span class="p">)]</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">parameterizations</span><span class="p">,</span> <span class="n">bw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">xticklabels</span><span class="p">,</span> <span class="c">#color=colors,</span>
                   <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;$\Psi$&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/modalities_10_0.png" src="../_images/modalities_10_0.png" />
</div>
<div class="section" id="prior-probability-on-parameterizations">
<h2>Prior probability on parameterizations<a class="headerlink" href="#prior-probability-on-parameterizations" title="Permalink to this headline">Â¶</a></h2>
<p>We use a uniform prior on all parameterizations except the <em>bimodal</em>
distribution. This is because through extensive testing, we have found
that using a uniform prior on the <em>bimodal</em> distribution causes slightly
noisy (+10% uniform noise) <em>included</em> and <em>excluded</em> splicing events to
be called &#8220;bimodal,&#8221; and we want to eliminate false postives in the
<em>bimodal</em> category.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">ncols</span><span class="p">,</span> <span class="mi">4</span><span class="o">*</span><span class="n">nrows</span><span class="p">),</span> <span class="n">sharey</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="n">n</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="n">n_params</span> <span class="o">=</span> <span class="mi">10</span>

<span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">model</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">color</span> <span class="o">=</span> <span class="n">modality_visualizer</span><span class="o">.</span><span class="n">modality_colors</span><span class="p">[</span><span class="n">model_name</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">model_name</span> <span class="o">==</span> <span class="s">&#39;bimodal&#39;</span><span class="p">:</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">parameters</span><span class="p">)))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">ymax</span><span class="o">/</span><span class="n">ymax</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">parameters</span><span class="p">))</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">parameters</span><span class="p">)</span>
    <span class="n">ymin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">ymax</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">ymax</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ymax</span><span class="p">,</span> <span class="s">&#39;o-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">xticklabels</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;$</span><span class="se">\\</span><span class="s">alpha={:.3f}$</span><span class="se">\n</span><span class="s">$</span><span class="se">\\</span><span class="s">beta={:.3f}$&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">alphas</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">betas</span><span class="p">)]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Prior probability&#39;</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xticklabels</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>

<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/modalities_13_0.png" src="../_images/modalities_13_0.png" />
</div>
<div class="section" id="testing-modality-estimation">
<h2>Testing modality estimation<a class="headerlink" href="#testing-modality-estimation" title="Permalink to this headline">Â¶</a></h2>
<p>To show that this framework is accurate, we start with a &#8220;perfect&#8221; event, which uses the last parameterization of each model, which is the &#8220;most extreme&#8221; version of that model under our assumptions.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">n</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">event</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">inclusion_model</span><span class="o">.</span><span class="n">rvs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>


<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s">&#39;ticks&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">event</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">bw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">modality_visualizer</span><span class="o">.</span><span class="n">modality_colors</span><span class="p">[</span><span class="s">&#39;included&#39;</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/modalities_16_0.png" src="../_images/modalities_16_0.png" />
<p>Let&#8217;s estimate the modality of this event by calculating the log-likelihood of each parameterization using <tt class="docutils literal"><span class="pre">modality_estimator._loglik</span></tt>, and summing the results using <tt class="docutils literal"><span class="pre">modality_estimator._logsumexp</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">logliks</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_loglik</span><span class="p">(</span><span class="n">event</span><span class="p">)</span>
<span class="n">logsumexps</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_logsumexp</span><span class="p">(</span><span class="n">logliks</span><span class="p">)</span>

<span class="c"># Visualize the events</span>
<span class="n">plotter</span> <span class="o">=</span> <span class="n">modality_visualizer</span><span class="o">.</span><span class="n">event_estimation</span><span class="p">(</span><span class="n">event</span><span class="p">,</span> <span class="n">logliks</span><span class="p">,</span> <span class="n">logsumexps</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/modalities_18_0.png" src="../_images/modalities_18_0.png" />
<p>The left panel shows a violinplot of the event. The middle panel shows the log-likelihoods of different models at increasing parameterizations. The right panel shows the total summed <a class="reference external" href="http://en.wikipedia.org/wiki/Bayes_factor">Bayes factors</a> of how much more likely a model is over the uniform distribution, where the cutoff for a guessing that something is a uniform distribution is a $log_2left(text{Bayes factor}right) &gt; 3$.</p>
<p>Let&#8217;s add add random noise to the &#8220;perfect&#8221; included event, and see how that changes the modality estimation.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s">&#39;ticks&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">event_noisy</span> <span class="o">=</span> <span class="n">event</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">event_noisy</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>

<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">event_noisy</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">bw</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">modality_visualizer</span><span class="o">.</span><span class="n">modality_colors</span><span class="p">[</span><span class="s">&#39;included&#39;</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>

<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/modalities_20_0.png" src="../_images/modalities_20_0.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">logliks</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_loglik</span><span class="p">(</span><span class="n">event_noisy</span><span class="p">)</span>
<span class="n">logsumexps</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_logsumexp</span><span class="p">(</span><span class="n">logliks</span><span class="p">)</span>

<span class="c"># Visualize the events</span>
<span class="n">plotter</span> <span class="o">=</span> <span class="n">modality_visualizer</span><span class="o">.</span><span class="n">event_estimation</span><span class="p">(</span><span class="n">event_noisy</span><span class="p">,</span> <span class="n">logliks</span><span class="p">,</span> <span class="n">logsumexps</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/modalities_21_0.png" src="../_images/modalities_21_0.png" />
<p>The included modality becomes less likely, but it still beats out the uniform and bimodal distributions, which is what we want.</p>
<p>Let&#8217;s use this idea of noise to estimate modalities. We&#8217;ll use 100 cells total, and do 3 iterations of adding noise at 0%, 25%, and 50%. We will visualize each estimation using the same <tt class="docutils literal"><span class="pre">modality_visualizer</span></tt> for the parameterizations. This will output quite a few plots.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># We will be writing incrementally to this &quot;modality_guesses&quot; file, because if</span>
<span class="c"># this is done with many iterations and many noise levels, the file gets too</span>
<span class="c"># big and the IPython kernel crashes</span>
<span class="n">modality_guesses_csv</span> <span class="o">=</span> <span class="s">&#39;modality_guesses.csv&#39;</span>

<span class="c"># ALways start fresh with a new csv file</span>
<span class="k">try</span><span class="p">:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">modality_guesses_csv</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
    <span class="k">pass</span>

<span class="n">n_cells</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">n_cell</span> <span class="o">=</span> <span class="n">n_cells</span>
<span class="n">n_noisys</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">50</span><span class="p">)</span>

<span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="s">&#39;true_modality&#39;</span><span class="p">,</span> <span class="s">&#39;guessed_modality&#39;</span><span class="p">,</span> <span class="s">&#39;n_rogue&#39;</span><span class="p">,</span> <span class="s">&#39;iteration&#39;</span><span class="p">]</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">modality_guesses_csv</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="k">for</span> <span class="n">n_noisy</span> <span class="ow">in</span> <span class="n">n_noisys</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
        <span class="n">rows</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">rv</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">rvs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">event</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">rv</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">n_cells</span><span class="p">))</span>
            <span class="n">event</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s">&#39;True modality: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>

            <span class="n">event</span><span class="p">[:</span><span class="n">n_noisy</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n_noisy</span><span class="p">)</span>

            <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">modality_estimator</span><span class="o">.</span><span class="n">bimodal_model</span><span class="o">.</span><span class="n">prob_parameters</span><span class="p">)</span>

            <span class="n">logliks</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_loglik</span><span class="p">(</span><span class="n">event</span><span class="p">)</span>
            <span class="n">logsumexps</span> <span class="o">=</span> <span class="n">modality_estimator</span><span class="o">.</span><span class="n">_logsumexp</span><span class="p">(</span><span class="n">logliks</span><span class="p">)</span>

            <span class="c"># Visualize the events</span>
            <span class="n">plotter</span> <span class="o">=</span> <span class="n">modality_visualizer</span><span class="o">.</span><span class="n">event_estimation</span><span class="p">(</span><span class="n">event</span><span class="p">,</span> <span class="n">logliks</span><span class="p">,</span> <span class="n">logsumexps</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s">&#39;{}_nnoisy{}_iter{}_bimodal_exponential_other_uniform_prior.pdf&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">n_cell</span><span class="p">,</span> <span class="n">n_noisy</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
            <span class="n">plotter</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
            <span class="n">rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="s">&#39;bimodal_exponential_other_uniform&#39;</span><span class="p">,</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">logsumexps</span><span class="o">.</span><span class="n">argmax</span><span class="p">(),</span> <span class="n">n_noisy</span><span class="p">,</span> <span class="n">i</span><span class="p">])</span>

        <span class="n">modality_guesses</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">modality_guesses_csv</span><span class="p">,</span> <span class="s">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">modality_guesses</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
<pre class='nb-text-output'>
/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/matplotlib/pyplot.py:424: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  max_open_warning, RuntimeWarning)

</pre><img alt="../_images/modalities_23_1.png" src="../_images/modalities_23_1.png" />
<img alt="../_images/modalities_23_2.png" src="../_images/modalities_23_2.png" />
<img alt="../_images/modalities_23_3.png" src="../_images/modalities_23_3.png" />
<img alt="../_images/modalities_23_4.png" src="../_images/modalities_23_4.png" />
<img alt="../_images/modalities_23_5.png" src="../_images/modalities_23_5.png" />
<img alt="../_images/modalities_23_6.png" src="../_images/modalities_23_6.png" />
<img alt="../_images/modalities_23_7.png" src="../_images/modalities_23_7.png" />
<img alt="../_images/modalities_23_8.png" src="../_images/modalities_23_8.png" />
<img alt="../_images/modalities_23_9.png" src="../_images/modalities_23_9.png" />
<img alt="../_images/modalities_23_10.png" src="../_images/modalities_23_10.png" />
<img alt="../_images/modalities_23_11.png" src="../_images/modalities_23_11.png" />
<img alt="../_images/modalities_23_12.png" src="../_images/modalities_23_12.png" />
<img alt="../_images/modalities_23_13.png" src="../_images/modalities_23_13.png" />
<img alt="../_images/modalities_23_14.png" src="../_images/modalities_23_14.png" />
<img alt="../_images/modalities_23_15.png" src="../_images/modalities_23_15.png" />
<img alt="../_images/modalities_23_16.png" src="../_images/modalities_23_16.png" />
<img alt="../_images/modalities_23_17.png" src="../_images/modalities_23_17.png" />
<img alt="../_images/modalities_23_18.png" src="../_images/modalities_23_18.png" />
<img alt="../_images/modalities_23_19.png" src="../_images/modalities_23_19.png" />
<img alt="../_images/modalities_23_20.png" src="../_images/modalities_23_20.png" />
<img alt="../_images/modalities_23_21.png" src="../_images/modalities_23_21.png" />
<img alt="../_images/modalities_23_22.png" src="../_images/modalities_23_22.png" />
<img alt="../_images/modalities_23_23.png" src="../_images/modalities_23_23.png" />
<img alt="../_images/modalities_23_24.png" src="../_images/modalities_23_24.png" />
<img alt="../_images/modalities_23_25.png" src="../_images/modalities_23_25.png" />
<img alt="../_images/modalities_23_26.png" src="../_images/modalities_23_26.png" />
<img alt="../_images/modalities_23_27.png" src="../_images/modalities_23_27.png" />
<img alt="../_images/modalities_23_28.png" src="../_images/modalities_23_28.png" />
<img alt="../_images/modalities_23_29.png" src="../_images/modalities_23_29.png" />
<img alt="../_images/modalities_23_30.png" src="../_images/modalities_23_30.png" />
<img alt="../_images/modalities_23_31.png" src="../_images/modalities_23_31.png" />
<img alt="../_images/modalities_23_32.png" src="../_images/modalities_23_32.png" />
<img alt="../_images/modalities_23_33.png" src="../_images/modalities_23_33.png" />
<img alt="../_images/modalities_23_34.png" src="../_images/modalities_23_34.png" />
<img alt="../_images/modalities_23_35.png" src="../_images/modalities_23_35.png" />
<img alt="../_images/modalities_23_36.png" src="../_images/modalities_23_36.png" />
<img alt="../_images/modalities_23_37.png" src="../_images/modalities_23_37.png" />
<img alt="../_images/modalities_23_38.png" src="../_images/modalities_23_38.png" />
<img alt="../_images/modalities_23_39.png" src="../_images/modalities_23_39.png" />
<img alt="../_images/modalities_23_40.png" src="../_images/modalities_23_40.png" />
<img alt="../_images/modalities_23_41.png" src="../_images/modalities_23_41.png" />
<img alt="../_images/modalities_23_42.png" src="../_images/modalities_23_42.png" />
<img alt="../_images/modalities_23_43.png" src="../_images/modalities_23_43.png" />
<img alt="../_images/modalities_23_44.png" src="../_images/modalities_23_44.png" />
<img alt="../_images/modalities_23_45.png" src="../_images/modalities_23_45.png" />
<img alt="../_images/modalities_23_46.png" src="../_images/modalities_23_46.png" />
<img alt="../_images/modalities_23_47.png" src="../_images/modalities_23_47.png" />
<img alt="../_images/modalities_23_48.png" src="../_images/modalities_23_48.png" />
<img alt="../_images/modalities_23_49.png" src="../_images/modalities_23_49.png" />
<img alt="../_images/modalities_23_50.png" src="../_images/modalities_23_50.png" />
<img alt="../_images/modalities_23_51.png" src="../_images/modalities_23_51.png" />
<img alt="../_images/modalities_23_52.png" src="../_images/modalities_23_52.png" />
<img alt="../_images/modalities_23_53.png" src="../_images/modalities_23_53.png" />
<img alt="../_images/modalities_23_54.png" src="../_images/modalities_23_54.png" />
<img alt="../_images/modalities_23_55.png" src="../_images/modalities_23_55.png" />
<img alt="../_images/modalities_23_56.png" src="../_images/modalities_23_56.png" />
<img alt="../_images/modalities_23_57.png" src="../_images/modalities_23_57.png" />
<img alt="../_images/modalities_23_58.png" src="../_images/modalities_23_58.png" />
<img alt="../_images/modalities_23_59.png" src="../_images/modalities_23_59.png" />
<img alt="../_images/modalities_23_60.png" src="../_images/modalities_23_60.png" />
<p>Now let&#8217;s plot the accuracy of the model on the increasingly noisy splicing events and their modality estimations.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">modality_guesses</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">modality_guesses_csv</span><span class="p">)</span>

<span class="n">model_accuracy</span> <span class="o">=</span> <span class="n">modality_guesses</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="s">&#39;n_rogue&#39;</span><span class="p">,</span>
                                           <span class="s">&#39;true_modality&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">true_modality</span> <span class="o">==</span> <span class="n">x</span><span class="o">.</span><span class="n">guessed_modality</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span><span class="o">/</span><span class="n">n_noisys</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">*</span><span class="n">n_cells</span>
<span class="n">model_accuracy</span> <span class="o">=</span> <span class="n">model_accuracy</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">model_accuracy</span> <span class="o">=</span> <span class="n">model_accuracy</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s">&#39;percentage&#39;</span><span class="p">})</span>
<span class="n">fg</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">factorplot</span><span class="p">(</span><span class="s">&#39;n_rogue&#39;</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">model_accuracy</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s">&#39;percentage&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s">&#39;true_modality&#39;</span><span class="p">,</span>
               <span class="n">hue_order</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;excluded&#39;</span><span class="p">,</span> <span class="s">&#39;middle&#39;</span><span class="p">,</span> <span class="s">&#39;included&#39;</span><span class="p">,</span> <span class="s">&#39;bimodal&#39;</span><span class="p">),</span> <span class="n">palette</span><span class="o">=</span><span class="s">&#39;deep&#39;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;point&#39;</span><span class="p">)</span>
<span class="n">fg</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s">&#39;percent_noise_vs_true_positives.pdf&#39;</span><span class="p">)</span>


<span class="n">false_positives</span> <span class="o">=</span> <span class="n">modality_guesses</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="s">&#39;n_rogue&#39;</span><span class="p">,</span>
                                            <span class="s">&#39;guessed_modality&#39;</span><span class="p">,</span> <span class="s">&#39;true_modality&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">true_modality</span> <span class="o">!=</span> <span class="n">x</span><span class="o">.</span><span class="n">guessed_modality</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span><span class="o">/</span><span class="mi">50</span>
<span class="c"># false_positives = false_positives.unstack()</span>
<span class="n">false_positives</span> <span class="o">=</span> <span class="n">false_positives</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">false_positives</span> <span class="o">=</span> <span class="n">false_positives</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s">&#39;counts&#39;</span><span class="p">})</span>
<span class="n">false_positives</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>

<span class="n">fg</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">factorplot</span><span class="p">(</span><span class="s">&#39;n_rogue&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s">&#39;counts&#39;</span><span class="p">,</span> <span class="n">row</span><span class="o">=</span><span class="s">&#39;prior&#39;</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&#39;true_modality&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s">&#39;guessed_modality&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">false_positives</span><span class="p">,</span>
               <span class="n">hue_order</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;excluded&#39;</span><span class="p">,</span> <span class="s">&#39;middle&#39;</span><span class="p">,</span> <span class="s">&#39;included&#39;</span><span class="p">,</span> <span class="s">&#39;bimodal&#39;</span><span class="p">,</span> <span class="s">&#39;uniform&#39;</span><span class="p">),</span> <span class="n">palette</span><span class="o">=</span><span class="s">&#39;deep&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="n">fg</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s">&#39;percent_noise_vs_false_positives_factorplot.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>---------------------------------------------------------------------------

AttributeError                            Traceback (most recent call last)

&lt;ipython-input-11-47af08ce33bf&gt; in &lt;module&gt;()
      2
      3 model_accuracy = modality_guesses.groupby([&#39;prior&#39;, &#39;n_rogue&#39;,
----&gt; 4                                            &#39;true_modality&#39;]).apply(lambda x: (x.true_modality == x.guessed_modality).sum())/n_noisys.max()*n_cells
      5 model_accuracy = model_accuracy.reset_index()
      6 model_accuracy = model_accuracy.rename(columns={0: &#39;percentage&#39;})


AttributeError: &#39;tuple&#39; object has no attribute &#39;max&#39;
</pre></div>
</div>
<p>The &#8220;middle&#8221; modalities become less accurate at about 30% noise, but from looking at the events, I&#8217;m happy with these getting &#8220;mis-categorized&#8221; because they do truly look like &#8220;uniform&#8221; events.</p>
</div>
</div>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="../_sources/tutorial/modalities.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <p>
    Stuck? Send questions, comments, and other feedback to:
    <ul>
        <li>The <a href="https://gitter.im/YeoLab/flotilla">Gitter</a> chat
            room for immediate questions</li>
        <li>The <a href="https://github.com/yeolab/flotilla/issues">Github
            issues page</a> for feature requests</li>
    </ul>
    Or send an email to obotvinn@ucsd.edu if you're oldskool like that.
    </p>

    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

  </body>
</html>