<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Shalek and Satija, et al (2013) &mdash; flotilla 0.2.8dev documentation</title>
    
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootswatch-3.3.4/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.2.8dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.8dev documentation" href="../index.html" />
    <link rel="up" title="Flotilla tutorial" href="../tutorial.html" />
    <link rel="prev" title="Storing supplemental data on Study objects" href="supplementaldata.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.8dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-8">v0.2.8 (........)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-7-april-20th-2015">v0.2.7 (April 20th, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-6-april-10th-2015">v0.2.6 (April 10th, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-5-march-3rd-2015">v0.2.5 (March 3rd, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Flotilla tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#create-a-datapackage">Create a datapackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#computing-on-splicing-data">Computing on splicing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorial.html#day-to-day-helpful-tips">Day to day helpful tips</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Shalek and Satija, <em>et al</em> (2013)</a><ul>
<li><a class="reference internal" href="#expression-data">Expression data</a></li>
<li><a class="reference internal" href="#expression-feature-data">Expression feature data</a></li>
<li><a class="reference internal" href="#splicing-data">Splicing Data</a><ul>
<li><a class="reference internal" href="#remove-multi-index-columns">Remove Multi-index columns</a></li>
<li><a class="reference internal" href="#splicing-feature-data">Splicing Feature Data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#metadata">Metadata</a></li>
<li><a class="reference internal" href="#mapping-stats-data">Mapping stats data</a></li>
<li><a class="reference internal" href="#create-a-flotilla-study">Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!</a></li>
<li><a class="reference internal" href="#figure-1">Figure 1</a><ul>
<li><a class="reference internal" href="#figure-1a">Figure 1a</a><ul>
<li><a class="reference internal" href="#without-flotilla-you-would-do">Without flotilla, you would do</a></li>
</ul>
</li>
<li><a class="reference internal" href="#figure-1b">Figure 1b</a><ul>
<li><a class="reference internal" href="#without-flotilla">Without <tt class="docutils literal"><span class="pre">flotilla</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#figure-1c">Figure 1c</a></li>
</ul>
</li>
<li><a class="reference internal" href="#figure-2">Figure 2</a><ul>
<li><a class="reference internal" href="#figure-2a">Figure 2a</a></li>
<li><a class="reference internal" href="#figure-2b">Figure 2b</a></li>
</ul>
</li>
<li><a class="reference internal" href="#figure-3">Figure 3</a><ul>
<li><a class="reference internal" href="#figure-3b-top-panel">Figure 3b, top panel</a></li>
<li><a class="reference internal" href="#figure-3b-bottom-panel">Figure 3b, bottom panel</a></li>
</ul>
</li>
<li><a class="reference internal" href="#figure-4">Figure 4</a><ul>
<li><a class="reference internal" href="#figure-4a">Figure 4a</a></li>
<li><a class="reference internal" href="#mark-immature-cells-as-a-new-subset">Mark immature cells as a new subset</a></li>
<li><a class="reference internal" href="#figure-4b">Figure 4b</a><ul>
<li><a class="reference internal" href="#elbow-method-for-determining-number-of-clusters">&#8220;Elbow method&#8221; for determining number of clusters</a></li>
<li><a class="reference internal" href="#their-pc-scores-and-clusters-for-the-genes">Their PC scores and clusters for the genes</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#supplementary-figures">Supplementary figures</a><ul>
<li><a class="reference internal" href="#supplementary-figure-1">Supplementary Figure 1</a></li>
<li><a class="reference internal" href="#supplementary-figures-2-and-3">Supplementary Figures 2 and 3</a></li>
<li><a class="reference internal" href="#supplementary-figure-4">Supplementary Figure 4</a></li>
<li><a class="reference internal" href="#supplementary-figure-5">Supplementary Figure 5</a><ul>
<li><a class="reference internal" href="#supplementary-figure-5d">Supplementary Figure 5d</a></li>
</ul>
</li>
<li><a class="reference internal" href="#supplementary-figures-6-20">Supplementary Figures 6-20</a></li>
</ul>
</li>
<li><a class="reference internal" href="#conclusions">Conclusions</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="../_sources/tutorial/shalek2013.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <p>In the interest of reproducibility, and to showcase our new package
<tt class="docutils literal"><span class="pre">`flotilla</span></tt> &lt;<a class="reference external" href="http://github.com/yeolab/flotilla">http://github.com/yeolab/flotilla</a>&gt;`_, I&#8217;ve reproduced
many figures from the landmark single-cell paper, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Single-cell
transcriptomics reveals bimodality in expression and splicing in immune
cells</a> by Shalek and
Satija, <em>et al</em>. <em>Nature</em> (2013).</p>
<p>Before we begin, let&#8217;s import everything we need.</p>
<div class="code python highlight-python"><div class="highlight"><pre># Turn on inline plots with IPython
%matplotlib inline

# Import the flotilla package for biological data analysis
import flotilla

# Import &quot;numerical python&quot; library for number crunching
import numpy as np

# Import &quot;panel data analysis&quot; library for tabular data
import pandas as pd

# Import statistical data visualization package
# Note: As of November 6th, 2014, you will need the &quot;master&quot; version of
# seaborn on github (v0.5.dev), installed via
# &quot;pip install git+ssh://git@github.com/mwaskom/seaborn.git
import seaborn as sns
</pre></div>
</div>
<div class="section" id="shalek-and-satija-et-al-2013">
<h1>Shalek and Satija, <em>et al</em> (2013)<a class="headerlink" href="#shalek-and-satija-et-al-2013" title="Permalink to this headline">¶</a></h1>
<p>In the 2013 paper, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/23685454">Single-cell transcriptomics reveals bimodality in
expression and splicing in immune
cells</a> (Shalek and
Satija, <em>et al</em>. <em>Nature</em> (2013)), Regev and colleagues performed
single-cell sequencing 18 bone marrow-derived dendritic cells (BMDCs),
in addition to 3 pooled samples.</p>
<div class="section" id="expression-data">
<h2>Expression data<a class="headerlink" href="#expression-data" title="Permalink to this headline">¶</a></h2>
<p>First, we will read in the expression data. These data were obtained
using,</p>
<div class="code python highlight-python"><div class="highlight"><pre>%%bash
wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>--2015-05-30 02:54:41--  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41265/suppl/GSE41265_allGenesTPM.txt.gz
           =&gt; `GSE41265_allGenesTPM.txt.gz&#39;
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 2607:f220:41e:250::10, 130.14.250.13
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|2607:f220:41e:250::10|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /geo/series/GSE41nnn/GSE41265/suppl ... done.
==&gt; SIZE GSE41265_allGenesTPM.txt.gz ... 1099290
==&gt; EPSV ... done.    ==&gt; RETR GSE41265_allGenesTPM.txt.gz ... done.
Length: 1099290 (1.0M) (unauthoritative)

     0K .......... .......... .......... .......... ..........  4% 1.14M 1s
    50K .......... .......... .......... .......... ..........  9% 5.77M 0s
   100K .......... .......... .......... .......... .......... 13% 6.20M 0s
   150K .......... .......... .......... .......... .......... 18% 57.9M 0s
   200K .......... .......... .......... .......... .......... 23% 6.31M 0s
   250K .......... .......... .......... .......... .......... 27% 6.57M 0s
   300K .......... .......... .......... .......... .......... 32% 80.6M 0s
   350K .......... .......... .......... .......... .......... 37% 64.2M 0s
   400K .......... .......... .......... .......... .......... 41% 6.79M 0s
   450K .......... .......... .......... .......... .......... 46% 84.6M 0s
   500K .......... .......... .......... .......... .......... 51% 89.8M 0s
   550K .......... .......... .......... .......... .......... 55% 7.91M 0s
   600K .......... .......... .......... .......... .......... 60% 55.7M 0s
   650K .......... .......... .......... .......... .......... 65% 68.4M 0s
   700K .......... .......... .......... .......... .......... 69% 80.7M 0s
   750K .......... .......... .......... .......... .......... 74%  160M 0s
   800K .......... .......... .......... .......... .......... 79% 8.46M 0s
   850K .......... .......... .......... .......... .......... 83% 44.6M 0s
   900K .......... .......... .......... .......... .......... 88% 84.5M 0s
   950K .......... .......... .......... .......... .......... 93% 78.9M 0s
  1000K .......... .......... .......... .......... .......... 97%  181M 0s
  1050K .......... .......... ...                             100%  198M=0.1s

2015-05-30 02:54:41 (10.3 MB/s) - `GSE41265_allGenesTPM.txt.gz&#39; saved [1099290]
</pre></div>
</div>
<p>We will also compare to the supplementary table 2 data, obtained using</p>
<div class="code python highlight-python"><div class="highlight"><pre>%%bash
wget http://www.nature.com/nature/journal/v498/n7453/extref/nature12172-s1.zip
unzip nature12172-s1.zip
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Archive:  nature12172-s1.zip
   creating: nature12172-s1/
  inflating: nature12172-s1/Supplementary_Table1.xls
  inflating: nature12172-s1/Supplementary_Table2.xlsx
  inflating: nature12172-s1/Supplementary_Table3.xls
  inflating: nature12172-s1/Supplementary_Table4.xls
  inflating: nature12172-s1/Supplementary_Table5.xls
  inflating: nature12172-s1/Supplementary_Table6.xls
  inflating: nature12172-s1/Supplementary_Table7.xlsx
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>--2015-05-30 02:54:41--  http://www.nature.com/nature/journal/v498/n7453/extref/nature12172-s1.zip
Resolving www.nature.com (www.nature.com)... 69.31.29.37, 69.31.29.198
Connecting to www.nature.com (www.nature.com)|69.31.29.37|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4634226 (4.4M) [application/zip]
Saving to: `nature12172-s1.zip&#39;

     0K .......... .......... .......... .......... ..........  1% 3.59M 1s
    50K .......... .......... .......... .......... ..........  2% 6.90M 1s
   100K .......... .......... .......... .......... ..........  3% 42.9M 1s
   150K .......... .......... .......... .......... ..........  4% 7.41M 1s
   200K .......... .......... .......... .......... ..........  5% 7.40M 1s
   250K .......... .......... .......... .......... ..........  6% 61.9M 1s
   300K .......... .......... .......... .......... ..........  7% 8.07M 0s
   350K .......... .......... .......... .......... ..........  8% 47.5M 0s
   400K .......... .......... .......... .......... ..........  9% 96.6M 0s
   450K .......... .......... .......... .......... .......... 11% 9.37M 0s
   500K .......... .......... .......... .......... .......... 12% 43.0M 0s
   550K .......... .......... .......... .......... .......... 13% 95.4M 0s
   600K .......... .......... .......... .......... .......... 14% 77.4M 0s
   650K .......... .......... .......... .......... .......... 15%  119M 0s
   700K .......... .......... .......... .......... .......... 16% 9.26M 0s
   750K .......... .......... .......... .......... .......... 17% 86.0M 0s
   800K .......... .......... .......... .......... .......... 18%  111M 0s
   850K .......... .......... .......... .......... .......... 19% 85.6M 0s
   900K .......... .......... .......... .......... .......... 20% 93.5M 0s
   950K .......... .......... .......... .......... .......... 22% 10.0M 0s
  1000K .......... .......... .......... .......... .......... 23% 84.3M 0s
  1050K .......... .......... .......... .......... .......... 24%  119M 0s
  1100K .......... .......... .......... .......... .......... 25%  114M 0s
  1150K .......... .......... .......... .......... .......... 26% 18.7M 0s
  1200K .......... .......... .......... .......... .......... 27% 96.9M 0s
  1250K .......... .......... .......... .......... .......... 28%  123M 0s
  1300K .......... .......... .......... .......... .......... 29%  103M 0s
  1350K .......... .......... .......... .......... .......... 30% 25.1M 0s
  1400K .......... .......... .......... .......... .......... 32% 43.3M 0s
  1450K .......... .......... .......... .......... .......... 33%  122M 0s
  1500K .......... .......... .......... .......... .......... 34%  130M 0s
  1550K .......... .......... .......... .......... .......... 35% 6.03M 0s
  1600K .......... .......... .......... .......... .......... 36%  134M 0s
  1650K .......... .......... .......... .......... .......... 37% 91.6M 0s
  1700K .......... .......... .......... .......... .......... 38% 72.3M 0s
  1750K .......... .......... .......... .......... .......... 39% 87.8M 0s
  1800K .......... .......... .......... .......... .......... 40%  121M 0s
  1850K .......... .......... .......... .......... .......... 41%  124M 0s
  1900K .......... .......... .......... .......... .......... 43% 89.1M 0s
  1950K .......... .......... .......... .......... .......... 44% 95.6M 0s
  2000K .......... .......... .......... .......... .......... 45% 15.4M 0s
  2050K .......... .......... .......... .......... .......... 46% 52.1M 0s
  2100K .......... .......... .......... .......... .......... 47%  108M 0s
  2150K .......... .......... .......... .......... .......... 48%  144M 0s
  2200K .......... .......... .......... .......... .......... 49% 95.6M 0s
  2250K .......... .......... .......... .......... .......... 50%  128M 0s
  2300K .......... .......... .......... .......... .......... 51%  122M 0s
  2350K .......... .......... .......... .......... .......... 53%  128M 0s
  2400K .......... .......... .......... .......... .......... 54%  129M 0s
  2450K .......... .......... .......... .......... .......... 55% 92.6M 0s
  2500K .......... .......... .......... .......... .......... 56% 98.1M 0s
  2550K .......... .......... .......... .......... .......... 57% 85.7M 0s
  2600K .......... .......... .......... .......... .......... 58% 21.9M 0s
  2650K .......... .......... .......... .......... .......... 59% 36.3M 0s
  2700K .......... .......... .......... .......... .......... 60% 90.4M 0s
  2750K .......... .......... .......... .......... .......... 61%  117M 0s
  2800K .......... .......... .......... .......... .......... 62%  125M 0s
  2850K .......... .......... .......... .......... .......... 64%  118M 0s
  2900K .......... .......... .......... .......... .......... 65% 92.5M 0s
  2950K .......... .......... .......... .......... .......... 66%  134M 0s
  3000K .......... .......... .......... .......... .......... 67%  190M 0s
  3050K .......... .......... .......... .......... .......... 68%  142M 0s
  3100K .......... .......... .......... .......... .......... 69%  133M 0s
  3150K .......... .......... .......... .......... .......... 70%  132M 0s
  3200K .......... .......... .......... .......... .......... 71%  177M 0s
  3250K .......... .......... .......... .......... .......... 72% 3.54M 0s
  3300K .......... .......... .......... .......... .......... 74%  109M 0s
  3350K .......... .......... .......... .......... .......... 75%  129M 0s
  3400K .......... .......... .......... .......... .......... 76% 86.4M 0s
  3450K .......... .......... .......... .......... .......... 77% 66.8M 0s
  3500K .......... .......... .......... .......... .......... 78%  121M 0s
  3550K .......... .......... .......... .......... .......... 79%  120M 0s
  3600K .......... .......... .......... .......... .......... 80% 95.4M 0s
  3650K .......... .......... .......... .......... .......... 81%  201M 0s
  3700K .......... .......... .......... .......... .......... 82%  122M 0s
  3750K .......... .......... .......... .......... .......... 83% 98.6M 0s
  3800K .......... .......... .......... .......... .......... 85% 81.6M 0s
  3850K .......... .......... .......... .......... .......... 86%  110M 0s
  3900K .......... .......... .......... .......... .......... 87% 10.2M 0s
  3950K .......... .......... .......... .......... .......... 88% 89.4M 0s
  4000K .......... .......... .......... .......... .......... 89% 84.7M 0s
  4050K .......... .......... .......... .......... .......... 90% 6.92M 0s
  4100K .......... .......... .......... .......... .......... 91% 78.2M 0s
  4150K .......... .......... .......... .......... .......... 92% 91.7M 0s
  4200K .......... .......... .......... .......... .......... 93% 91.3M 0s
  4250K .......... .......... .......... .......... .......... 95%  111M 0s
  4300K .......... .......... .......... .......... .......... 96% 93.4M 0s
  4350K .......... .......... .......... .......... .......... 97%  116M 0s
  4400K .......... .......... .......... .......... .......... 98% 78.8M 0s
  4450K .......... .......... .......... .......... .......... 99% 95.2M 0s
  4500K .......... .......... .....                           100%  269M=0.1s

2015-05-30 02:54:42 (32.0 MB/s) - `nature12172-s1.zip&#39; saved [4634226/4634226]
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s">&quot;GSE41265_allGenesTPM.txt.gz&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s">&quot;gzip&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
      <th>S5</th>
      <th>S6</th>
      <th>S7</th>
      <th>S8</th>
      <th>S9</th>
      <th>S10</th>
      <th>...</th>
      <th>S12</th>
      <th>S13</th>
      <th>S14</th>
      <th>S15</th>
      <th>S16</th>
      <th>S17</th>
      <th>S18</th>
      <th>P1</th>
      <th>P2</th>
      <th>P3</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>XKR4</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.019906</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>AB338584</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>B3GAT2</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.023441</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.029378</td>
      <td>0.000000</td>
      <td>0.055452</td>
      <td>0.000000</td>
      <td>0.029448</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.031654</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>42.150208</td>
      <td>0.680327</td>
      <td>0.022996</td>
      <td>0.110236</td>
    </tr>
    <tr>
      <th>NPL</th>
      <td>72.008590</td>
      <td>0.000000</td>
      <td>128.062012</td>
      <td>0.095082</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>112.310234</td>
      <td>104.329122</td>
      <td>0.119230</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.116802</td>
      <td>0.104200</td>
      <td>0.106188</td>
      <td>0.229197</td>
      <td>0.110582</td>
      <td>0.000000</td>
      <td>7.109356</td>
      <td>6.727028</td>
      <td>14.525447</td>
    </tr>
    <tr>
      <th>T2</th>
      <td>0.109249</td>
      <td>0.172009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.182703</td>
      <td>0.076012</td>
      <td>0.078698</td>
      <td>0.000000</td>
      <td>0.093698</td>
      <td>0.076583</td>
      <td>...</td>
      <td>0.693459</td>
      <td>0.010137</td>
      <td>0.081936</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.086879</td>
      <td>0.068174</td>
      <td>0.062063</td>
      <td>0.000000</td>
      <td>0.050605</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div><p>These data are in the &#8220;transcripts per million,&#8221; aka TPM unit. See
<a class="reference external" href="http://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/">this</a>
blog post if that sounds weird to you.</p>
<p>These data are formatted with samples on the columns, and genes on the
rows. But we want the opposite, with samples on the rows and genes on
the columns. This follows
<tt class="docutils literal"><span class="pre">`scikit-learn</span></tt> &lt;<a class="reference external" href="http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset">http://scikit-learn.org/stable/tutorial/basic/tutorial.html#loading-an-example-dataset</a>&gt;`_&#8217;s
standard of data matrices with size (<tt class="docutils literal"><span class="pre">n_samples</span></tt>, <tt class="docutils literal"><span class="pre">n_features</span></tt>) as
each gene is a feature. So we will simply transpose this.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">T</span>
<span class="n">expression</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>GENE</th>
      <th>XKR4</th>
      <th>AB338584</th>
      <th>B3GAT2</th>
      <th>NPL</th>
      <th>T2</th>
      <th>T</th>
      <th>PDE10A</th>
      <th>1700010I14RIK</th>
      <th>6530411M01RIK</th>
      <th>PABPC6</th>
      <th>...</th>
      <th>AK085062</th>
      <th>DHX9</th>
      <th>RNASET2B</th>
      <th>FGFR1OP</th>
      <th>CCR6</th>
      <th>BRP44L</th>
      <th>AK014435</th>
      <th>AK015714</th>
      <th>SFT2D1</th>
      <th>PRR18</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>72.008590</td>
      <td>0.109249</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.774638</td>
      <td>23.520936</td>
      <td>0.000000</td>
      <td>0</td>
      <td>460.316773</td>
      <td>0</td>
      <td>0.000000</td>
      <td>39.442566</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.172009</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.367391</td>
      <td>1.887873</td>
      <td>0.000000</td>
      <td>0</td>
      <td>823.890290</td>
      <td>0</td>
      <td>0.000000</td>
      <td>4.967412</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>0</td>
      <td>0</td>
      <td>0.023441</td>
      <td>128.062012</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.249858</td>
      <td>0.313510</td>
      <td>0.166772</td>
      <td>0</td>
      <td>1002.354241</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.095082</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.354157</td>
      <td>0.000000</td>
      <td>0.887003</td>
      <td>0</td>
      <td>1230.766795</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.131215</td>
      <td>0</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.182703</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0.039263</td>
      <td>0.000000</td>
      <td>131.077131</td>
      <td>0</td>
      <td>1614.749122</td>
      <td>0</td>
      <td>0.242179</td>
      <td>95.485743</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27723 columns</p>
</div><p>The authors filtered the expression data based on having at least 3
single cells express genes with at TPM (transcripts per million, ) &gt; 1.
We can express this in using the
<tt class="docutils literal"><span class="pre">`pandas</span></tt> &lt;<a class="reference external" href="http://pandas.pydata.org">http://pandas.pydata.org</a>&gt;`_ DataFrames easily.</p>
<p>First, from reading the paper and looking at the data, I know there are
18 single cells, and there are 18 samples that start with the letter
&#8220;S.&#8221; So I will extract the single samples from the <tt class="docutils literal"><span class="pre">index</span></tt> (row names)
using a <tt class="docutils literal"><span class="pre">lambda</span></tt>, a tiny function which in this case, tells me whether
or not that sample id begins with the letter &#8220;S&#8221;.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">singles_ids</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">expression</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;S&#39;</span><span class="p">))]</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">singles_ids</span><span class="p">))</span>
<span class="n">singles</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">]</span>

<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">singles</span><span class="p">[</span><span class="n">singles</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">]</span>
<span class="n">expression_filtered</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">expression_filtered</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">expression_filtered</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">&#39;number of single cells:&#39;</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">6312</span><span class="p">)</span>
</pre></div>
</div>
<p>Hmm, that&#8217;s strange. The paper states that they had 6313 genes after
filtering, but I get 6312. Even using &#8220;<tt class="docutils literal"><span class="pre">singles</span> <span class="pre">&gt;=</span> <span class="pre">1</span></tt>&#8221; doesn&#8217;t help.</p>
<p>(I also tried this with the expression table provided in the
supplementary data as &#8220;<tt class="docutils literal"><span class="pre">SupplementaryTable2.xlsx</span></tt>,&#8221; and got the same
results.)</p>
<p>Now that we&#8217;ve taken care of importing and filtering the expression
data, let&#8217;s do the feature data of the expression data.</p>
</div>
<div class="section" id="expression-feature-data">
<h2>Expression feature data<a class="headerlink" href="#expression-feature-data" title="Permalink to this headline">¶</a></h2>
<p>This is similar to the <tt class="docutils literal"><span class="pre">fData</span></tt> from <tt class="docutils literal"><span class="pre">BioconductoR</span></tt>, where there&#8217;s
some additional data on your features that you want to look at. They
uploaded information about the features in their OTHER expression
matrix, uploaded as a supplementary file, <tt class="docutils literal"><span class="pre">Supplementary_Table2.xlsx</span></tt>.</p>
<p>Notice that this is a <tt class="docutils literal"><span class="pre">csv</span></tt> and not an <tt class="docutils literal"><span class="pre">xlsx</span></tt>. This is because Excel
mangled the gene IDS that started with <tt class="docutils literal"><span class="pre">201*</span></tt> and assumed they were
dates :(</p>
<p>The workaround I did was to add another column to the sheet with the
formula <tt class="docutils literal"><span class="pre">=&quot;'&quot;</span> <span class="pre">&amp;</span> <span class="pre">A1</span></tt>, press <tt class="docutils literal"><span class="pre">Command</span></tt>-<tt class="docutils literal"><span class="pre">Shift</span></tt>-<tt class="docutils literal"><span class="pre">End</span></tt> to select the
end of the rows, and then do <tt class="docutils literal"><span class="pre">Ctrl</span></tt>-<tt class="docutils literal"><span class="pre">D</span></tt> to &#8220;fill down&#8221; to the bottom
(thanks to
<a class="reference external" href="http://superuser.com/questions/298276/excel-keyboard-shortcut-to-copy-fill-down-for-all-cells-with-non-blank-adjacent">this</a>
stackoverflow post for teaching me how to Excel). Then, I saved the file
as a <tt class="docutils literal"><span class="pre">csv</span></tt> for maximum portability and compatibility.</p>
<p>So sorry, this requires some non-programming editing! But I&#8217;ve posted
the csv to our <a class="reference external" href="https://github.com/YeoLab/shalek2013">github repo</a>
with all the data, and we&#8217;ll access it from there.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">&#39;https://raw.githubusercontent.com/YeoLab/shalek2013/master/Supplementary_Table2.csv&#39;</span><span class="p">,</span>
                            <span class="c"># Need to specify the index column as both the first and the last columns,</span>
                            <span class="c"># Because the last column is the &quot;Gene Category&quot;</span>
                            <span class="n">index_col</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">parse_dates</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">infer_datetime_format</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c"># This was also in features x samples format, so we need to transpose</span>
<span class="n">expression2</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">T</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>'GENE</th>
      <th>'0610007L01RIK</th>
      <th>'0610007P14RIK</th>
      <th>'0610007P22RIK</th>
      <th>'0610008F07RIK</th>
      <th>'0610009B22RIK</th>
      <th>'0610009D07RIK</th>
      <th>'0610009O20RIK</th>
      <th>'0610010B08RIK</th>
      <th>'0610010F05RIK</th>
      <th>'0610010K06RIK</th>
      <th>...</th>
      <th>'ZWILCH</th>
      <th>'ZWINT</th>
      <th>'ZXDA</th>
      <th>'ZXDB</th>
      <th>'ZXDC</th>
      <th>'ZYG11A</th>
      <th>'ZYG11B</th>
      <th>'ZYX</th>
      <th>'ZZEF1</th>
      <th>'ZZZ3</th>
    </tr>
    <tr>
      <th>Gene Category</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>...</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>27.181570</td>
      <td>0.166794</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>178.852732</td>
      <td>0</td>
      <td>0.962417</td>
      <td>0.000000</td>
      <td>143.359550</td>
      <td>...</td>
      <td>0.000000</td>
      <td>302.361227</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.027717</td>
      <td>297.918756</td>
      <td>37.685501</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>37.682691</td>
      <td>0.263962</td>
      <td>0</td>
      <td>0</td>
      <td>0.207921</td>
      <td>0.141099</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.255617</td>
      <td>...</td>
      <td>0.000000</td>
      <td>96.033724</td>
      <td>0.020459</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.042430</td>
      <td>0.242888</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>0.056916</td>
      <td>78.622459</td>
      <td>0</td>
      <td>0</td>
      <td>0.145680</td>
      <td>0.396363</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.024692</td>
      <td>72.775846</td>
      <td>...</td>
      <td>0.000000</td>
      <td>427.915555</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.040407</td>
      <td>6.753530</td>
      <td>0.132011</td>
      <td>0.017615</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>55.649250</td>
      <td>0.228866</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>88.798158</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>93.825442</td>
      <td>...</td>
      <td>0.000000</td>
      <td>9.788557</td>
      <td>0.017787</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.013452</td>
      <td>0.274689</td>
      <td>9.724890</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>0.000000</td>
      <td>0.093117</td>
      <td>0</td>
      <td>0</td>
      <td>131.326008</td>
      <td>155.936361</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.031029</td>
      <td>...</td>
      <td>0.204522</td>
      <td>26.575760</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1.101589</td>
      <td>59.256094</td>
      <td>44.430726</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27723 columns</p>
</div><p>Now we need to strip the single-quote I added to all the gene names:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">new_index</span><span class="p">,</span> <span class="n">indexer</span> <span class="o">=</span> <span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&quot;&#39;&quot;</span><span class="p">),</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">new_index</span>
<span class="n">expression2</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>'GENE</th>
      <th>0610007L01RIK</th>
      <th>0610007P14RIK</th>
      <th>0610007P22RIK</th>
      <th>0610008F07RIK</th>
      <th>0610009B22RIK</th>
      <th>0610009D07RIK</th>
      <th>0610009O20RIK</th>
      <th>0610010B08RIK</th>
      <th>0610010F05RIK</th>
      <th>0610010K06RIK</th>
      <th>...</th>
      <th>ZWILCH</th>
      <th>ZWINT</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZYG11A</th>
      <th>ZYG11B</th>
      <th>ZYX</th>
      <th>ZZEF1</th>
      <th>ZZZ3</th>
    </tr>
    <tr>
      <th>Gene Category</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>...</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
      <th>NaN</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td>27.181570</td>
      <td>0.166794</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>178.852732</td>
      <td>0</td>
      <td>0.962417</td>
      <td>0.000000</td>
      <td>143.359550</td>
      <td>...</td>
      <td>0.000000</td>
      <td>302.361227</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.027717</td>
      <td>297.918756</td>
      <td>37.685501</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>37.682691</td>
      <td>0.263962</td>
      <td>0</td>
      <td>0</td>
      <td>0.207921</td>
      <td>0.141099</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.255617</td>
      <td>...</td>
      <td>0.000000</td>
      <td>96.033724</td>
      <td>0.020459</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.042430</td>
      <td>0.242888</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>0.056916</td>
      <td>78.622459</td>
      <td>0</td>
      <td>0</td>
      <td>0.145680</td>
      <td>0.396363</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.024692</td>
      <td>72.775846</td>
      <td>...</td>
      <td>0.000000</td>
      <td>427.915555</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.040407</td>
      <td>6.753530</td>
      <td>0.132011</td>
      <td>0.017615</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>55.649250</td>
      <td>0.228866</td>
      <td>0</td>
      <td>0</td>
      <td>0.000000</td>
      <td>88.798158</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>93.825442</td>
      <td>...</td>
      <td>0.000000</td>
      <td>9.788557</td>
      <td>0.017787</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0.013452</td>
      <td>0.274689</td>
      <td>9.724890</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>0.000000</td>
      <td>0.093117</td>
      <td>0</td>
      <td>0</td>
      <td>131.326008</td>
      <td>155.936361</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.031029</td>
      <td>...</td>
      <td>0.204522</td>
      <td>26.575760</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>1.101589</td>
      <td>59.256094</td>
      <td>44.430726</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27723 columns</p>
</div><p>We want to create a <tt class="docutils literal"><span class="pre">pandas.DataFrame</span></tt> from the &#8220;Gene Category&#8221; row
for our <tt class="docutils literal"><span class="pre">expression_feature_data</span></tt>, which we will do via:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_ids</span><span class="p">,</span> <span class="n">gene_category</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">expression2</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">gene_categories</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">gene_category</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s">&#39;gene_category&#39;</span><span class="p">)</span>
<span class="n">gene_categories</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>0610007L01RIK             NaN
0610007P14RIK             NaN
0610007P22RIK             NaN
0610008F07RIK             NaN
0610009B22RIK             NaN
0610009D07RIK             NaN
0610009O20RIK             NaN
0610010B08RIK             NaN
0610010F05RIK             NaN
0610010K06RIK             NaN
0610010K14RIK             NaN
0610010O12RIK             NaN
0610011F06RIK             NaN
0610011L14RIK             NaN
0610012G03RIK             NaN
0610012H03RIK             NaN
0610030E20RIK             NaN
0610031J06RIK             NaN
0610037L13RIK             NaN
0610037P05RIK             NaN
0610038B21RIK             NaN
0610039K10RIK             NaN
0610040B10RIK             NaN
0610040J01RIK             NaN
0910001L09RIK             NaN
100043387                 NaN
1100001G20RIK             NaN
1110001A16RIK             NaN
1110001J03RIK             NaN
1110002B05RIK             NaN
                     ...
ZSCAN20                   NaN
ZSCAN21                   NaN
ZSCAN22                   NaN
ZSCAN29                   NaN
ZSCAN30                   NaN
ZSCAN4B                   NaN
ZSCAN4C                   NaN
ZSCAN4D                   NaN
ZSCAN4E                   NaN
ZSCAN4F                   NaN
ZSCAN5B                   NaN
ZSWIM1                    NaN
ZSWIM2                    NaN
ZSWIM3                    NaN
ZSWIM4                    NaN
ZSWIM5                    NaN
ZSWIM6                    NaN
ZSWIM7                    NaN
ZUFSP            LPS Response
ZW10                      NaN
ZWILCH                    NaN
ZWINT                     NaN
ZXDA                      NaN
ZXDB                      NaN
ZXDC                      NaN
ZYG11A                    NaN
ZYG11B                    NaN
ZYX                       NaN
ZZEF1                     NaN
ZZZ3                      NaN
Name: gene_category, dtype: object
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression_feature_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_categories</span><span class="p">)</span>
<span class="n">expression_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0610007L01RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>NaN</td>
    </tr>
    <tr>
      <th>0610009B22RIK</th>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
<div class="section" id="splicing-data">
<h2>Splicing Data<a class="headerlink" href="#splicing-data" title="Permalink to this headline">¶</a></h2>
<p>We obtain the splicing data from this study from the supplementary
information, specifically the <tt class="docutils literal"><span class="pre">Supplementary_Table4.xls</span></tt></p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table4.xls&#39;</span><span class="p">,</span> <span class="s">&#39;splicingTable.txt&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)

&lt;ipython-input-11-6956dd3a6ad6&gt; in &lt;module&gt;()
----&gt; 1 splicing = pd.read_excel(&#39;nature12172-s1/Supplementary_Table4.xls&#39;, &#39;splicingTable.txt&#39;, index_col=(0,1))
      2 splicing.head()


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/excel.pyc in read_excel(io, sheetname, **kwds)
    149     engine = kwds.pop(&#39;engine&#39;, None)
    150
--&gt; 151     return ExcelFile(io, engine=engine).parse(sheetname=sheetname, **kwds)
    152
    153


/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/pandas/io/excel.pyc in __init__(self, io, **kwds)
    167     def __init__(self, io, **kwds):
    168
--&gt; 169         import xlrd  # throw an ImportError if we need to
    170
    171         ver = tuple(map(int, xlrd.__VERSION__.split(&quot;.&quot;)[:2]))


ImportError: No module named xlrd
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">T</span>
<span class="n">splicing</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>Event name</th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
    <tr>
      <th>gene</th>
      <th>Os9</th>
      <th>Vta1</th>
      <th>Bclaf1</th>
      <th>Bclaf1</th>
      <th>P4ha1</th>
      <th>Bsg</th>
      <th>Ptbp1</th>
      <th>Igf1</th>
      <th>Elk3</th>
      <th>Nbr1</th>
      <th>...</th>
      <th>Afg3l1</th>
      <th>Tep1</th>
      <th>Fgd2</th>
      <th>Ttc17</th>
      <th>Tmed1</th>
      <th>Tmed1</th>
      <th>Sbno2</th>
      <th>Synj2</th>
      <th>Tbc1d13</th>
      <th>Usp47</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> 0.84</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.42</td>
      <td>  NaN</td>
      <td> 0.57</td>
      <td> 0.31</td>
      <td> 0.93</td>
      <td> 0.57</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td> 0.98</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.55</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.20</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>  NaN</td>
      <td> 0.84</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.94</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.73</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S6</th>
      <td> 0.01</td>
      <td> 0.91</td>
      <td> 0.14</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.61</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S7</th>
      <td>  NaN</td>
      <td> 0.87</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.62</td>
      <td>  NaN</td>
      <td> 0.85</td>
      <td> 0.73</td>
      <td> 0.55</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S8</th>
      <td>  NaN</td>
      <td> 0.86</td>
      <td> 0.02</td>
      <td> 0.98</td>
      <td> 0.03</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.89</td>
      <td> 0.82</td>
      <td> 0.83</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S9</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.97</td>
      <td>  NaN</td>
      <td> 0.97</td>
      <td>  NaN</td>
      <td> 0.90</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S10</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.06</td>
      <td> 0.98</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S11</th>
      <td> 0.03</td>
      <td> 0.93</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.97</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S13</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S14</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.88</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S15</th>
      <td> 0.02</td>
      <td> 0.96</td>
      <td> 0.01</td>
      <td> 0.06</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.44</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td> 0.91</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S16</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td> 0.27</td>
      <td> 0.99</td>
      <td> 0.99</td>
      <td> 0.98</td>
      <td> 0.98</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>S17</th>
      <td> 0.01</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.96</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.99</td>
      <td> 0.98</td>
      <td> 0.67</td>
      <td> 0.07</td>
    </tr>
    <tr>
      <th>S18</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>10,000 cell Rep1 (P1)</th>
      <td> 0.27</td>
      <td> 0.83</td>
      <td> 0.40</td>
      <td> 0.62</td>
      <td> 0.43</td>
      <td> 0.78</td>
      <td>  NaN</td>
      <td> 0.60</td>
      <td> 0.76</td>
      <td> 0.52</td>
      <td>...</td>
      <td> 0.92</td>
      <td>  NaN</td>
      <td> 0.81</td>
      <td> 0.77</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.84</td>
      <td> 0.50</td>
      <td> 0.56</td>
      <td>  NaN</td>
    </tr>
    <tr>
      <th>10,000 cell Rep2 (P2)</th>
      <td> 0.37</td>
      <td> 0.85</td>
      <td> 0.49</td>
      <td> 0.63</td>
      <td> 0.36</td>
      <td> 0.72</td>
      <td> 0.47</td>
      <td> 0.60</td>
      <td> 0.73</td>
      <td> 0.68</td>
      <td>...</td>
      <td> 0.67</td>
      <td> 0.15</td>
      <td> 0.52</td>
      <td> 0.67</td>
      <td> 0.63</td>
      <td> 0.73</td>
      <td> 0.82</td>
      <td> 0.90</td>
      <td> 0.71</td>
      <td> 0.55</td>
    </tr>
    <tr>
      <th>10,000 cell Rep3 (P3)</th>
      <td> 0.31</td>
      <td> 0.64</td>
      <td> 0.59</td>
      <td> 0.70</td>
      <td> 0.52</td>
      <td> 0.79</td>
      <td>  NaN</td>
      <td> 0.65</td>
      <td> 0.42</td>
      <td> 0.64</td>
      <td>...</td>
      <td> 0.58</td>
      <td> 0.79</td>
      <td> 0.74</td>
      <td> 0.85</td>
      <td> 0.73</td>
      <td> 0.39</td>
      <td> 0.56</td>
      <td>  NaN</td>
      <td> 0.64</td>
      <td>  NaN</td>
    </tr>
  </tbody>
</table>
<p>20 rows × 352 columns</p>
</div><p>The three pooled samples aren&#8217;t named consistently with the expression
data, so we have to fix that.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">&#39;P&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">)]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="n">Index</span><span class="p">([</span><span class="s">u&#39;10,000 cell Rep1 (P1)&#39;</span><span class="p">,</span> <span class="s">u&#39;10,000 cell Rep2 (P2)&#39;</span><span class="p">,</span> <span class="s">u&#39;10,000 cell Rep3 (P3)&#39;</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s">&#39;object&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Since the pooled sample IDs are inconsistent with the <tt class="docutils literal"><span class="pre">expression</span></tt>
data, we have to change them. We can get the &#8220;P&#8221; and the number after
that using regular expressions, called <tt class="docutils literal"><span class="pre">re</span></tt> in the Python standard
library, e.g.:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">re</span>
<span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="s">&#39;10,000 cell Rep1 (P1)&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="s">&#39;P1&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">long_pooled_name_to_short</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="k">if</span> <span class="s">&#39;P&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">x</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s">r&#39;P\d&#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>


<span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="n">array</span><span class="p">([</span><span class="s">u&#39;S1&#39;</span><span class="p">,</span> <span class="s">u&#39;S2&#39;</span><span class="p">,</span> <span class="s">u&#39;S3&#39;</span><span class="p">,</span> <span class="s">u&#39;S4&#39;</span><span class="p">,</span> <span class="s">u&#39;S5&#39;</span><span class="p">,</span> <span class="s">u&#39;S6&#39;</span><span class="p">,</span> <span class="s">u&#39;S7&#39;</span><span class="p">,</span> <span class="s">u&#39;S8&#39;</span><span class="p">,</span> <span class="s">u&#39;S9&#39;</span><span class="p">,</span>
       <span class="s">u&#39;S10&#39;</span><span class="p">,</span> <span class="s">u&#39;S11&#39;</span><span class="p">,</span> <span class="s">u&#39;S13&#39;</span><span class="p">,</span> <span class="s">u&#39;S14&#39;</span><span class="p">,</span> <span class="s">u&#39;S15&#39;</span><span class="p">,</span> <span class="s">u&#39;S16&#39;</span><span class="p">,</span> <span class="s">u&#39;S17&#39;</span><span class="p">,</span> <span class="s">u&#39;S18&#39;</span><span class="p">,</span>
       <span class="s">u&#39;P1&#39;</span><span class="p">,</span> <span class="s">u&#39;P2&#39;</span><span class="p">,</span> <span class="s">u&#39;P3&#39;</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">)</span>
</pre></div>
</div>
<p>And now we assign this new index as our index to the <tt class="docutils literal"><span class="pre">splicing</span></tt>
dataframe</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">splicing</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">long_pooled_name_to_short</span><span class="p">)</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>Event name</th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
    <tr>
      <th>gene</th>
      <th>Os9</th>
      <th>Vta1</th>
      <th>Bclaf1</th>
      <th>Bclaf1</th>
      <th>P4ha1</th>
      <th>Bsg</th>
      <th>Ptbp1</th>
      <th>Igf1</th>
      <th>Elk3</th>
      <th>Nbr1</th>
      <th>...</th>
      <th>Afg3l1</th>
      <th>Tep1</th>
      <th>Fgd2</th>
      <th>Ttc17</th>
      <th>Tmed1</th>
      <th>Tmed1</th>
      <th>Sbno2</th>
      <th>Synj2</th>
      <th>Tbc1d13</th>
      <th>Usp47</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> 0.84</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.42</td>
      <td>NaN</td>
      <td> 0.57</td>
      <td> 0.31</td>
      <td> 0.93</td>
      <td> 0.57</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td> 0.98</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.55</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.20</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>  NaN</td>
      <td> 0.84</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.94</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.73</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 352 columns</p>
</div><div class="section" id="remove-multi-index-columns">
<h3>Remove Multi-index columns<a class="headerlink" href="#remove-multi-index-columns" title="Permalink to this headline">¶</a></h3>
<p>Currently, <tt class="docutils literal"><span class="pre">flotilla</span></tt> only supports non-multi-index Dataframes. This
means that we need to change the columns of <tt class="docutils literal"><span class="pre">splicing</span></tt> to just the
unique event name. We&#8217;ll save this data as <tt class="docutils literal"><span class="pre">splicing_feature_data</span></tt>,
which will rename the crazy feature id to the reasonable gene name.</p>
</div>
<div class="section" id="splicing-feature-data">
<h3>Splicing Feature Data<a class="headerlink" href="#splicing-feature-data" title="Permalink to this headline">¶</a></h3>
<p>First, let&#8217;s extract the event names and gene names from <tt class="docutils literal"><span class="pre">splicing</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">event_names</span><span class="p">,</span> <span class="n">gene_names</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">splicing</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">event_names</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">u&#39;chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-&#39;</span><span class="p">,</span>
 <span class="s">u&#39;chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_names</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="s">u&#39;Os9&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Vta1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Bclaf1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Bclaf1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;P4ha1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Bsg&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Ptbp1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Igf1&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Elk3&#39;</span><span class="p">,</span>
 <span class="s">u&#39;Nbr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we can rename the columns of <tt class="docutils literal"><span class="pre">splicing</span></tt> easily</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">event_names</span>
<span class="n">splicing</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <th>chr10:79173370:79173665:+@chr10:79174001:79174029:+@chr10:79174239:79174726:+</th>
      <th>chr10:79322526:79322700:+@chr10:79322862:79322939:+@chr10:79323569:79323862:+</th>
      <th>chr10:87376364:87376545:+@chr10:87378043:87378094:+@chr10:87393420:87399792:+</th>
      <th>chr10:92747514:92747722:-@chr10:92727625:92728425:-@chr10:92717434:92717556:-</th>
      <th>chr11:101438508:101438565:+@chr11:101439246:101439351:+@chr11:101441899:101443267:+</th>
      <th>...</th>
      <th>chr8:126022488:126022598:+@chr8:126023892:126024007:+@chr8:126025133:126025333:+</th>
      <th>chr14:51455667:51455879:-@chr14:51453589:51453752:-@chr14:51453129:51453242:-</th>
      <th>chr17:29497858:29498102:+@chr17:29500656:29500887:+@chr17:29501856:29502226:+</th>
      <th>chr2:94198908:94199094:-@chr2:94182784:94182954:-@chr2:94172950:94173209:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313558:-@chr9:21311823:21312835:-</th>
      <th>chr9:21314438:21314697:-@chr9:21313375:21313795:-@chr9:21311823:21312835:-</th>
      <th>chr10:79545360:79545471:-@chr10:79542698:79544127:-@chr10:79533365:79535263:-</th>
      <th>chr17:5975579:5975881:+@chr17:5985972:5986242:+@chr17:5990136:5990361:+</th>
      <th>chr2:29997782:29997941:+@chr2:30002172:30002382:+@chr2:30002882:30003045:+</th>
      <th>chr7:119221306:119221473:+@chr7:119223686:119223745:+@chr7:119225944:119226075:+</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> 0.84</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.42</td>
      <td>NaN</td>
      <td> 0.57</td>
      <td> 0.31</td>
      <td> 0.93</td>
      <td> 0.57</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S2</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td> 0.98</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S3</th>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.02</td>
      <td> 0.55</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.20</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S4</th>
      <td>  NaN</td>
      <td> 0.84</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td> 0.04</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>S5</th>
      <td>  NaN</td>
      <td> 0.95</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td> 0.94</td>
      <td>NaN</td>
      <td>  NaN</td>
      <td> 0.73</td>
      <td>  NaN</td>
      <td>  NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 352 columns</p>
</div><p>Now let&#8217;s create <tt class="docutils literal"><span class="pre">splicing_feature_data</span></tt> to map these event names to
the gene names, and to the <tt class="docutils literal"><span class="pre">gene_category</span></tt> from before.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">event_names</span><span class="p">)</span>
<span class="n">splicing_feature_data</span><span class="p">[</span><span class="s">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_names</span>
<span class="n">splicing_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <td>    Os9</td>
    </tr>
    <tr>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <td>   Vta1</td>
    </tr>
    <tr>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <td> Bclaf1</td>
    </tr>
    <tr>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <td> Bclaf1</td>
    </tr>
    <tr>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <td>  P4ha1</td>
    </tr>
  </tbody>
</table>
</div><p>One thing we need to keep in mind is that the gene names in the
<tt class="docutils literal"><span class="pre">expression</span></tt> data were uppercase. We can convert our gene names to
uppercase with,`</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing_feature_data</span><span class="p">[</span><span class="s">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">splicing_feature_data</span><span class="p">[</span><span class="s">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="n">splicing_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <td>    OS9</td>
    </tr>
    <tr>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <td>   VTA1</td>
    </tr>
    <tr>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <td> BCLAF1</td>
    </tr>
    <tr>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <td> BCLAF1</td>
    </tr>
    <tr>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <td>  P4HA1</td>
    </tr>
  </tbody>
</table>
</div><p>Now let&#8217;s get the <tt class="docutils literal"><span class="pre">gene_category</span></tt> of these genes by doing a
<a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/merging.html">join</a> on
the splicing data and the expression data.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="n">splicing_feature_data</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">expression_feature_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s">&#39;gene_name&#39;</span><span class="p">)</span>
<span class="n">splicing_feature_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_name</th>
      <th>gene_category</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr10:126534988:126535177:-@chr10:126533971:126534135:-@chr10:126533686:126533798:-</th>
      <td>    OS9</td>
      <td>          NaN</td>
    </tr>
    <tr>
      <th>chr10:14403870:14403945:-@chr10:14395740:14395848:-@chr10:14387738:14387914:-</th>
      <td>   VTA1</td>
      <td>          NaN</td>
    </tr>
    <tr>
      <th>chr10:20051892:20052067:+@chr10:20052202:20052363:+@chr10:20053198:20053697:+</th>
      <td> BCLAF1</td>
      <td>          NaN</td>
    </tr>
    <tr>
      <th>chr10:20052864:20053378:+@chr10:20054305:20054451:+@chr10:20059515:20059727:+</th>
      <td> BCLAF1</td>
      <td>          NaN</td>
    </tr>
    <tr>
      <th>chr10:58814831:58815007:+@chr10:58817088:58817158:+@chr10:58818098:58818168:+@chr10:58824609:58824708:+</th>
      <td>  P4HA1</td>
      <td> LPS Response</td>
    </tr>
  </tbody>
</table>
</div><p>Now we have the <strong>gene_category</strong> encoded in the splicing data as well!</p>
</div>
</div>
<div class="section" id="metadata">
<h2>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">¶</a></h2>
<p>Now let&#8217;s get into creating a metadata dataframe. We&#8217;ll use the index
from the <tt class="docutils literal"><span class="pre">expression_filtered</span></tt> data to create the minimum required
column, <tt class="docutils literal"><span class="pre">'phenotype'</span></tt>, which has the name of the phenotype of that
cell. And we&#8217;ll also add the column <tt class="docutils literal"><span class="pre">'pooled'</span></tt> to indicate whether
this sample is pooled or not.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">expression_filtered</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;phenotype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;BDMC&#39;</span>
<span class="n">metadata</span><span class="p">[</span><span class="s">&#39;pooled&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;P&#39;</span><span class="p">))</span>

<span class="n">metadata</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
      <th>pooled</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S2</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S3</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S4</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S5</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S6</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S7</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S8</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S9</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S10</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S11</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S12</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S13</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S14</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S15</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S16</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S17</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>S18</th>
      <td> BDMC</td>
      <td> False</td>
    </tr>
    <tr>
      <th>P1</th>
      <td> BDMC</td>
      <td>  True</td>
    </tr>
    <tr>
      <th>P2</th>
      <td> BDMC</td>
      <td>  True</td>
    </tr>
    <tr>
      <th>P3</th>
      <td> BDMC</td>
      <td>  True</td>
    </tr>
  </tbody>
</table>
</div></div>
<div class="section" id="mapping-stats-data">
<h2>Mapping stats data<a class="headerlink" href="#mapping-stats-data" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table1.xls&#39;</span><span class="p">,</span> <span class="n">sheetname</span><span class="o">=</span><span class="s">&#39;SuppTable1 2.txt&#39;</span><span class="p">)</span>
<span class="n">mapping_stats</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>PF_READS</th>
      <th>PCT_MAPPED_GENOME</th>
      <th>PCT_RIBOSOMAL_BASES</th>
      <th>MEDIAN_CV_COVERAGE</th>
      <th>MEDIAN_5PRIME_BIAS</th>
      <th>MEDIAN_3PRIME_BIAS</th>
      <th>MEDIAN_5PRIME_TO_3PRIME_BIAS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0 </th>
      <td>       S1</td>
      <td> 21326048</td>
      <td> 0.706590</td>
      <td> 0.006820</td>
      <td> 0.509939</td>
      <td> 0.092679</td>
      <td> 0.477321</td>
      <td> 0.247741</td>
    </tr>
    <tr>
      <th>1 </th>
      <td>       S2</td>
      <td> 27434011</td>
      <td> 0.745385</td>
      <td> 0.004111</td>
      <td> 0.565732</td>
      <td> 0.056583</td>
      <td> 0.321053</td>
      <td> 0.244062</td>
    </tr>
    <tr>
      <th>2 </th>
      <td>       S3</td>
      <td> 31142391</td>
      <td> 0.722087</td>
      <td> 0.006428</td>
      <td> 0.540341</td>
      <td> 0.079551</td>
      <td> 0.382286</td>
      <td> 0.267367</td>
    </tr>
    <tr>
      <th>3 </th>
      <td>       S4</td>
      <td> 26231852</td>
      <td> 0.737854</td>
      <td> 0.004959</td>
      <td> 0.530978</td>
      <td> 0.067041</td>
      <td> 0.351670</td>
      <td> 0.279782</td>
    </tr>
    <tr>
      <th>4 </th>
      <td>       S5</td>
      <td> 29977214</td>
      <td> 0.746466</td>
      <td> 0.006121</td>
      <td> 0.525598</td>
      <td> 0.066543</td>
      <td> 0.353995</td>
      <td> 0.274252</td>
    </tr>
    <tr>
      <th>5 </th>
      <td>       S6</td>
      <td> 24148387</td>
      <td> 0.730079</td>
      <td> 0.008794</td>
      <td> 0.529650</td>
      <td> 0.072095</td>
      <td> 0.413696</td>
      <td> 0.225929</td>
    </tr>
    <tr>
      <th>6 </th>
      <td>       S7</td>
      <td> 24078116</td>
      <td> 0.730638</td>
      <td> 0.007945</td>
      <td> 0.540913</td>
      <td> 0.051991</td>
      <td> 0.358597</td>
      <td> 0.201984</td>
    </tr>
    <tr>
      <th>7 </th>
      <td>       S8</td>
      <td> 25032126</td>
      <td> 0.739989</td>
      <td> 0.004133</td>
      <td> 0.512725</td>
      <td> 0.058783</td>
      <td> 0.373509</td>
      <td> 0.212337</td>
    </tr>
    <tr>
      <th>8 </th>
      <td>       S9</td>
      <td> 22257682</td>
      <td> 0.747427</td>
      <td> 0.004869</td>
      <td> 0.521622</td>
      <td> 0.063566</td>
      <td> 0.334294</td>
      <td> 0.240641</td>
    </tr>
    <tr>
      <th>9 </th>
      <td>      S10</td>
      <td> 29436289</td>
      <td> 0.748795</td>
      <td> 0.005499</td>
      <td> 0.560454</td>
      <td> 0.036219</td>
      <td> 0.306729</td>
      <td> 0.187479</td>
    </tr>
    <tr>
      <th>10</th>
      <td>      S11</td>
      <td> 31130278</td>
      <td> 0.741882</td>
      <td> 0.002740</td>
      <td> 0.558882</td>
      <td> 0.049581</td>
      <td> 0.349191</td>
      <td> 0.211787</td>
    </tr>
    <tr>
      <th>11</th>
      <td>      S12</td>
      <td> 21161595</td>
      <td> 0.750782</td>
      <td> 0.006837</td>
      <td> 0.756339</td>
      <td> 0.013878</td>
      <td> 0.324264</td>
      <td> 0.195430</td>
    </tr>
    <tr>
      <th>12</th>
      <td>      S13</td>
      <td> 28612833</td>
      <td> 0.733976</td>
      <td> 0.011718</td>
      <td> 0.598687</td>
      <td> 0.035392</td>
      <td> 0.357447</td>
      <td> 0.198566</td>
    </tr>
    <tr>
      <th>13</th>
      <td>      S14</td>
      <td> 26351189</td>
      <td> 0.748323</td>
      <td> 0.004106</td>
      <td> 0.517518</td>
      <td> 0.070293</td>
      <td> 0.381095</td>
      <td> 0.259122</td>
    </tr>
    <tr>
      <th>14</th>
      <td>      S15</td>
      <td> 25739575</td>
      <td> 0.748421</td>
      <td> 0.003353</td>
      <td> 0.526238</td>
      <td> 0.050938</td>
      <td> 0.324207</td>
      <td> 0.212366</td>
    </tr>
    <tr>
      <th>15</th>
      <td>      S16</td>
      <td> 26802346</td>
      <td> 0.739833</td>
      <td> 0.009370</td>
      <td> 0.520287</td>
      <td> 0.071503</td>
      <td> 0.358758</td>
      <td> 0.240009</td>
    </tr>
    <tr>
      <th>16</th>
      <td>      S17</td>
      <td> 26343522</td>
      <td> 0.749358</td>
      <td> 0.003155</td>
      <td> 0.673195</td>
      <td> 0.024121</td>
      <td> 0.301588</td>
      <td> 0.245854</td>
    </tr>
    <tr>
      <th>17</th>
      <td>      S18</td>
      <td> 25290073</td>
      <td> 0.749358</td>
      <td> 0.007465</td>
      <td> 0.562382</td>
      <td> 0.048528</td>
      <td> 0.314776</td>
      <td> 0.215160</td>
    </tr>
    <tr>
      <th>18</th>
      <td> 10k_rep1</td>
      <td> 28247826</td>
      <td> 0.688553</td>
      <td> 0.018993</td>
      <td> 0.547000</td>
      <td> 0.056113</td>
      <td> 0.484393</td>
      <td> 0.140333</td>
    </tr>
    <tr>
      <th>19</th>
      <td> 10k_rep2</td>
      <td> 39303876</td>
      <td> 0.690313</td>
      <td> 0.017328</td>
      <td> 0.547621</td>
      <td> 0.055600</td>
      <td> 0.474634</td>
      <td> 0.142889</td>
    </tr>
    <tr>
      <th>20</th>
      <td> 10k_rep3</td>
      <td> 29831281</td>
      <td> 0.710875</td>
      <td> 0.010610</td>
      <td> 0.518053</td>
      <td> 0.066053</td>
      <td> 0.488738</td>
      <td> 0.168180</td>
    </tr>
    <tr>
      <th>21</th>
      <td>   MB_SC1</td>
      <td> 13848219</td>
      <td> 0.545000</td>
      <td> 0.007000</td>
      <td> 0.531495</td>
      <td> 0.127934</td>
      <td> 0.207841</td>
      <td> 0.728980</td>
    </tr>
    <tr>
      <th>22</th>
      <td>   MB_SC2</td>
      <td> 13550218</td>
      <td> 0.458000</td>
      <td> 0.010800</td>
      <td> 0.569271</td>
      <td> 0.102581</td>
      <td> 0.179407</td>
      <td> 0.694747</td>
    </tr>
    <tr>
      <th>23</th>
      <td>   MB_SC3</td>
      <td> 26765848</td>
      <td> 0.496000</td>
      <td> 0.007900</td>
      <td> 0.535192</td>
      <td> 0.141893</td>
      <td> 0.231068</td>
      <td> 0.722080</td>
    </tr>
  </tbody>
</table>
</div></div>
<div class="section" id="create-a-flotilla-study">
<h2>Create a <tt class="docutils literal"><span class="pre">flotilla</span></tt> Study!<a class="headerlink" href="#create-a-flotilla-study" title="Permalink to this headline">¶</a></h2>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">Study</span><span class="p">(</span><span class="c"># The metadata describing phenotype and pooled samples</span>
                       <span class="n">metadata</span><span class="p">,</span>

                       <span class="c"># A version for this data</span>
                       <span class="n">version</span><span class="o">=</span><span class="s">&#39;0.1.0&#39;</span><span class="p">,</span>

                       <span class="c"># Dataframe of the filtered expression data</span>
                       <span class="n">expression_data</span><span class="o">=</span><span class="n">expression_filtered</span><span class="p">,</span>

                       <span class="c"># Dataframe of the feature data of the genes</span>
                       <span class="n">expression_feature_data</span><span class="o">=</span><span class="n">expression_feature_data</span><span class="p">,</span>

                       <span class="c"># Dataframe of the splicing data</span>
                       <span class="n">splicing_data</span><span class="o">=</span><span class="n">splicing</span><span class="p">,</span>

                       <span class="c"># Dataframe of the feature data of the splicing events</span>
                       <span class="n">splicing_feature_data</span><span class="o">=</span><span class="n">splicing_feature_data</span><span class="p">,</span>

                       <span class="c"># Specify &quot;gene_name&quot; as the column we want to rename the splicing ids to</span>
                       <span class="n">splicing_feature_rename_col</span><span class="o">=</span><span class="s">&quot;gene_name&quot;</span><span class="p">,</span>

                       <span class="c"># Specify &quot;gene_name&quot; as the column that links splicing ids to expression ids</span>
                       <span class="n">splicing_feature_expression_id_col</span><span class="o">=</span><span class="s">&quot;gene_name&quot;</span><span class="p">,</span>

                       <span class="c"># Dataframe of the mapping stats data</span>
                       <span class="n">mapping_stats_data</span><span class="o">=</span><span class="n">mapping_stats</span><span class="p">,</span>

                       <span class="c"># Which column in &quot;mapping_stats&quot; has the number of reads</span>
                       <span class="n">mapping_stats_number_mapped_col</span><span class="o">=</span><span class="s">&#39;PF_READS&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>2014-12-10 15:36:38 Initializing Study
2014-12-10 15:36:38 Initializing Predictor configuration manager for Study
2014-12-10 15:36:38 Predictor ExtraTreesClassifier is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesClassifier&#39;&gt;
2014-12-10 15:36:38 Added ExtraTreesClassifier to default predictors
2014-12-10 15:36:38 Predictor ExtraTreesRegressor is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesRegressor&#39;&gt;
2014-12-10 15:36:38 Added ExtraTreesRegressor to default predictors
2014-12-10 15:36:38 Predictor GradientBoostingClassifier is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingClassifier&#39;&gt;
2014-12-10 15:36:38 Added GradientBoostingClassifier to default predictors
2014-12-10 15:36:38 Predictor GradientBoostingRegressor is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingRegressor&#39;&gt;
2014-12-10 15:36:38 Added GradientBoostingRegressor to default predictors
2014-12-10 15:36:38 Loading metadata
2014-12-10 15:36:38 Loading expression data
2014-12-10 15:36:38 Initializing expression
2014-12-10 15:36:38 Done initializing expression
2014-12-10 15:36:38 Loading splicing data
2014-12-10 15:36:38 Initializing splicing
2014-12-10 15:36:38 Done initializing splicing
2014-12-10 15:36:38 Successfully initialized a Study object!
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>No phenotype to color mapping was provided, so coming up with reasonable defaults
No phenotype to marker (matplotlib plotting symbol) was provided, so each phenotype will be plotted as a circle in the PCA visualizations.
</pre></div>
</div>
<p>As a side note, you can save this study to disk now, so you can
&#8220;<tt class="docutils literal"><span class="pre">embark</span></tt>&#8221; later:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Wrote datapackage to /Users/olga/flotilla_projects/shalek2013/datapackage.json
</pre></div>
</div>
<p>Note that this is saved to my home directory, in
<tt class="docutils literal"><span class="pre">~/flotilla_projects/&lt;study_name&gt;/</span></tt> (the &#8220;<tt class="docutils literal"><span class="pre">~</span></tt>&#8221; stands for my &#8220;home
directory&#8221;, in this case <tt class="docutils literal"><span class="pre">/Users/olga</span></tt>). This will be saved in your
home directory, too.</p>
<p>The <tt class="docutils literal"><span class="pre">datapackage.json</span></tt> file is what holds all the information relative
to the study, and loosely follows the <a class="reference external" href="http://data.okfn.org/doc/data-package">datapackage
spec</a> created by the Open
Knowledge Foundation.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">cat</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span><span class="o">/</span><span class="n">datapackage</span><span class="o">.</span><span class="n">json</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">{</span>
  <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;shalek2013&quot;</span><span class="p">,</span>
  <span class="s">&quot;title&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
  <span class="s">&quot;datapackage_version&quot;</span><span class="p">:</span> <span class="s">&quot;0.1.1&quot;</span><span class="p">,</span>
  <span class="s">&quot;sources&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
  <span class="s">&quot;licenses&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
  <span class="s">&quot;resources&quot;</span><span class="p">:</span> <span class="p">[</span>
    <span class="p">{</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;splicing.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;splicing&quot;</span><span class="p">,</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s">&quot;number_mapped_col&quot;</span><span class="p">:</span> <span class="s">&quot;PF_READS&quot;</span><span class="p">,</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;mapping_stats.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;mapping_stats&quot;</span><span class="p">,</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;expression_feature&quot;</span><span class="p">,</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;rename_col&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
      <span class="s">&quot;ignore_subset_cols&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;expression_feature.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;expression&quot;</span><span class="p">,</span>
      <span class="s">&quot;log_base&quot;</span><span class="p">:</span> <span class="n">null</span><span class="p">,</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;thresh&quot;</span><span class="p">:</span> <span class="o">-</span><span class="n">Infinity</span><span class="p">,</span>
      <span class="s">&quot;plus_one&quot;</span><span class="p">:</span> <span class="n">false</span><span class="p">,</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;expression.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;splicing_feature&quot;</span><span class="p">,</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;rename_col&quot;</span><span class="p">:</span> <span class="s">&quot;gene_name&quot;</span><span class="p">,</span>
      <span class="s">&quot;ignore_subset_cols&quot;</span><span class="p">:</span> <span class="p">[],</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;splicing_feature.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;expression_id_col&quot;</span><span class="p">:</span> <span class="s">&quot;gene_name&quot;</span><span class="p">,</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">},</span>
    <span class="p">{</span>
      <span class="s">&quot;pooled_col&quot;</span><span class="p">:</span> <span class="s">&quot;pooled&quot;</span><span class="p">,</span>
      <span class="s">&quot;name&quot;</span><span class="p">:</span> <span class="s">&quot;metadata&quot;</span><span class="p">,</span>
      <span class="s">&quot;phenotype_to_marker&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s">&quot;BDMC&quot;</span><span class="p">:</span> <span class="s">&quot;o&quot;</span>
      <span class="p">},</span>
      <span class="s">&quot;format&quot;</span><span class="p">:</span> <span class="s">&quot;csv&quot;</span><span class="p">,</span>
      <span class="s">&quot;minimum_samples&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
      <span class="s">&quot;phenotype_to_color&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s">&quot;BDMC&quot;</span><span class="p">:</span> <span class="s">&quot;#1b9e77&quot;</span>
      <span class="p">},</span>
      <span class="s">&quot;path&quot;</span><span class="p">:</span> <span class="s">&quot;metadata.csv.gz&quot;</span><span class="p">,</span>
      <span class="s">&quot;phenotype_col&quot;</span><span class="p">:</span> <span class="s">&quot;phenotype&quot;</span><span class="p">,</span>
      <span class="s">&quot;phenotype_order&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s">&quot;BDMC&quot;</span>
      <span class="p">],</span>
      <span class="s">&quot;compression&quot;</span><span class="p">:</span> <span class="s">&quot;gzip&quot;</span>
    <span class="p">}</span>
  <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>One thing to note is that when you save, the version number is bumped
up. <tt class="docutils literal"><span class="pre">study.version</span></tt> (the one we just made) is <tt class="docutils literal"><span class="pre">0.1.0</span></tt>, but the one
we saved is <tt class="docutils literal"><span class="pre">0.1.1</span></tt>, since we could have made some changes to the
data.</p>
<p>Let&#8217;s look at what else is in this folder:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">ls</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">olga</span><span class="o">/</span><span class="n">flotilla_projects</span><span class="o">/</span><span class="n">shalek2013</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>datapackage.json           expression_feature.csv     mapping_stats.csv.gz       splicing.csv               splicing_feature.csv.gz
expression.csv             expression_feature.csv.gz  metadata.csv               splicing.csv.gz
expression.csv.gz          mapping_stats.csv          metadata.csv.gz            splicing_feature.csv
</pre></div>
</div>
<p>So this is where all the other files are. Good to know!</p>
<p>We can &#8220;embark&#8221; on this newly-saved study now very painlessly, without
having to open and process all those files again:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study2</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">embark</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>2014-12-10 15:34:27 Reading datapackage from /Users/olga/flotilla_projects/shalek2013/datapackage.json
2014-12-10 15:34:27 Parsing datapackage to create a Study object
2014-12-10 15:34:27 Initializing Study
2014-12-10 15:34:27 Initializing Predictor configuration manager for Study
2014-12-10 15:34:27 Predictor ExtraTreesClassifier is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesClassifier&#39;&gt;
2014-12-10 15:34:27 Added ExtraTreesClassifier to default predictors
2014-12-10 15:34:27 Predictor ExtraTreesRegressor is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesRegressor&#39;&gt;
2014-12-10 15:34:27 Added ExtraTreesRegressor to default predictors
2014-12-10 15:34:27 Predictor GradientBoostingClassifier is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingClassifier&#39;&gt;
2014-12-10 15:34:27 Added GradientBoostingClassifier to default predictors
2014-12-10 15:34:27 Predictor GradientBoostingRegressor is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingRegressor&#39;&gt;
2014-12-10 15:34:27 Added GradientBoostingRegressor to default predictors
2014-12-10 15:34:27 Loading metadata
2014-12-10 15:34:27 Loading expression data
2014-12-10 15:34:27 Initializing expression
2014-12-10 15:34:27 Done initializing expression
2014-12-10 15:34:27 Loading splicing data
2014-12-10 15:34:27 Initializing splicing
2014-12-10 15:34:27 Done initializing splicing
2014-12-10 15:34:27 Successfully initialized a Study object!
</pre></div>
</div>
<p>Now we can start creating figures!</p>
</div>
<div class="section" id="figure-1">
<h2>Figure 1<a class="headerlink" href="#figure-1" title="Permalink to this headline">¶</a></h2>
<p>Here, we will attempt to re-create the sub-panels in <a class="reference external" href="http://www.nature.com/nature/journal/v498/n7453/fig_tab/nature12172_F1.html">Figure
1</a>,
where the original is:</p>
<div class="figure align-center">
<img alt="Original Figure 1" src="http://www.nature.com/nature/journal/v498/n7453/images/nature12172-f1.2.jpg" />
<p class="caption">Original Figure 1</p>
</div>
<div class="section" id="figure-1a">
<h3>Figure 1a<a class="headerlink" href="#figure-1a" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="s">&#39;P1&#39;</span><span class="p">,</span> <span class="s">&#39;P2&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>/usr/local/lib/python2.7/site-packages/matplotlib/figure.py:1644: UserWarning: This figure includes Axes that are not compatible with tight_layout, so its results might be incorrect.
  warnings.warn(&quot;This figure includes Axes that are not &quot;
</pre></div>
</div>
<img alt="../_images/shalek2013_72_1.png" src="../_images/shalek2013_72_1.png" />
<div class="section" id="without-flotilla-you-would-do">
<h4>Without flotilla, you would do<a class="headerlink" href="#without-flotilla-you-would-do" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s">&#39;ticks&#39;</span><span class="p">)</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">expression_filtered</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="s">&#39;P1&#39;</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">expression_filtered</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="s">&#39;P2&#39;</span><span class="p">]</span>
<span class="n">jointgrid</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">joint_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>
<span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">axis</span><span class="p">()</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_74_0.png" src="../_images/shalek2013_74_0.png" />
</div>
</div>
<div class="section" id="figure-1b">
<h3>Figure 1b<a class="headerlink" href="#figure-1b" title="Permalink to this headline">¶</a></h3>
<p>Paper: <span class="math">\(r=0.54\)</span>. Not sure at all what&#8217;s going on here.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="s">&#39;S1&#39;</span><span class="p">,</span> <span class="s">&#39;S2&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_77_0.png" src="../_images/shalek2013_77_0.png" />
<div class="section" id="without-flotilla">
<h4>Without <tt class="docutils literal"><span class="pre">flotilla</span></tt><a class="headerlink" href="#without-flotilla" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s">&#39;ticks&#39;</span><span class="p">)</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">expression_filtered</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="s">&#39;S1&#39;</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">expression_filtered</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="s">&#39;S2&#39;</span><span class="p">]</span>
<span class="n">jointgrid</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">joint_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>

<span class="c"># Adjust xmin, ymin to 0</span>
<span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">axis</span><span class="p">()</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_79_0.png" src="../_images/shalek2013_79_0.png" />
<p>By the way, you can do other kinds of plots with <tt class="docutils literal"><span class="pre">flotilla</span></tt>, like a
kernel density estimate (&#8220;<tt class="docutils literal"><span class="pre">kde</span></tt>&#8221;) plot:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="s">&#39;S1&#39;</span><span class="p">,</span> <span class="s">&#39;S2&#39;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;kde&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_81_0.png" src="../_images/shalek2013_81_0.png" />
<p>Or a binned hexagon plot (&#8220;<tt class="docutils literal"><span class="pre">hexbin&quot;</span></tt>):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="s">&#39;S1&#39;</span><span class="p">,</span> <span class="s">&#39;S2&#39;</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;hexbin&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_83_0.png" src="../_images/shalek2013_83_0.png" />
<p>Any inputs that are valid to <tt class="docutils literal"><span class="pre">seaborn</span></tt>&#8216;s
<tt class="docutils literal"><span class="pre">`jointplot</span></tt> &lt;<a class="reference external" href="http://web.stanford.edu/~mwaskom/software/seaborn/generated/seaborn.jointplot.html#seaborn.jointplot">http://web.stanford.edu/~mwaskom/software/seaborn/generated/seaborn.jointplot.html#seaborn.jointplot</a>&gt;`_
are valid.</p>
</div>
</div>
<div class="section" id="figure-1c">
<h3>Figure 1c<a class="headerlink" href="#figure-1c" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">x</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="s">&#39;P1&#39;</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">y</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s">&quot;Average singles&quot;</span>

<span class="n">jointgrid</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">joint_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>

<span class="c"># Adjust xmin, ymin to 0</span>
<span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">axis</span><span class="p">()</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_86_0.png" src="../_images/shalek2013_86_0.png" />
</div>
</div>
<div class="section" id="figure-2">
<h2>Figure 2<a class="headerlink" href="#figure-2" title="Permalink to this headline">¶</a></h2>
<p>Next, we will attempt to recreate the figures from <a class="reference external" href="http://www.nature.com/nature/journal/v498/n7453/fig_tab/nature12172_F2.html">Figure
2</a>:</p>
<div class="figure align-center">
<img alt="Original figure 2" src="http://www.nature.com/nature/journal/v498/n7453/images/nature12172-f2.2.jpg" />
<p class="caption">Original figure 2</p>
</div>
<div class="section" id="figure-2a">
<h3>Figure 2a<a class="headerlink" href="#figure-2a" title="Permalink to this headline">¶</a></h3>
<p>For this figure, we will need the &#8220;LPS Response&#8221; and &#8220;Housekeeping&#8221; gene
annotations, from the <tt class="docutils literal"><span class="pre">expression_feature_data</span></tt> that we created.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="c"># Get colors for plotting the gene categories</span>
<span class="n">dark2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s">&#39;Dark2&#39;</span><span class="p">)</span>

<span class="n">singles</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span>
<span class="c"># Get only gene categories for genes in the singles data</span>
<span class="n">singles</span><span class="p">,</span> <span class="n">gene_categories</span> <span class="o">=</span> <span class="n">singles</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_data</span><span class="o">.</span><span class="n">gene_category</span><span class="p">,</span> <span class="n">join</span><span class="o">=</span><span class="s">&#39;left&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">mean</span> <span class="o">=</span> <span class="n">singles</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">std</span> <span class="o">=</span> <span class="n">singles</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

<span class="n">jointgrid</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">mean</span><span class="p">,</span> <span class="n">std</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;#262626&#39;</span><span class="p">,</span> <span class="n">joint_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">category</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_categories</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">gene_categories</span><span class="p">)):</span>
    <span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">mean</span><span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">index</span><span class="p">],</span> <span class="n">std</span><span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">index</span><span class="p">],</span> <span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">dark2</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;Standard deviation in single cells $\mu$&#39;</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Average expression in single cells $\sigma$&#39;</span><span class="p">)</span>

<span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">axis</span><span class="p">()</span>
<span class="n">vmax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">xmax</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>
<span class="n">vmin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">xmin</span><span class="p">,</span> <span class="n">ymin</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;steelblue&#39;</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">.</span><span class="mi">25</span><span class="o">*</span><span class="n">vmax</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;grey&#39;</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="p">)</span>
<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>

<span class="n">jointgrid</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">fill_betweenx</span><span class="p">((</span><span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">250</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=-</span><span class="mi">1</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_91_0.png" src="../_images/shalek2013_91_0.png" />
<p>I couldn&#8217;t find the data for the <a href="#id13"><span class="problematic" id="id14">``</span></a>hESC``s for the right-side panel of
Fig. 2a, so I couldn&#8217;t remake the figure.</p>
</div>
<div class="section" id="figure-2b">
<h3>Figure 2b<a class="headerlink" href="#figure-2b" title="Permalink to this headline">¶</a></h3>
<p>In the paper, they use <em>&#8220;522 most highly expressed genes (single-cell
average TPM &gt; 250)&#8221;</em>, but I wasn&#8217;t able to replicate their numbers. If I
use the pre-filtered expression data that I fed into flotilla, then I
get 297 genes:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mean</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">highly_expressed_genes</span> <span class="o">=</span> <span class="n">mean</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">mean</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">250</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">highly_expressed_genes</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="mi">297</span>
</pre></div>
</div>
<p>Which is much less. If I use the original, unfiltered data, then I get
the <em>&#8220;522&#8221;</em> number, but this seems strange because they did the
filtering step of <em>&#8220;discarded genes not appreciably expressed
(transcripts per million (TPM) &gt; 1) in at least three individual cells,
retaining 6,313 genes for further analysis&#8221;</em>, and yet they went back to
the original data to get this new subset.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">250</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">522</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">expression_highly_expressed</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">,</span> <span class="n">expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">250</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">mean</span> <span class="o">=</span> <span class="n">expression_highly_expressed</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

<span class="n">std</span> <span class="o">=</span> <span class="n">expression_highly_expressed</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>

<span class="n">mean_bins</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">mean</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">11</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>

<span class="c"># Coefficient of variation</span>
<span class="n">cv</span> <span class="o">=</span> <span class="n">std</span><span class="o">/</span><span class="n">mean</span>
<span class="n">cv</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="n">genes</span> <span class="o">=</span> <span class="n">mean</span><span class="o">.</span><span class="n">index</span>


<span class="c"># for name, df in shalek2013.expression.singles.groupby(dict(zip(genes, mean_bins)), axis=1):</span>
<span class="k">def</span> <span class="nf">calculate_cells_per_tpm_per_cv</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cv</span><span class="p">):</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">df_aligned</span><span class="p">,</span> <span class="n">cv_aligned</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">cv</span><span class="p">,</span> <span class="n">join</span><span class="o">=</span><span class="s">&#39;inner&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">cv_aligned</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">n_cells</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">cv</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">n_cells</span><span class="p">[</span><span class="n">cv_aligned</span><span class="o">.</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_aligned</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">cv_aligned</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">n_cells</span>

<span class="n">grouped</span> <span class="o">=</span> <span class="n">expression_highly_expressed</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">mean_bins</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">cells_per_tpm_per_cv</span> <span class="o">=</span> <span class="n">grouped</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">calculate_cells_per_tpm_per_cv</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="n">cv</span><span class="p">)</span>
</pre></div>
</div>
<p>Here&#8217;s how you would make the original figure from the paper:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">cells_per_tpm_per_cv</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;ln(TPM, binned)&#39;</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_100_0.png" src="../_images/shalek2013_100_0.png" />
<p>Doesn&#8217;t quite look the same. Maybe the y-axis labels were opposite, and
higher up on the y-axis was less variant? Because I see a blob of color
for (1,2] TPM (by the way, the figure in the paper is not TPM+1 as
previous figures were)</p>
<p>This is how you would make a modified version of the figure, which also
plots the coefficient of variation on a side-plot, which I like because
it shows the CV changes directly on the heatmap. Also, technically this
is <span class="math">\(\ln\)</span>(TPM+1).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">gridspec</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">cv_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">heatmap_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">cells_per_tpm_per_cv</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">heatmap_ax</span><span class="p">)</span>
<span class="n">heatmap_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">heatmap_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;$\ln$(TPM+1), binned&#39;</span><span class="p">)</span>

<span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">cv</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">set_xscale</span><span class="p">(</span><span class="s">&#39;log&#39;</span><span class="p">)</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">cv</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;#262626&#39;</span><span class="p">)</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">fill_betweenx</span><span class="p">(</span><span class="n">cv</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">cv</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="n">y</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;#262626&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;CV = $\mu/\sigma$&#39;</span><span class="p">)</span>
<span class="n">cv_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">cv_ax</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_102_0.png" src="../_images/shalek2013_102_0.png" />
</div>
</div>
<div class="section" id="figure-3">
<h2>Figure 3<a class="headerlink" href="#figure-3" title="Permalink to this headline">¶</a></h2>
<p>We will attempt to re-create the sub-panel figures from <a class="reference external" href="http://www.nature.com/nature/journal/v498/n7453/fig_tab/nature12172_F3.html">Figure
3</a>:</p>
<div class="figure align-center">
<img alt="Original Figure 3" src="http://www.nature.com/nature/journal/v498/n7453/images/nature12172-f3.2.jpg" />
<p class="caption">Original Figure 3</p>
</div>
<p>Since we can&#8217;t re-do the microscopy (Figure 3a) or the RNA-FISH counts
(Figure 3c), we will make Figures 3b. These histograms are simple to do
outside of <tt class="docutils literal"><span class="pre">flotilla</span></tt>, so we do not have them within flotilla.</p>
<div class="section" id="figure-3b-top-panel">
<h3>Figure 3b, top panel<a class="headerlink" href="#figure-3b-top-panel" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">singles</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flat</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.05</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_106_0.png" src="../_images/shalek2013_106_0.png" />
</div>
<div class="section" id="figure-3b-bottom-panel">
<h3>Figure 3b, bottom panel<a class="headerlink" href="#figure-3b-bottom-panel" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">study</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flat</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.05</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;grey&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_108_0.png" src="../_images/shalek2013_108_0.png" />
</div>
</div>
<div class="section" id="figure-4">
<h2>Figure 4<a class="headerlink" href="#figure-4" title="Permalink to this headline">¶</a></h2>
<p>We will attempt to re-create the sub-panel figures from <a class="reference external" href="http://www.nature.com/nature/journal/v498/n7453/fig_tab/nature12172_F4.html">Figure
4</a>:</p>
<div class="figure align-center">
<img alt="Original Figure 4" src="http://www.nature.com/nature/journal/v498/n7453/images/nature12172-f4.2.jpg" />
<p class="caption">Original Figure 4</p>
</div>
<div class="section" id="figure-4a">
<h3>Figure 4a<a class="headerlink" href="#figure-4a" title="Permalink to this headline">¶</a></h3>
<p>Here, we can use the &#8220;<tt class="docutils literal"><span class="pre">interactive_pca</span></tt>&#8221; function we have to explore
different dimensionality reductions on the data.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">interactive_pca</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>featurewise : False
y_pc : 2
data_type : expression
show_point_labels : False
sample_subset : all_samples
feature_subset : variant
plot_violins : False
x_pc : 1
list_link :
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>&lt;function flotilla.visualize.ipython_interact.do_interact&gt;
</pre></div>
</div>
<img alt="../_images/shalek2013_113_2.png" src="../_images/shalek2013_113_2.png" />
<p>A &#8220;sequences shortened&#8221; version of this is available as a gif:</p>
<div class="figure align-center">
<img alt="Imgur" src="http://i.imgur.com/fJKPQ7W.gif" />
<p class="caption">Imgur</p>
</div>
<p>Equivalently, I could have written out the plotting command by hand,
instead of using <tt class="docutils literal"><span class="pre">study.interactive_pca</span></tt>:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">feature_subset</span><span class="o">=</span><span class="s">&#39;gene_category: LPS Response&#39;</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="s">&#39;not (pooled)&#39;</span><span class="p">,</span> <span class="n">plot_violins</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>&lt;flotilla.visualize.decomposition.DecompositionViz at 0x1125d52d0&gt;
</pre></div>
</div>
<img alt="../_images/shalek2013_116_1.png" src="../_images/shalek2013_116_1.png" />
</div>
<div class="section" id="mark-immature-cells-as-a-new-subset">
<h3>Mark immature cells as a new subset<a class="headerlink" href="#mark-immature-cells-as-a-new-subset" title="Permalink to this headline">¶</a></h3>
<p>As in the paper, the cells S12, S13, and S16 appear in a cluster that is
separate from the remaining cells. From the paper, these were the
&#8220;matured&#8221; bone-marrow derived dendritic cells, after stimulation with a
lipopolysaccharide. We can mark these as mature in our metadata,</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">mature</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;S12&#39;</span><span class="p">,</span> <span class="s">&#39;S13&#39;</span><span class="p">,</span> <span class="s">&#39;S16&#39;</span><span class="p">]</span>
<span class="n">study</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;maturity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">&#39;mature&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">mature</span> <span class="k">else</span> <span class="s">&#39;immature&#39;</span><span class="p">)</span>
<span class="n">study</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
      <th>pooled</th>
      <th>outlier</th>
      <th>maturity</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> BDMC</td>
      <td> False</td>
      <td> False</td>
      <td> immature</td>
    </tr>
    <tr>
      <th>S2</th>
      <td> BDMC</td>
      <td> False</td>
      <td> False</td>
      <td> immature</td>
    </tr>
    <tr>
      <th>S3</th>
      <td> BDMC</td>
      <td> False</td>
      <td> False</td>
      <td> immature</td>
    </tr>
    <tr>
      <th>S4</th>
      <td> BDMC</td>
      <td> False</td>
      <td> False</td>
      <td> immature</td>
    </tr>
    <tr>
      <th>S5</th>
      <td> BDMC</td>
      <td> False</td>
      <td> False</td>
      <td> immature</td>
    </tr>
  </tbody>
</table>
</div><p>Then, we can set <strong>maturity</strong> as the column we use for coloring the PCA,
since before it was the &#8220;phenotype&#8221; column.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_col</span> <span class="o">=</span> <span class="s">&#39;maturity&#39;</span>
<span class="n">study</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
<span class="n">study</span> <span class="o">=</span> <span class="n">flotilla</span><span class="o">.</span><span class="n">embark</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Wrote datapackage to /Users/olga/flotilla_projects/shalek2013/datapackage.json2014-12-10 15:41:07   Reading datapackage from /Users/olga/flotilla_projects/shalek2013/datapackage.json
2014-12-10 15:41:07 Parsing datapackage to create a Study object
2014-12-10 15:41:07 Initializing Study
2014-12-10 15:41:07 Initializing Predictor configuration manager for Study
2014-12-10 15:41:07 Predictor ExtraTreesClassifier is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesClassifier&#39;&gt;
2014-12-10 15:41:07 Added ExtraTreesClassifier to default predictors
2014-12-10 15:41:07 Predictor ExtraTreesRegressor is of type &lt;class &#39;sklearn.ensemble.forest.ExtraTreesRegressor&#39;&gt;
2014-12-10 15:41:07 Added ExtraTreesRegressor to default predictors
2014-12-10 15:41:07 Predictor GradientBoostingClassifier is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingClassifier&#39;&gt;
2014-12-10 15:41:07 Added GradientBoostingClassifier to default predictors
2014-12-10 15:41:07 Predictor GradientBoostingRegressor is of type &lt;class &#39;sklearn.ensemble.gradient_boosting.GradientBoostingRegressor&#39;&gt;
2014-12-10 15:41:07 Added GradientBoostingRegressor to default predictors
2014-12-10 15:41:07 Loading metadata
2014-12-10 15:41:07 Loading expression data
2014-12-10 15:41:07 Initializing expression
2014-12-10 15:41:07 Done initializing expression
2014-12-10 15:41:07 Loading splicing data
2014-12-10 15:41:07 Initializing splicing
2014-12-10 15:41:07 Done initializing splicing
2014-12-10 15:41:07 Successfully initialized a Study object!
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>No color was assigned to the phenotype immature, assigning a random colorNo color was assigned to the phenotype mature, assigning a random colorimmature does not have marker style, falling back on &quot;o&quot; (circle)mature does not have marker style, falling back on &quot;o&quot; (circle)
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">feature_subset</span><span class="o">=</span><span class="s">&#39;gene_category: LPS Response&#39;</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="s">&#39;not (pooled)&#39;</span><span class="p">,</span> <span class="n">plot_violins</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>&lt;flotilla.visualize.decomposition.DecompositionViz at 0x118468090&gt;
</pre></div>
</div>
<img alt="../_images/shalek2013_122_1.png" src="../_images/shalek2013_122_1.png" />
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">study</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s">&#39;shalek2013&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Wrote datapackage to /Users/olga/flotilla_projects/shalek2013/datapackage.json
</pre></div>
</div>
<p>Without <tt class="docutils literal"><span class="pre">flotilla</span></tt>, <tt class="docutils literal"><span class="pre">plot_pca</span></tt> is quite a bit of code:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">cycle</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="kn">from</span> <span class="nn">sklearn</span> <span class="kn">import</span> <span class="n">decomposition</span>
<span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">matplotlib.gridspec</span> <span class="kn">import</span> <span class="n">GridSpec</span><span class="p">,</span> <span class="n">GridSpecFromSubplotSpec</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>

<span class="kn">from</span> <span class="nn">flotilla.visualize.color</span> <span class="kn">import</span> <span class="n">dark2</span>
<span class="kn">from</span> <span class="nn">flotilla.visualize.generic</span> <span class="kn">import</span> <span class="n">violinplot</span>


<span class="k">class</span> <span class="nc">DataFrameReducerBase</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Just like scikit-learn&#39;s reducers, but with prettied up DataFrames.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="o">**</span><span class="n">decomposer_kwargs</span><span class="p">):</span>

        <span class="c"># This magically initializes the reducer like DataFramePCA or DataFrameNMF</span>
        <span class="k">if</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s">&quot;Too few features (n={}) to reduce&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="n">n_components</span><span class="p">,</span>
                                                   <span class="o">**</span><span class="n">decomposer_kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reduced_space</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">relabel_pcs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="k">return</span> <span class="s">&quot;pc_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
        <span class="k">except</span> <span class="ne">AssertionError</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Try again as a pandas DataFrame&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;Input X was not a pandas DataFrame, &#39;</span>
                             <span class="s">&#39;was of type {} instead&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">))))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">X</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">components_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">components_</span><span class="p">,</span>
                                        <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relabel_pcs</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_</span><span class="p">)</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relabel_pcs</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">relabel_pcs</span><span class="p">,</span>
                                                            <span class="mi">0</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
        <span class="n">component_space</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
            <span class="n">component_space</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">component_space</span><span class="p">,</span>
                                           <span class="n">index</span><span class="o">=</span><span class="n">X</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">relabel_pcs</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">component_space</span>

    <span class="k">def</span> <span class="nf">fit_transform</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Try again as a pandas DataFrame&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;Input X was not a pandas DataFrame, &#39;</span>
                             <span class="s">&#39;was of type {} instead&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">))))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>


<span class="k">class</span> <span class="nc">DataFramePCA</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="n">decomposition</span><span class="o">.</span><span class="n">PCA</span><span class="p">):</span>
    <span class="k">pass</span>


<span class="k">class</span> <span class="nc">DataFrameNMF</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="n">decomposition</span><span class="o">.</span><span class="n">NMF</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        duplicated fit code for DataFrameNMF because sklearn&#39;s NMF cheats for</span>
<span class="sd">        efficiency and calls fit_transform. MRO resolves the closest</span>
<span class="sd">        (in this package)</span>
<span class="sd">        _single_fit_transform first and so there&#39;s a recursion error:</span>

<span class="sd">            def fit(self, X, y=None, **params):</span>
<span class="sd">                self._single_fit_transform(X, **params)</span>
<span class="sd">                return self</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="o">==</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Try again as a pandas DataFrame&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;Input X was not a pandas DataFrame, &#39;</span>
                             <span class="s">&#39;was of type {} instead&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">X</span><span class="p">))))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">X</span>
        <span class="c"># notice this is fit_transform, not fit</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">decomposition</span><span class="o">.</span><span class="n">NMF</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">components_</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">components_</span><span class="p">,</span>
                                        <span class="n">columns</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">relabel_pcs</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>


<span class="k">class</span> <span class="nc">DataFrameICA</span><span class="p">(</span><span class="n">DataFrameReducerBase</span><span class="p">,</span> <span class="n">decomposition</span><span class="o">.</span><span class="n">FastICA</span><span class="p">):</span>
    <span class="k">pass</span>

<span class="k">class</span> <span class="nc">DecompositionViz</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plots the reduced space from a decomposed dataset. Does not perform any</span>
<span class="sd">    reductions of its own</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reduced_space</span><span class="p">,</span> <span class="n">components_</span><span class="p">,</span>
                 <span class="n">explained_variance_ratio_</span><span class="p">,</span>
                 <span class="n">feature_renamer</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">singles</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">pooled</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">outliers</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">featurewise</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">violinplot_kws</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">label_to_color</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">label_to_marker</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">scale_by_variance</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">x_pc</span><span class="o">=</span><span class="s">&#39;pc_1&#39;</span><span class="p">,</span>
                 <span class="n">y_pc</span><span class="o">=</span><span class="s">&#39;pc_2&#39;</span><span class="p">,</span> <span class="n">n_vectors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">distance</span><span class="o">=</span><span class="s">&#39;L1&#39;</span><span class="p">,</span>
                 <span class="n">n_top_pc_features</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">max_char_width</span><span class="o">=</span><span class="mi">30</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot the results of a decomposition visualization</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        reduced_space : pandas.DataFrame</span>
<span class="sd">            A (n_samples, n_dimensions) DataFrame of the post-dimensionality</span>
<span class="sd">            reduction data</span>
<span class="sd">        components_ : pandas.DataFrame</span>
<span class="sd">            A (n_features, n_dimensions) DataFrame of how much each feature</span>
<span class="sd">            contributes to the components (trailing underscore to be</span>
<span class="sd">            consistent with scikit-learn)</span>
<span class="sd">        explained_variance_ratio_ : pandas.Series</span>
<span class="sd">            A (n_dimensions,) Series of how much variance each component</span>
<span class="sd">            explains. (trailing underscore to be consistent with scikit-learn)</span>
<span class="sd">        feature_renamer : function, optional</span>
<span class="sd">            A function which takes the name of the feature and renames it,</span>
<span class="sd">            e.g. from an ENSEMBL ID to a HUGO known gene symbol. If not</span>
<span class="sd">            provided, the original name is used.</span>
<span class="sd">        groupby : mapping function | dict, optional</span>
<span class="sd">            A mapping of the samples to a label, e.g. sample IDs to</span>
<span class="sd">            phenotype, for the violinplots. If None, all samples are treated</span>
<span class="sd">            the same and are colored the same.</span>
<span class="sd">        singles : pandas.DataFrame, optional</span>
<span class="sd">            For violinplots only. If provided and &#39;plot_violins&#39; is True,</span>
<span class="sd">            will plot the raw (not reduced) measurement values as violin plots.</span>
<span class="sd">        pooled : pandas.DataFrame, optional</span>
<span class="sd">            For violinplots only. If provided, pooled samples are plotted as</span>
<span class="sd">            black dots within their label.</span>
<span class="sd">        outliers : pandas.DataFrame, optional</span>
<span class="sd">            For violinplots only. If provided, outlier samples are plotted as</span>
<span class="sd">            a grey shadow within their label.</span>
<span class="sd">        featurewise : bool, optional</span>
<span class="sd">            If True, then the &quot;samples&quot; are features, e.g. genes instead of</span>
<span class="sd">            samples, and the &quot;features&quot; are the samples, e.g. the cells</span>
<span class="sd">            instead of the gene ids. Essentially, the transpose of the</span>
<span class="sd">            original matrix. If True, then violins aren&#39;t plotted. (default</span>
<span class="sd">            False)</span>
<span class="sd">        order : list-like</span>
<span class="sd">            The order of the labels for the violinplots, e.g. if the data is</span>
<span class="sd">            from a differentiation timecourse, then this would be the labels</span>
<span class="sd">            of the phenotypes, in the differentiation order.</span>
<span class="sd">        violinplot_kws : dict</span>
<span class="sd">            Any additional parameters to violinplot</span>
<span class="sd">        data_type : &#39;expression&#39; | &#39;splicing&#39;, optional</span>
<span class="sd">            For violinplots only. The kind of data that was originally used</span>
<span class="sd">            for the reduction. (default &#39;expression&#39;)</span>
<span class="sd">        label_to_color : dict, optional</span>
<span class="sd">            A mapping of the label, e.g. the phenotype, to the desired</span>
<span class="sd">            plotting color (default None, auto-assigned with the groupby)</span>
<span class="sd">        label_to_marker : dict, optional</span>
<span class="sd">            A mapping of the label, e.g. the phenotype, to the desired</span>
<span class="sd">            plotting symbol (default None, auto-assigned with the groupby)</span>
<span class="sd">        scale_by_variance : bool, optional</span>
<span class="sd">            If True, scale the x- and y-axes by their explained_variance_ratio_</span>
<span class="sd">            (default True)</span>
<span class="sd">        {x,y}_pc : str, optional</span>
<span class="sd">            Principal component to plot on the x- and y-axis. (default &quot;pc_1&quot;</span>
<span class="sd">            and &quot;pc_2&quot;)</span>
<span class="sd">        n_vectors : int, optional</span>
<span class="sd">            Number of vectors to plot of the principal components. (default 20)</span>
<span class="sd">        distance : &#39;L1&#39; | &#39;L2&#39;, optional</span>
<span class="sd">            The distance metric to use to plot the vector lengths. L1 is</span>
<span class="sd">            &quot;Cityblock&quot;, i.e. the sum of the x and y coordinates, and L2 is</span>
<span class="sd">            the traditional Euclidean distance. (default &quot;L1&quot;)</span>
<span class="sd">        n_top_pc_features : int, optional</span>
<span class="sd">            THe number of top features from the principal components to plot.</span>
<span class="sd">            (default 50)</span>
<span class="sd">        max_char_width : int, optional</span>
<span class="sd">            Maximum character width of a feature name. Useful for crazy long</span>
<span class="sd">            feature IDs like MISO IDs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reduced_space</span> <span class="o">=</span> <span class="n">reduced_space</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">components_</span> <span class="o">=</span> <span class="n">components_</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="o">=</span> <span class="n">explained_variance_ratio_</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">singles</span> <span class="o">=</span> <span class="n">singles</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pooled</span> <span class="o">=</span> <span class="n">pooled</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">outliers</span> <span class="o">=</span> <span class="n">outliers</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">groupby</span> <span class="o">=</span> <span class="n">groupby</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">order</span> <span class="o">=</span> <span class="n">order</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">violinplot_kws</span> <span class="o">=</span> <span class="n">violinplot_kws</span> <span class="k">if</span> <span class="n">violinplot_kws</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> \
            <span class="k">else</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data_type</span> <span class="o">=</span> <span class="n">data_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span> <span class="o">=</span> <span class="n">label_to_color</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">label_to_marker</span> <span class="o">=</span> <span class="n">label_to_marker</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_vectors</span> <span class="o">=</span> <span class="n">n_vectors</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span> <span class="o">=</span> <span class="n">x_pc</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span> <span class="o">=</span> <span class="n">y_pc</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">distance</span> <span class="o">=</span> <span class="n">distance</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_top_pc_features</span> <span class="o">=</span> <span class="n">n_top_pc_features</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">featurewise</span> <span class="o">=</span> <span class="n">featurewise</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span> <span class="o">=</span> <span class="n">feature_renamer</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_char_width</span> <span class="o">=</span> <span class="n">max_char_width</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">colors</span> <span class="o">=</span> <span class="n">cycle</span><span class="p">(</span><span class="n">dark2</span><span class="p">)</span>

            <span class="k">def</span> <span class="nf">color_factory</span><span class="p">():</span>
                <span class="k">return</span> <span class="n">colors</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">color_factory</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">label_to_marker</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">markers</span> <span class="o">=</span> <span class="n">cycle</span><span class="p">([</span><span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="s">&#39;^&#39;</span><span class="p">,</span> <span class="s">&#39;s&#39;</span><span class="p">,</span> <span class="s">&#39;v&#39;</span><span class="p">,</span> <span class="s">&#39;*&#39;</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">,</span> <span class="s">&#39;h&#39;</span><span class="p">])</span>

            <span class="k">def</span> <span class="nf">marker_factory</span><span class="p">():</span>
                <span class="k">return</span> <span class="n">markers</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">label_to_marker</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">marker_factory</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">groupby</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">groupby</span> <span class="o">=</span> <span class="nb">dict</span><span class="o">.</span><span class="n">fromkeys</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reduced_space</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s">&#39;all&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">grouped</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reduced_space</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">groupby</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">order</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">color_ordered</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">order</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">color_ordered</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                  <span class="bp">self</span><span class="o">.</span><span class="n">grouped</span><span class="o">.</span><span class="n">groups</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">ix</span><span class="p">[[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">]]</span>

        <span class="c"># Get the explained variance</span>
        <span class="k">if</span> <span class="n">explained_variance_ratio_</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">vars</span> <span class="o">=</span> <span class="n">explained_variance_ratio_</span><span class="p">[[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">]]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">vars</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="mf">1.</span><span class="p">,</span> <span class="mf">1.</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">scale_by_variance</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span><span class="o">.</span><span class="n">multiply</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vars</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c"># sort features by magnitude/contribution to transformation</span>
        <span class="n">reduced_space</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reduced_space</span><span class="p">[[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">]]</span>
        <span class="n">farthest_sample</span> <span class="o">=</span> <span class="n">reduced_space</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
        <span class="n">whole_space</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">)</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
        <span class="n">scale</span> <span class="o">=</span> <span class="o">.</span><span class="mi">25</span> <span class="o">*</span> <span class="n">farthest_sample</span> <span class="o">/</span> <span class="n">whole_space</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span> <span class="o">*=</span> <span class="n">scale</span>

        <span class="nb">ord</span> <span class="o">=</span> <span class="mi">2</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">distance</span> <span class="o">==</span> <span class="s">&#39;L2&#39;</span> <span class="k">else</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">magnitudes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="nb">ord</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">magnitudes</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">top_features</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings_labels</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">pc</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">pcs</span><span class="p">:</span>
            <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">x</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
            <span class="n">half_features</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">n_top_pc_features</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_top_pc_features</span><span class="p">:</span>
                <span class="n">a</span> <span class="o">=</span> <span class="n">x</span><span class="p">[:</span><span class="n">half_features</span><span class="p">]</span>
                <span class="n">b</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="o">-</span><span class="n">half_features</span><span class="p">:]</span>
                <span class="n">labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">b</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">r_</span><span class="p">[</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">labels</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">index</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span> <span class="o">=</span> <span class="n">x</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings_labels</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span> <span class="o">=</span> <span class="n">labels</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">top_features</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">,</span> <span class="n">plot_violins</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                 <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">show_vectors</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                 <span class="n">show_vector_labels</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                 <span class="n">markersize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="n">gs_x</span> <span class="o">=</span> <span class="mi">14</span>
        <span class="n">gs_y</span> <span class="o">=</span> <span class="mi">12</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reduced_fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">GridSpec</span><span class="p">(</span><span class="n">gs_x</span><span class="p">,</span> <span class="n">gs_y</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">GridSpecFromSubplotSpec</span><span class="p">(</span><span class="n">gs_x</span><span class="p">,</span> <span class="n">gs_y</span><span class="p">,</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_subplotspec</span><span class="p">())</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reduced_fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gcf</span><span class="p">()</span>

        <span class="n">ax_components</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">5</span><span class="p">])</span>
        <span class="n">ax_loading1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[:,</span> <span class="mi">6</span><span class="p">:</span><span class="mi">8</span><span class="p">])</span>
        <span class="n">ax_loading2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[:,</span> <span class="mi">10</span><span class="p">:</span><span class="mi">14</span><span class="p">])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">plot_samples</span><span class="p">(</span><span class="n">show_point_labels</span><span class="o">=</span><span class="n">show_point_labels</span><span class="p">,</span>
                          <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">show_vectors</span><span class="o">=</span><span class="n">show_vectors</span><span class="p">,</span>
                          <span class="n">show_vector_labels</span><span class="o">=</span><span class="n">show_vector_labels</span><span class="p">,</span>
                          <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="n">legend</span><span class="p">,</span>
                          <span class="n">ax</span><span class="o">=</span><span class="n">ax_components</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_loadings</span><span class="p">(</span><span class="n">pc</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_loading1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_loadings</span><span class="p">(</span><span class="n">pc</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_loading2</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reduced_fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">plot_violins</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">featurewise</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">singles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_violins</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">shorten</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_char_width</span><span class="p">:</span>
            <span class="k">return</span> <span class="s">&#39;{}...&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">[:</span><span class="bp">self</span><span class="o">.</span><span class="n">max_char_width</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">x</span>

    <span class="k">def</span> <span class="nf">plot_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                     <span class="n">title</span><span class="o">=</span><span class="s">&#39;DataFramePCA&#39;</span><span class="p">,</span> <span class="n">show_vectors</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                     <span class="n">show_vector_labels</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                     <span class="n">three_d</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a pandas dataframe, performs DataFramePCA and plots the results in a</span>
<span class="sd">        convenient single function.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        groupby : groupby</span>
<span class="sd">            How to group the samples by color/label</span>
<span class="sd">        label_to_color : dict</span>
<span class="sd">            Group labels to a matplotlib color E.g. if you&#39;ve already chosen</span>
<span class="sd">            specific colors to indicate a particular group. Otherwise will</span>
<span class="sd">            auto-assign colors</span>
<span class="sd">        label_to_marker : dict</span>
<span class="sd">            Group labels to matplotlib marker</span>
<span class="sd">        title : str</span>
<span class="sd">            title of the plot</span>
<span class="sd">        show_vectors : bool</span>
<span class="sd">            Whether or not to draw the vectors indicating the supporting</span>
<span class="sd">            principal components</span>
<span class="sd">        show_vector_labels : bool</span>
<span class="sd">            whether or not to draw the names of the vectors</span>
<span class="sd">        show_point_labels : bool</span>
<span class="sd">            Whether or not to label the scatter points</span>
<span class="sd">        markersize : int</span>
<span class="sd">            size of the scatter markers on the plot</span>
<span class="sd">        text_group : list of str</span>
<span class="sd">            Group names that you want labeled with text</span>
<span class="sd">        three_d : bool</span>
<span class="sd">            if you want hte plot in 3d (need to set up the axes beforehand)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        For each vector in data:</span>
<span class="sd">        x, y, marker, distance</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>

        <span class="c"># Plot the samples</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">grouped</span><span class="p">:</span>
            <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">label_to_color</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">marker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">label_to_marker</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">]</span>
            <span class="n">y</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s">&#39;None&#39;</span><span class="p">,</span>
                    <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span>
                    <span class="n">markeredgewidth</span><span class="o">=.</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                    <span class="n">pooled_ids</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pooled</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
                    <span class="n">pooled_x</span><span class="p">,</span> <span class="n">pooled_y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">pooled_ids</span><span class="p">],</span> <span class="n">y</span><span class="p">[</span><span class="n">pooled_ids</span><span class="p">]</span>
                    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pooled_x</span><span class="p">,</span> <span class="n">pooled_y</span><span class="p">,</span> <span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span>
                            <span class="n">markeredgecolor</span><span class="o">=</span><span class="s">&#39;k&#39;</span><span class="p">,</span> <span class="n">markeredgewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                            <span class="n">label</span><span class="o">=</span><span class="s">&#39;{} pooled&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">),</span>
                            <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">outliers</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                    <span class="n">outlier_ids</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">outliers</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
                    <span class="n">outlier_x</span><span class="p">,</span> <span class="n">outlier_y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="n">outlier_ids</span><span class="p">],</span> <span class="n">y</span><span class="p">[</span><span class="n">outlier_ids</span><span class="p">]</span>
                    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">outlier_x</span><span class="p">,</span> <span class="n">outlier_y</span><span class="p">,</span> <span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
                            <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span>
                            <span class="n">markeredgecolor</span><span class="o">=</span><span class="s">&#39;lightgrey&#39;</span><span class="p">,</span> <span class="n">markeredgewidth</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                            <span class="n">label</span><span class="o">=</span><span class="s">&#39;{} outlier&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">),</span>
                            <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">if</span> <span class="n">show_point_labels</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">args</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
                    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span>

        <span class="c"># Plot vectors, if asked</span>
        <span class="k">if</span> <span class="n">show_vectors</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">vector_label</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">magnitudes</span><span class="p">[:</span><span class="bp">self</span><span class="o">.</span><span class="n">n_vectors</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
                <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">loadings</span><span class="p">[</span><span class="n">vector_label</span><span class="p">]</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">x</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">y</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;k&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">show_vector_labels</span><span class="p">:</span>
                    <span class="n">x_offset</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">copysign</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
                    <span class="n">y_offset</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">copysign</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
                    <span class="n">horizontalalignment</span> <span class="o">=</span> <span class="s">&#39;left&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s">&#39;right&#39;</span>
                    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="n">renamed</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span><span class="p">(</span><span class="n">vector_label</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">renamed</span> <span class="o">=</span> <span class="n">vector_label</span>
                    <span class="n">ax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">renamed</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">),</span>
                                <span class="n">textcoords</span><span class="o">=</span><span class="s">&#39;offset points&#39;</span><span class="p">,</span>
                                <span class="n">xytext</span><span class="o">=</span><span class="p">(</span><span class="n">x_offset</span><span class="p">,</span> <span class="n">y_offset</span><span class="p">),</span>
                                <span class="n">horizontalalignment</span><span class="o">=</span><span class="n">horizontalalignment</span><span class="p">)</span>

        <span class="c"># Label x and y axes</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span>
            <span class="s">&#39;Principal Component {} (Explains {:.2f}% Of Variance)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">),</span> <span class="mi">100</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">vars</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">x_pc</span><span class="p">]))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span>
            <span class="s">&#39;Principal Component {} (Explains {:.2f}% Of Variance)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">),</span> <span class="mi">100</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">vars</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">y_pc</span><span class="p">]))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">legend</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">plot_loadings</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pc</span><span class="o">=</span><span class="s">&#39;pc_1&#39;</span><span class="p">,</span> <span class="n">n_features</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="n">loadings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pc_loadings_labels</span><span class="p">[</span><span class="n">pc</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">loadings</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">loadings</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="s">&#39;o&#39;</span><span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">loadings</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">n_features</span><span class="p">)))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s">&quot;Component &quot;</span> <span class="o">+</span> <span class="n">pc</span><span class="p">)</span>

        <span class="n">x_offset</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">loadings</span><span class="p">)</span> <span class="o">*</span> <span class="o">.</span><span class="mo">05</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="n">loadings</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">-</span> <span class="n">x_offset</span><span class="p">,</span>
                    <span class="n">right</span><span class="o">=</span><span class="n">loadings</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">+</span> <span class="n">x_offset</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="n">labels</span>

        <span class="n">labels</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">shorten</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
        <span class="c"># ax.set_yticklabels(map(shorten, labels))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">lab</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">():</span>
            <span class="n">lab</span><span class="o">.</span><span class="n">set_rotation</span><span class="p">(</span><span class="mi">90</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">plot_explained_variance</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">&quot;PCA explained variance&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;If the reducer is a form of PCA, then plot the explained variance</span>
<span class="sd">        ratio by the components.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># Plot the explained variance ratio</span>
        <span class="k">assert</span> <span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span>
        <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
        <span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">,</span> <span class="s">&#39;o-&#39;</span><span class="p">)</span>

        <span class="n">xticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">xticks</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xticks</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;Principal component&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Fraction explained variance&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">plot_violins</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Make violinplots of each feature</span>

<span class="sd">        Must be called after plot_samples because it depends on the existence</span>
<span class="sd">        of the &quot;self.magnitudes&quot; attribute.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ncols</span> <span class="o">=</span> <span class="mi">4</span>
        <span class="n">nrows</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">vector_labels</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">magnitudes</span><span class="p">[:</span><span class="bp">self</span><span class="o">.</span><span class="n">n_vectors</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">Index</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">top_features</span><span class="p">))))</span>
        <span class="k">while</span> <span class="n">ncols</span> <span class="o">*</span> <span class="n">nrows</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">vector_labels</span><span class="p">):</span>
            <span class="n">nrows</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">violins_fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="n">ncols</span><span class="p">,</span>
                                              <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span> <span class="o">*</span> <span class="n">ncols</span><span class="p">,</span> <span class="mi">4</span> <span class="o">*</span> <span class="n">nrows</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">renamed_vectors</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">feature_renamer</span><span class="p">,</span> <span class="n">vector_labels</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">renamed_vectors</span> <span class="o">=</span> <span class="n">vector_labels</span>
        <span class="n">labels</span> <span class="o">=</span> <span class="p">[(</span><span class="n">y</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">renamed_vectors</span><span class="p">,</span>
                                                  <span class="n">vector_labels</span><span class="p">))]</span>

        <span class="k">for</span> <span class="p">(</span><span class="n">renamed</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">),</span> <span class="n">ax</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">):</span>
            <span class="n">singles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">singles</span><span class="p">[</span><span class="n">feature_id</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">singles</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> \
                <span class="k">else</span> <span class="bp">None</span>
            <span class="n">pooled</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pooled</span><span class="p">[</span><span class="n">feature_id</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pooled</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="k">else</span> \
                <span class="bp">None</span>
            <span class="n">outliers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">outliers</span><span class="p">[</span><span class="n">feature_id</span><span class="p">]</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">outliers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> \
                <span class="k">else</span> <span class="bp">None</span>
            <span class="n">title</span> <span class="o">=</span> <span class="s">&#39;{}</span><span class="se">\n</span><span class="s">{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">feature_id</span><span class="p">,</span> <span class="n">renamed</span><span class="p">)</span>
            <span class="n">violinplot</span><span class="p">(</span><span class="n">singles</span><span class="p">,</span> <span class="n">pooled_data</span><span class="o">=</span><span class="n">pooled</span><span class="p">,</span> <span class="n">outliers</span><span class="o">=</span><span class="n">outliers</span><span class="p">,</span>
                       <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">groupby</span><span class="p">,</span> <span class="n">color_ordered</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">color_ordered</span><span class="p">,</span>
                       <span class="n">order</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">order</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
                       <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data_type</span><span class="p">,</span>
                       <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">violinplot_kws</span><span class="p">)</span>

        <span class="c"># Clear any unused axes</span>
        <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">:</span>
            <span class="c"># Check if the plotting space is empty</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">collections</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">lines</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s">&#39;off&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">violins_fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="c"># Notice we&#39;re using the original data, nothing from &quot;study&quot;</span>
<span class="n">lps_response_genes</span> <span class="o">=</span> <span class="n">expression_feature_data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">expression_feature_data</span><span class="o">.</span><span class="n">gene_category</span> <span class="o">==</span> <span class="s">&#39;LPS Response&#39;</span><span class="p">]</span>
<span class="n">subset</span> <span class="o">=</span> <span class="n">expression_filtered</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">singles_ids</span><span class="p">,</span> <span class="n">lps_response_genes</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">subset_standardized</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">StandardScaler</span><span class="p">()</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">subset</span><span class="p">),</span>
                                       <span class="n">index</span><span class="o">=</span><span class="n">subset</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">subset</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>


<span class="n">pca</span> <span class="o">=</span> <span class="n">DataFramePCA</span><span class="p">(</span><span class="n">subset_standardized</span><span class="p">)</span>
<span class="n">visualizer</span> <span class="o">=</span> <span class="n">DecompositionViz</span><span class="p">(</span><span class="n">pca</span><span class="o">.</span><span class="n">reduced_space</span><span class="p">,</span> <span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="p">,</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span><span class="p">)</span>
<span class="n">visualizer</span><span class="p">();</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_125_0.png" src="../_images/shalek2013_125_0.png" />
</div>
<div class="section" id="figure-4b">
<h3>Figure 4b<a class="headerlink" href="#figure-4b" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">lps_response_genes</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_subsets</span><span class="p">[</span><span class="s">&#39;gene_category: LPS Response&#39;</span><span class="p">]</span>
<span class="n">lps_response</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">lps_response_genes</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">lps_response</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>GENE</th>
      <th>1110018G07RIK</th>
      <th>1110038F14RIK</th>
      <th>1200009I06RIK</th>
      <th>1600014C10RIK</th>
      <th>1810029B16RIK</th>
      <th>2210009G21RIK</th>
      <th>2810474O19RIK</th>
      <th>3110001I22RIK</th>
      <th>4921513D23RIK</th>
      <th>4930523C07RIK</th>
      <th>...</th>
      <th>ZC3H12C</th>
      <th>ZC3HAV1</th>
      <th>ZCCHC2</th>
      <th>ZCCHC6</th>
      <th>ZDHHC21</th>
      <th>ZFP36</th>
      <th>ZFP800</th>
      <th>ZHX2</th>
      <th>ZNFX1</th>
      <th>ZUFSP</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>S1</th>
      <td> 3.711442</td>
      <td> 0.000000</td>
      <td> 3.275468</td>
      <td> 0.000000</td>
      <td> 5.609305</td>
      <td> 0</td>
      <td> 0.000000</td>
      <td> 3.828860</td>
      <td> 1.314573</td>
      <td> 3.778275</td>
      <td>...</td>
      <td> 3.972904</td>
      <td> 3.509979</td>
      <td> 0.035344</td>
      <td> 3.042277</td>
      <td> 4.425735</td>
      <td> 4.092559</td>
      <td> 4.025124</td>
      <td> 0.779382</td>
      <td> 2.998800</td>
      <td> 0.000000</td>
    </tr>
    <tr>
      <th>S2</th>
      <td> 4.361671</td>
      <td> 0.147643</td>
      <td> 0.000000</td>
      <td> 0.000000</td>
      <td> 5.478071</td>
      <td> 0</td>
      <td> 3.407342</td>
      <td> 0.000000</td>
      <td> 1.531443</td>
      <td> 0.000000</td>
      <td>...</td>
      <td> 4.794306</td>
      <td> 4.984262</td>
      <td> 2.251330</td>
      <td> 1.018315</td>
      <td> 4.955713</td>
      <td> 0.356008</td>
      <td> 4.297776</td>
      <td> 0.032569</td>
      <td> 3.091207</td>
      <td> 5.000843</td>
    </tr>
    <tr>
      <th>S3</th>
      <td> 0.000000</td>
      <td> 3.737014</td>
      <td> 2.987093</td>
      <td> 0.063526</td>
      <td> 5.320993</td>
      <td> 0</td>
      <td> 3.372359</td>
      <td> 0.058163</td>
      <td> 1.105115</td>
      <td> 0.025043</td>
      <td>...</td>
      <td> 4.882749</td>
      <td> 0.807258</td>
      <td> 0.094925</td>
      <td> 0.126673</td>
      <td> 3.952273</td>
      <td> 1.956983</td>
      <td> 0.000000</td>
      <td> 0.000000</td>
      <td> 3.794063</td>
      <td> 2.928699</td>
    </tr>
    <tr>
      <th>S4</th>
      <td> 2.719587</td>
      <td> 0.000000</td>
      <td> 0.045823</td>
      <td> 0.000000</td>
      <td> 0.488049</td>
      <td> 0</td>
      <td> 5.127847</td>
      <td> 0.000000</td>
      <td> 2.303969</td>
      <td> 0.000000</td>
      <td>...</td>
      <td> 4.833354</td>
      <td> 4.538699</td>
      <td> 0.137427</td>
      <td> 2.025546</td>
      <td> 4.193989</td>
      <td> 2.372572</td>
      <td> 0.121924</td>
      <td> 0.000000</td>
      <td> 0.230278</td>
      <td> 0.430168</td>
    </tr>
    <tr>
      <th>S5</th>
      <td> 2.982073</td>
      <td> 0.000000</td>
      <td> 2.829152</td>
      <td> 0.000000</td>
      <td> 5.093188</td>
      <td> 0</td>
      <td> 0.065122</td>
      <td> 4.635671</td>
      <td> 1.015640</td>
      <td> 0.461296</td>
      <td>...</td>
      <td> 4.446634</td>
      <td> 0.157178</td>
      <td> 0.616401</td>
      <td> 0.000000</td>
      <td> 4.039816</td>
      <td> 0.000000</td>
      <td> 4.714087</td>
      <td> 1.565475</td>
      <td> 0.860254</td>
      <td> 4.866979</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 630 columns</p>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">lps_response_corr</span> <span class="o">=</span> <span class="n">lps_response</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
</pre></div>
</div>
<div class="section" id="elbow-method-for-determining-number-of-clusters">
<h4>&#8220;Elbow method&#8221; for determining number of clusters<a class="headerlink" href="#elbow-method-for-determining-number-of-clusters" title="Permalink to this headline">¶</a></h4>
<p>The authors state that they used the &#8220;Elbow method&#8221; to determine the
<a class="reference external" href="http://en.wikipedia.org/wiki/Determining_the_number_of_clusters_in_a_data_set">number of cluster
centers</a>.
Essentially, you try a bunch of different <span class="math">\(k\)</span>, and see where
there is a flattening out of the metric, like an elbow. There&#8217;s a few
different variations on which metric to use, such as using the average
distance to the cluster center, or the explained variance. Let&#8217;s try the
distance to cluster center first, because <tt class="docutils literal"><span class="pre">scikit-learn</span></tt> makes it
easy.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">KMeans</span>

<span class="c">##### cluster data into K=1..10 clusters #####</span>
<span class="n">ks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

<span class="n">X</span> <span class="o">=</span> <span class="n">lps_response_corr</span><span class="o">.</span><span class="n">values</span>

<span class="n">kmeans</span> <span class="o">=</span> <span class="p">[</span><span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="n">k</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ks</span><span class="p">]</span>

<span class="c"># Scikit-learn makes this easy by computing the distance to the nearest center</span>
<span class="n">dist_to_center</span> <span class="o">=</span> <span class="p">[</span><span class="n">km</span><span class="o">.</span><span class="n">inertia_</span> <span class="k">for</span> <span class="n">km</span> <span class="ow">in</span> <span class="n">kmeans</span><span class="p">]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ks</span><span class="p">,</span> <span class="n">dist_to_center</span><span class="p">,</span> <span class="s">&#39;o-&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Sum of distance to nearest cluster center&#39;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_131_0.png" src="../_images/shalek2013_131_0.png" />
<p>Not quite sure where the elbow is here. looks like there&#8217;s a big drop
off after <span class="math">\(k=1\)</span>, but that could just be an illusion. Since they
didn&#8217;t specify which version of the elbow method they used, I&#8217;m not
going to investigate this further, and just see if we can see what they
see with the <span class="math">\(k=5\)</span> clusters that they found was optimal.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">kmeans</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">lps_response_corr_clusters</span> <span class="o">=</span> <span class="n">kmeans</span><span class="o">.</span><span class="n">fit_predict</span><span class="p">(</span><span class="n">lps_response_corr</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="n">lps_response_corr_clusters</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span>
       <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span>
       <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span>
       <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>
       <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span>
       <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>
       <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span>
       <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span>
       <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">int32</span><span class="p">)</span>
</pre></div>
</div>
<p>Now let&#8217;s create a dataframe with these genes in their cluster orders.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_to_cluster</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">lps_response_corr</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">lps_response_corr_clusters</span><span class="p">))</span>

<span class="n">dfs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">lps_response_corr</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">gene_to_cluster</span><span class="p">):</span>
    <span class="n">dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="n">lps_response_corr_ordered_by_clusters</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

<span class="c"># Make symmetric, since we created this dataframe by smashing rows on top of each other, we need to reorder the columns</span>
<span class="n">lps_response_corr_ordered_by_clusters</span> <span class="o">=</span> <span class="n">lps_response_corr_ordered_by_clusters</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="n">lps_response_corr_ordered_by_clusters</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
<span class="n">lps_response_corr_ordered_by_clusters</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>GENE</th>
      <th>1110038F14RIK</th>
      <th>2210009G21RIK</th>
      <th>A430084P05RIK</th>
      <th>AA960436</th>
      <th>AK141659</th>
      <th>AK163103</th>
      <th>ALCAM</th>
      <th>ALPK2</th>
      <th>ARMC8</th>
      <th>BC147527</th>
      <th>...</th>
      <th>TNFAIP2</th>
      <th>TNFSF4</th>
      <th>TOR1AIP1</th>
      <th>TRA2A</th>
      <th>TRIM26</th>
      <th>TRIM34</th>
      <th>TTC39C</th>
      <th>USP12</th>
      <th>ZC3H12C</th>
      <th>ZC3HAV1</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1110038F14RIK</th>
      <td> 1.000000</td>
      <td> 0.175230</td>
      <td> 0.043846</td>
      <td> 0.240304</td>
      <td> 0.150073</td>
      <td>-0.007459</td>
      <td>-0.075510</td>
      <td> 0.001210</td>
      <td> 0.078638</td>
      <td>-0.073983</td>
      <td>...</td>
      <td>-0.053588</td>
      <td>-0.085191</td>
      <td> 0.322774</td>
      <td> 0.096905</td>
      <td>-0.370932</td>
      <td> 0.451829</td>
      <td> 0.387727</td>
      <td>-0.257008</td>
      <td> 0.274102</td>
      <td>-0.163423</td>
    </tr>
    <tr>
      <th>2210009G21RIK</th>
      <td> 0.175230</td>
      <td> 1.000000</td>
      <td> 0.301786</td>
      <td> 0.454579</td>
      <td>-0.106546</td>
      <td>-0.122179</td>
      <td> 0.177472</td>
      <td> 0.215454</td>
      <td> 0.540303</td>
      <td> 0.078574</td>
      <td>...</td>
      <td> 0.230309</td>
      <td>-0.158622</td>
      <td> 0.019694</td>
      <td> 0.142045</td>
      <td> 0.053967</td>
      <td> 0.483106</td>
      <td>-0.085604</td>
      <td> 0.279262</td>
      <td> 0.153934</td>
      <td> 0.160710</td>
    </tr>
    <tr>
      <th>A430084P05RIK</th>
      <td> 0.043846</td>
      <td> 0.301786</td>
      <td> 1.000000</td>
      <td> 0.001150</td>
      <td> 0.060210</td>
      <td>-0.173020</td>
      <td> 0.150884</td>
      <td> 0.429134</td>
      <td> 0.131837</td>
      <td>-0.069652</td>
      <td>...</td>
      <td>-0.341757</td>
      <td>-0.296639</td>
      <td>-0.192074</td>
      <td>-0.360383</td>
      <td> 0.025340</td>
      <td> 0.033636</td>
      <td>-0.227960</td>
      <td>-0.166541</td>
      <td> 0.200579</td>
      <td> 0.064736</td>
    </tr>
    <tr>
      <th>AA960436</th>
      <td> 0.240304</td>
      <td> 0.454579</td>
      <td> 0.001150</td>
      <td> 1.000000</td>
      <td>-0.361780</td>
      <td> 0.206889</td>
      <td> 0.174208</td>
      <td> 0.075687</td>
      <td> 0.394432</td>
      <td> 0.163830</td>
      <td>...</td>
      <td> 0.175022</td>
      <td>-0.271395</td>
      <td> 0.272221</td>
      <td>-0.222182</td>
      <td> 0.181522</td>
      <td>-0.094028</td>
      <td> 0.218182</td>
      <td> 0.396040</td>
      <td>-0.159072</td>
      <td> 0.048122</td>
    </tr>
    <tr>
      <th>AK141659</th>
      <td> 0.150073</td>
      <td>-0.106546</td>
      <td> 0.060210</td>
      <td>-0.361780</td>
      <td> 1.000000</td>
      <td>-0.287830</td>
      <td>-0.370827</td>
      <td> 0.143026</td>
      <td>-0.019682</td>
      <td>-0.157671</td>
      <td>...</td>
      <td>-0.295535</td>
      <td> 0.194073</td>
      <td>-0.232992</td>
      <td> 0.061276</td>
      <td>-0.032583</td>
      <td> 0.411637</td>
      <td>-0.182131</td>
      <td> 0.018036</td>
      <td> 0.175434</td>
      <td>-0.288042</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 630 columns</p>
</div><p>The next step is to get the principal-component reduced data, using only
the LPS response genes. We can do this in <tt class="docutils literal"><span class="pre">flotilla</span></tt> using
<tt class="docutils literal"><span class="pre">study.expression.reduce</span></tt>.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reduced</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">reduce</span><span class="p">(</span><span class="n">singles_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">lps_response_genes</span><span class="p">)</span>
</pre></div>
</div>
<p>We can get the principal components using <tt class="docutils literal"><span class="pre">reduced.components_</span></tt>
(similar interface as <tt class="docutils literal"><span class="pre">scikit-learn</span></tt>).</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">reduced</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>MOV10</th>
      <th>PPAP2B</th>
      <th>LASS6</th>
      <th>TMCO3</th>
      <th>CPD</th>
      <th>AK138792</th>
      <th>TARM1</th>
      <th>P4HA1</th>
      <th>CD180</th>
      <th>SMG7</th>
      <th>...</th>
      <th>OAS1B</th>
      <th>OAS1G</th>
      <th>AK151815</th>
      <th>GTPBP2</th>
      <th>PRPF38A</th>
      <th>SLC7A11</th>
      <th>PCDH7</th>
      <th>GNA13</th>
      <th>PTPRJ</th>
      <th>ATF3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>pc_1</th>
      <td> 0.035299</td>
      <td> 0.038725</td>
      <td>-0.006343</td>
      <td> 0.014219</td>
      <td> 0.033734</td>
      <td>-0.079831</td>
      <td> 0.032886</td>
      <td> 0.034783</td>
      <td> 0.033719</td>
      <td>-0.048453</td>
      <td>...</td>
      <td>-0.022490</td>
      <td> 0.031091</td>
      <td>-0.021397</td>
      <td> 0.034917</td>
      <td> 0.001745</td>
      <td> 0.058000</td>
      <td> 0.007748</td>
      <td> 0.000767</td>
      <td> 0.016012</td>
      <td> 0.018020</td>
    </tr>
    <tr>
      <th>pc_2</th>
      <td> 0.055310</td>
      <td> 0.002925</td>
      <td>-0.043986</td>
      <td>-0.024020</td>
      <td>-0.061957</td>
      <td>-0.016327</td>
      <td> 0.002882</td>
      <td>-0.003178</td>
      <td> 0.050055</td>
      <td> 0.038601</td>
      <td>...</td>
      <td> 0.012240</td>
      <td> 0.052127</td>
      <td> 0.009120</td>
      <td> 0.077015</td>
      <td> 0.072064</td>
      <td>-0.080902</td>
      <td>-0.056607</td>
      <td> 0.068444</td>
      <td>-0.072533</td>
      <td> 0.068088</td>
    </tr>
    <tr>
      <th>pc_3</th>
      <td> 0.000374</td>
      <td> 0.099514</td>
      <td>-0.039636</td>
      <td> 0.003997</td>
      <td>-0.000575</td>
      <td>-0.042212</td>
      <td>-0.056827</td>
      <td> 0.015571</td>
      <td>-0.039811</td>
      <td> 0.005398</td>
      <td>...</td>
      <td>-0.010524</td>
      <td>-0.009277</td>
      <td>-0.102462</td>
      <td>-0.043913</td>
      <td>-0.052513</td>
      <td>-0.030622</td>
      <td> 0.022607</td>
      <td>-0.002503</td>
      <td> 0.023997</td>
      <td>-0.054205</td>
    </tr>
    <tr>
      <th>pc_4</th>
      <td> 0.022491</td>
      <td> 0.002342</td>
      <td> 0.009422</td>
      <td>-0.034725</td>
      <td> 0.025866</td>
      <td>-0.009656</td>
      <td>-0.027689</td>
      <td>-0.089803</td>
      <td>-0.046888</td>
      <td> 0.002274</td>
      <td>...</td>
      <td>-0.003404</td>
      <td>-0.070307</td>
      <td>-0.007025</td>
      <td> 0.003407</td>
      <td>-0.048078</td>
      <td> 0.028099</td>
      <td> 0.032970</td>
      <td>-0.066284</td>
      <td> 0.010371</td>
      <td>-0.006108</td>
    </tr>
    <tr>
      <th>pc_5</th>
      <td>-0.025743</td>
      <td>-0.009200</td>
      <td>-0.030187</td>
      <td>-0.061283</td>
      <td> 0.010464</td>
      <td> 0.032668</td>
      <td> 0.012223</td>
      <td>-0.047623</td>
      <td>-0.047351</td>
      <td> 0.045909</td>
      <td>...</td>
      <td>-0.074817</td>
      <td> 0.044218</td>
      <td>-0.000884</td>
      <td>-0.000597</td>
      <td>-0.033893</td>
      <td>-0.018108</td>
      <td>-0.012669</td>
      <td>-0.025833</td>
      <td>-0.044248</td>
      <td>-0.001995</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 630 columns</p>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">pc_components</span> <span class="o">=</span> <span class="n">reduced</span><span class="o">.</span><span class="n">components_</span><span class="o">.</span><span class="n">ix</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span> <span class="n">lps_response_corr_ordered_by_clusters</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>
<span class="n">pc_components</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>pc_1</th>
      <th>pc_2</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1110038F14RIK</th>
      <td>-0.007729</td>
      <td>-0.005858</td>
    </tr>
    <tr>
      <th>2210009G21RIK</th>
      <td>-0.006981</td>
      <td> 0.002916</td>
    </tr>
    <tr>
      <th>A430084P05RIK</th>
      <td> 0.014550</td>
      <td> 0.022191</td>
    </tr>
    <tr>
      <th>AA960436</th>
      <td> 0.002159</td>
      <td> 0.014470</td>
    </tr>
    <tr>
      <th>AK141659</th>
      <td> 0.016902</td>
      <td>-0.009151</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">matplotlib</span> <span class="kn">as</span> <span class="nn">mpl</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">2</span><span class="p">],</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">1</span><span class="p">])</span>
<span class="n">corr_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">corr_cbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">pc_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])</span>
<span class="n">pc_cbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="mi">1</span><span class="p">:])</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">lps_response_corr_ordered_by_clusters</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">corr_ax</span><span class="p">,</span> <span class="n">cbar_ax</span><span class="o">=</span><span class="n">corr_cbar_ax</span><span class="p">,</span>
            <span class="n">cbar_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s">&#39;horizontal&#39;</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">pc_components</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">PRGn</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">pc_ax</span><span class="p">,</span> <span class="n">cbar_ax</span><span class="o">=</span><span class="n">pc_cbar_ax</span><span class="p">,</span>
            <span class="n">cbar_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s">&#39;horizontal&#39;</span><span class="p">))</span>

<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">pc_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">pc_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>&lt;matplotlib.text.Text at 0x10f37d5d0&gt;
</pre></div>
</div>
<img alt="../_images/shalek2013_141_1.png" src="../_images/shalek2013_141_1.png" />
<p>This looks pretty similar, maybe just rearranged cluster order. Let&#8217;s
check what their data looks like when you plot this.</p>
</div>
<div class="section" id="their-pc-scores-and-clusters-for-the-genes">
<h4>Their PC scores and clusters for the genes<a class="headerlink" href="#their-pc-scores-and-clusters-for-the-genes" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_pc_clusters</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table5.xls&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Annotation</th>
      <th>Cluster</th>
      <th>PC1 Score</th>
      <th>PC2 Score</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>LNPEP</th>
      <td>    NaN</td>
      <td> 1</td>
      <td> 0.232368</td>
      <td> 0.677266</td>
    </tr>
    <tr>
      <th>TOR1AIP2</th>
      <td>  Antiv</td>
      <td> 1</td>
      <td>-0.075934</td>
      <td> 1.485877</td>
    </tr>
    <tr>
      <th>TNFSF4</th>
      <td>    NaN</td>
      <td> 1</td>
      <td> 0.497893</td>
      <td>-0.562412</td>
    </tr>
    <tr>
      <th>CFB</th>
      <td> Inflam</td>
      <td> 1</td>
      <td>-0.394318</td>
      <td> 1.277749</td>
    </tr>
    <tr>
      <th>H2-T10</th>
      <td>    NaN</td>
      <td> 1</td>
      <td> 0.514947</td>
      <td>-0.698538</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">data</span> <span class="o">=</span> <span class="n">lps_response_corr</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">2</span><span class="p">],</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">1</span><span class="p">])</span>
<span class="n">corr_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">corr_cbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="n">pc_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">:])</span>
<span class="n">pc_cbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="mi">1</span><span class="p">:])</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">square</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">corr_ax</span><span class="p">,</span> <span class="n">cbar_ax</span><span class="o">=</span><span class="n">corr_cbar_ax</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s">&#39;horizontal&#39;</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">ix</span><span class="p">[:,</span> <span class="p">[</span><span class="s">&#39;PC1 Score&#39;</span><span class="p">,</span> <span class="s">&#39;PC2 Score&#39;</span><span class="p">]],</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">mpl</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">PRGn</span><span class="p">,</span>
            <span class="n">ax</span><span class="o">=</span><span class="n">pc_ax</span><span class="p">,</span> <span class="n">cbar_ax</span><span class="o">=</span><span class="n">pc_cbar_ax</span><span class="p">,</span> <span class="n">cbar_kws</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s">&#39;horizontal&#39;</span><span class="p">),</span> <span class="n">xticklabels</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">corr_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

<span class="n">pc_ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">pc_ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="p">);</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_145_0.png" src="../_images/shalek2013_145_0.png" />
<p>Sure enough, if I use their annotations, I get exactly that. Though
there were two genes in their file that I didn&#8217;t have in the
<tt class="docutils literal"><span class="pre">lps_response_corr</span></tt> data:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">lps_response_corr</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)

&lt;ipython-input-53-f11acbabb21a&gt; in &lt;module&gt;()
----&gt; 1 gene_pc_clusters.index.difference(lps_response_corr.index)


/usr/local/lib/python2.7/site-packages/pandas/core/index.pyc in difference(self, other)
   1325             result_name = self.name if self.name == other.name else None
   1326
-&gt; 1327         theDiff = sorted(set(self) - set(other))
   1328         return Index(theDiff, name=result_name)
   1329


TypeError: can&#39;t compare datetime.datetime to unicode
</pre></div>
</div>
<p>Oh joy, another <tt class="docutils literal"><span class="pre">datetime</span></tt> error, just like we had with
<tt class="docutils literal"><span class="pre">expression2</span></tt>... Looking back at the original Excel file, there is one
gene that Excel mangled to be a date:</p>
<div class="figure align-center">
<img alt="" src="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_supplementary_table_5_datetime_error.png" />
</div>
<p>Please, can we start using just plain ole <a href="#id17"><span class="problematic" id="id18">``</span></a>.csv``s for supplementary
data! Excel does NOT preserve strings if they start with numbers, and
instead thinks they are dates.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">collections</span>
<span class="n">collections</span><span class="o">.</span><span class="n">Counter</span><span class="p">(</span><span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">type</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Counter({&lt;type &#39;unicode&#39;&gt;: 631, &lt;type &#39;datetime.datetime&#39;&gt;: 1})
</pre></div>
</div>
<p>Yep, it&#8217;s just that one that got mangled.... oh well.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">gene_pc_clusters_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">unicode</span><span class="p">),</span> <span class="n">gene_pc_clusters</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
<span class="n">gene_pc_clusters_genes</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">lps_response_corr</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="p">{</span><span class="s">u&#39;RPS6KA2&#39;</span><span class="p">}</span>
</pre></div>
</div>
<p>So, &#8220;<tt class="docutils literal"><span class="pre">RPS6KA2</span></tt>&#8221; is the only gene that was in their list of genes and
not in mine.</p>
</div>
</div>
</div>
<div class="section" id="supplementary-figures">
<h2>Supplementary figures<a class="headerlink" href="#supplementary-figures" title="Permalink to this headline">¶</a></h2>
<p>Now we get to have even more fun by plotting the Supplementary figures!
:D</p>
<p>Ironically, the supplementary figures are usually way easier to access
(like not behind a paywall), and yet they&#8217;re usually the documents that
really have the crucial information about the experiments.</p>
<div class="section" id="supplementary-figure-1">
<h3>Supplementary Figure 1<a class="headerlink" href="#supplementary-figure-1" title="Permalink to this headline">¶</a></h3>
<div class="figure align-center">
<img alt="Supplementary figure 1, a correlation plot" src="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig1.png" />
<p class="caption">Supplementary figure 1, a correlation plot</p>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">singles_mean</span> <span class="o">=</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">singles_mean</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s">&#39;Single cell average&#39;</span>

<span class="c"># Need to convert &quot;average_singles&quot; to a DataFrame instead of a single-row Series</span>
<span class="n">singles_mean</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">singles_mean</span><span class="p">)</span>
<span class="n">singles_mean</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Single cell average</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>NPL</th>
      <td> 1.075740</td>
    </tr>
    <tr>
      <th>QK</th>
      <td> 2.019888</td>
    </tr>
    <tr>
      <th>AK163153</th>
      <td> 1.429369</td>
    </tr>
    <tr>
      <th>PARK2</th>
      <td> 0.596479</td>
    </tr>
    <tr>
      <th>AGPAT4</th>
      <td> 2.021294</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">data_for_correlations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">singles</span><span class="p">,</span> <span class="n">singles_mean</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">study</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">pooled</span><span class="p">])</span>

<span class="c"># Take the transpose of the data, because the plotting algorithm calculates correlations between columns,</span>
<span class="c"># And we want the correlations between samples, not features</span>
<span class="n">data_for_correlations</span> <span class="o">=</span> <span class="n">data_for_correlations</span><span class="o">.</span><span class="n">T</span>
<span class="n">data_for_correlations</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>

<span class="c"># %time sns.corrplot(data_for_correlations)</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>S1</th>
      <th>S2</th>
      <th>S3</th>
      <th>S4</th>
      <th>S5</th>
      <th>S6</th>
      <th>S7</th>
      <th>S8</th>
      <th>S9</th>
      <th>S10</th>
      <th>...</th>
      <th>S13</th>
      <th>S14</th>
      <th>S15</th>
      <th>S16</th>
      <th>S17</th>
      <th>S18</th>
      <th>Single cell average</th>
      <th>P1</th>
      <th>P2</th>
      <th>P3</th>
    </tr>
    <tr>
      <th>GENE</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>NPL</th>
      <td> 4.290577</td>
      <td> 0.000000</td>
      <td> 4.860293</td>
      <td> 0.090829</td>
      <td> 0.000000</td>
      <td> 0.000000</td>
      <td> 4.730129</td>
      <td> 4.657090</td>
      <td> 0.112641</td>
      <td> 0.000000</td>
      <td>...</td>
      <td> 0.110470</td>
      <td> 0.099121</td>
      <td> 0.100920</td>
      <td> 0.206361</td>
      <td> 0.104884</td>
      <td> 0.000000</td>
      <td> 1.075740</td>
      <td> 2.093019</td>
      <td> 2.044724</td>
      <td> 2.742480</td>
    </tr>
    <tr>
      <th>QK</th>
      <td> 5.038477</td>
      <td> 4.183371</td>
      <td> 3.847854</td>
      <td> 0.066797</td>
      <td> 3.305915</td>
      <td> 0.114225</td>
      <td> 3.730270</td>
      <td> 2.750103</td>
      <td> 0.134389</td>
      <td> 0.760353</td>
      <td>...</td>
      <td> 3.395885</td>
      <td> 2.294456</td>
      <td> 0.301120</td>
      <td> 3.547688</td>
      <td> 2.185832</td>
      <td> 0.040923</td>
      <td> 2.019888</td>
      <td> 3.869102</td>
      <td> 3.690982</td>
      <td> 3.671838</td>
    </tr>
    <tr>
      <th>AK163153</th>
      <td> 1.249363</td>
      <td> 1.947622</td>
      <td> 1.082463</td>
      <td> 1.119633</td>
      <td> 1.267464</td>
      <td> 0.901824</td>
      <td> 1.033401</td>
      <td> 0.978591</td>
      <td> 1.220720</td>
      <td> 1.035237</td>
      <td>...</td>
      <td> 2.103135</td>
      <td> 1.110511</td>
      <td> 1.202271</td>
      <td> 4.446612</td>
      <td> 1.367261</td>
      <td> 0.428320</td>
      <td> 1.429369</td>
      <td> 0.605094</td>
      <td> 0.392494</td>
      <td> 0.284990</td>
    </tr>
    <tr>
      <th>PARK2</th>
      <td> 0.540694</td>
      <td> 0.500426</td>
      <td> 0.604097</td>
      <td> 0.418703</td>
      <td> 0.000000</td>
      <td> 0.601280</td>
      <td> 0.404931</td>
      <td> 0.552874</td>
      <td> 0.343271</td>
      <td> 0.844120</td>
      <td>...</td>
      <td> 0.755072</td>
      <td> 1.109400</td>
      <td> 0.807534</td>
      <td> 0.586962</td>
      <td> 0.485122</td>
      <td> 0.091469</td>
      <td> 0.596479</td>
      <td> 0.815242</td>
      <td> 0.267032</td>
      <td> 0.645365</td>
    </tr>
    <tr>
      <th>AGPAT4</th>
      <td> 0.095072</td>
      <td> 5.868557</td>
      <td> 4.137252</td>
      <td> 0.066015</td>
      <td> 0.000000</td>
      <td> 4.750107</td>
      <td> 0.069345</td>
      <td> 4.130618</td>
      <td> 3.328758</td>
      <td> 0.000000</td>
      <td>...</td>
      <td> 0.000000</td>
      <td> 4.430612</td>
      <td> 0.000000</td>
      <td> 0.000000</td>
      <td> 4.219120</td>
      <td> 0.171028</td>
      <td> 2.021294</td>
      <td> 2.854144</td>
      <td> 2.139655</td>
      <td> 2.806291</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 22 columns</p>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">corrplot</span><span class="p">(</span><span class="n">data_for_correlations</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/shalek2013_159_0.png" src="../_images/shalek2013_159_0.png" />
<p>Notice that this is mostly red, while in the figure from the paper, it
was both blue and red. This is because the colormap started at 0.2 (not
negative), and was centered with white at about 0.6. I see that they&#8217;re
trying to emphasize how much more correlated the pooled samples are to
each other, but I think a simple sequential map would have been more
effective.</p>
</div>
<div class="section" id="supplementary-figures-2-and-3">
<h3>Supplementary Figures 2 and 3<a class="headerlink" href="#supplementary-figures-2-and-3" title="Permalink to this headline">¶</a></h3>
<p>Supplementary <a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig2.png">Figure
2</a>
and <a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig3.png">Figure
3</a>
are from FISH and raw sequence data, and are out of the scope of this
computational reproduction.</p>
</div>
<div class="section" id="supplementary-figure-4">
<h3>Supplementary Figure 4<a class="headerlink" href="#supplementary-figure-4" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig4.png">Supplementary Figure
4</a>
was from published data, however the citation in the Supplementary
Information (#23) was a <a class="reference external" href="http://link.springer.com/book/10.1007%2F978-3-642-51175-2">machine-learning
book</a>, and
#23 in the main text citations was a <a class="reference external" href="http://www.sciencemag.org/content/303/5659/799.full">review of probabilistic graphical
models</a>, neither
of which have the mouse embryonic stem cells or mouse embryonic
fibroblasts used in the figure.</p>
</div>
<div class="section" id="supplementary-figure-5">
<h3>Supplementary Figure 5<a class="headerlink" href="#supplementary-figure-5" title="Permalink to this headline">¶</a></h3>
<p>For this figure, we can only plot 5d, since it&#8217;s derived directly from a
table in their dataset.</p>
<p>Warning: these data are going to require some serious cleaning. Yay data
janitorial duties!</p>
<div class="figure align-center">
<img alt="" src="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig5.png" />
</div>
<div class="section" id="supplementary-figure-5d">
<h4>Supplementary Figure 5d<a class="headerlink" href="#supplementary-figure-5d" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s">&#39;nature12172-s1/Supplementary_Table7.xlsx&#39;</span><span class="p">)</span>
<span class="n">barcoded</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TPM</th>
      <th>Unnamed: 1</th>
      <th>Unnamed: 2</th>
      <th>Unnamed: 3</th>
      <th>Unique Barcodes</th>
      <th>Unnamed: 5</th>
      <th>Unnamed: 6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GENE</th>
      <td>    MB_S1</td>
      <td>    MB_S2</td>
      <td>     MB_S3</td>
      <td>NaN</td>
      <td> MB_S1</td>
      <td> MB_S2</td>
      <td> MB_S3</td>
    </tr>
    <tr>
      <th>0610007L01RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>  5.595054</td>
      <td>NaN</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td> 76.25091</td>
      <td> 38.77614</td>
      <td> 0.1823286</td>
      <td>NaN</td>
      <td>    23</td>
      <td>     8</td>
      <td>     0</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td> 24.26729</td>
      <td> 50.24694</td>
      <td>  17.74422</td>
      <td>NaN</td>
      <td>    14</td>
      <td>     5</td>
      <td>     6</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>         0</td>
      <td>NaN</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
    </tr>
  </tbody>
</table>
</div><p>The first three columns are TPM calculated from the three samples that
have molecular barcodes, and the last three columns are the integer
counts of molecular barcodes from the three molecular barcode samples.</p>
<p>Let&#8217;s remove the &#8220;Unnamed: 3&#8221; column which is all NaNs. We&#8217;ll do that
with the <tt class="docutils literal"><span class="pre">.dropna</span></tt> method, specifying <tt class="docutils literal"><span class="pre">axis=1</span></tt> for columns and
<tt class="docutils literal"><span class="pre">how=&quot;all&quot;</span></tt> to make sure only columns that have ALL NaNs are removed.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded</span> <span class="o">=</span> <span class="n">barcoded</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">barcoded</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TPM</th>
      <th>Unnamed: 1</th>
      <th>Unnamed: 2</th>
      <th>Unique Barcodes</th>
      <th>Unnamed: 5</th>
      <th>Unnamed: 6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GENE</th>
      <td>    MB_S1</td>
      <td>    MB_S2</td>
      <td>     MB_S3</td>
      <td> MB_S1</td>
      <td> MB_S2</td>
      <td> MB_S3</td>
    </tr>
    <tr>
      <th>0610007L01RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>  5.595054</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td> 76.25091</td>
      <td> 38.77614</td>
      <td> 0.1823286</td>
      <td>    23</td>
      <td>     8</td>
      <td>     0</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td> 24.26729</td>
      <td> 50.24694</td>
      <td>  17.74422</td>
      <td>    14</td>
      <td>     5</td>
      <td>     6</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>         0</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
    </tr>
  </tbody>
</table>
</div><p>Next, let&#8217;s drop that pesky &#8220;GENE&#8221; row. Don&#8217;t worry, we&#8217;ll get the
sample ID names back next.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded</span> <span class="o">=</span> <span class="n">barcoded</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s">&#39;GENE&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">barcoded</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TPM</th>
      <th>Unnamed: 1</th>
      <th>Unnamed: 2</th>
      <th>Unique Barcodes</th>
      <th>Unnamed: 5</th>
      <th>Unnamed: 6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0610007L01RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>  5.595054</td>
      <td>  0</td>
      <td>  0</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td> 76.25091</td>
      <td> 38.77614</td>
      <td> 0.1823286</td>
      <td> 23</td>
      <td>  8</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td> 24.26729</td>
      <td> 50.24694</td>
      <td>  17.74422</td>
      <td> 14</td>
      <td>  5</td>
      <td> 6</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>        0</td>
      <td>        0</td>
      <td>         0</td>
      <td>  0</td>
      <td>  0</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610009B22RIK</th>
      <td> 67.12981</td>
      <td> 115.1393</td>
      <td>  55.98812</td>
      <td> 11</td>
      <td> 18</td>
      <td> 8</td>
    </tr>
  </tbody>
</table>
</div><p>We&#8217;ll create a <tt class="docutils literal"><span class="pre">pandas.MultiIndex</span></tt> from the tuples of
<tt class="docutils literal"><span class="pre">(sample_id,</span> <span class="pre">measurement_type)</span></tt> pair.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">columns</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_tuples</span><span class="p">([(</span><span class="s">&#39;MB_S1&#39;</span><span class="p">,</span> <span class="s">&#39;TPM&#39;</span><span class="p">),</span>
           <span class="p">(</span><span class="s">&#39;MB_S2&#39;</span><span class="p">,</span> <span class="s">&#39;TPM&#39;</span><span class="p">),</span>
           <span class="p">(</span><span class="s">&#39;MB_S3&#39;</span><span class="p">,</span> <span class="s">&#39;TPM&#39;</span><span class="p">),</span>
           <span class="p">(</span><span class="s">&#39;MB_S1&#39;</span><span class="p">,</span> <span class="s">&#39;Unique Barcodes&#39;</span><span class="p">),</span>
           <span class="p">(</span><span class="s">&#39;MB_S2&#39;</span><span class="p">,</span> <span class="s">&#39;Unique Barcodes&#39;</span><span class="p">),</span>
           <span class="p">(</span><span class="s">&#39;MB_S3&#39;</span><span class="p">,</span> <span class="s">&#39;Unique Barcodes&#39;</span><span class="p">)])</span>
<span class="n">barcoded</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">columns</span>
<span class="n">barcoded</span> <span class="o">=</span> <span class="n">barcoded</span><span class="o">.</span><span class="n">sort_index</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">barcoded</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr>
      <th></th>
      <th colspan="2" halign="left">MB_S1</th>
      <th colspan="2" halign="left">MB_S2</th>
      <th colspan="2" halign="left">MB_S3</th>
    </tr>
    <tr>
      <th></th>
      <th>TPM</th>
      <th>Unique Barcodes</th>
      <th>TPM</th>
      <th>Unique Barcodes</th>
      <th>TPM</th>
      <th>Unique Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0610007L01RIK</th>
      <td>        0</td>
      <td>  0</td>
      <td>        0</td>
      <td>  0</td>
      <td>  5.595054</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610007P14RIK</th>
      <td> 76.25091</td>
      <td> 23</td>
      <td> 38.77614</td>
      <td>  8</td>
      <td> 0.1823286</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610007P22RIK</th>
      <td> 24.26729</td>
      <td> 14</td>
      <td> 50.24694</td>
      <td>  5</td>
      <td>  17.74422</td>
      <td> 6</td>
    </tr>
    <tr>
      <th>0610008F07RIK</th>
      <td>        0</td>
      <td>  0</td>
      <td>        0</td>
      <td>  0</td>
      <td>         0</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>0610009B22RIK</th>
      <td> 67.12981</td>
      <td> 11</td>
      <td> 115.1393</td>
      <td> 18</td>
      <td>  55.98812</td>
      <td> 8</td>
    </tr>
  </tbody>
</table>
</div><p>For the next move, we&#8217;re going to do some crazy <tt class="docutils literal"><span class="pre">pandas</span></tt>-fu. First
we&#8217;re going to transpose, then <tt class="docutils literal"><span class="pre">reset_index</span></tt> of the transpose. Just so
you know what this looks like, it&#8217;s this.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>level_0</th>
      <th>level_1</th>
      <th>0610007L01RIK</th>
      <th>0610007P14RIK</th>
      <th>0610007P22RIK</th>
      <th>0610008F07RIK</th>
      <th>0610009B22RIK</th>
      <th>0610009D07RIK</th>
      <th>0610009O20RIK</th>
      <th>0610010B08RIK</th>
      <th>...</th>
      <th>ZWILCH</th>
      <th>ZWINT</th>
      <th>ZXDA</th>
      <th>ZXDB</th>
      <th>ZXDC</th>
      <th>ZYG11A</th>
      <th>ZYG11B</th>
      <th>ZYX</th>
      <th>ZZEF1</th>
      <th>ZZZ3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> MB_S1</td>
      <td>             TPM</td>
      <td>        0</td>
      <td>  76.25091</td>
      <td> 24.26729</td>
      <td> 0</td>
      <td> 67.12981</td>
      <td> 132.2392</td>
      <td> 17.03907</td>
      <td> 0.01375923</td>
      <td>...</td>
      <td> 0</td>
      <td> 206.8494</td>
      <td>        0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0.01985733</td>
      <td> 55.28996</td>
      <td> 0.09482778</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>1</th>
      <td> MB_S1</td>
      <td> Unique Barcodes</td>
      <td>        0</td>
      <td>        23</td>
      <td>       14</td>
      <td> 0</td>
      <td>       11</td>
      <td>       29</td>
      <td>        3</td>
      <td>          1</td>
      <td>...</td>
      <td> 0</td>
      <td>       33</td>
      <td>        0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>          0</td>
      <td>        6</td>
      <td>          0</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>2</th>
      <td> MB_S2</td>
      <td>             TPM</td>
      <td>        0</td>
      <td>  38.77614</td>
      <td> 50.24694</td>
      <td> 0</td>
      <td> 115.1393</td>
      <td> 49.16287</td>
      <td>        0</td>
      <td>          0</td>
      <td>...</td>
      <td> 0</td>
      <td>  48.7729</td>
      <td>        0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>   7.894789</td>
      <td> 135.1977</td>
      <td>          0</td>
      <td> 4.272594</td>
    </tr>
    <tr>
      <th>3</th>
      <td> MB_S2</td>
      <td> Unique Barcodes</td>
      <td>        0</td>
      <td>         8</td>
      <td>        5</td>
      <td> 0</td>
      <td>       18</td>
      <td>       11</td>
      <td>        0</td>
      <td>          0</td>
      <td>...</td>
      <td> 0</td>
      <td>       10</td>
      <td>        0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>          0</td>
      <td>        7</td>
      <td>          0</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>4</th>
      <td> MB_S3</td>
      <td>             TPM</td>
      <td> 5.595054</td>
      <td> 0.1823286</td>
      <td> 17.74422</td>
      <td> 0</td>
      <td> 55.98812</td>
      <td> 203.6302</td>
      <td>        0</td>
      <td>  0.4914763</td>
      <td>...</td>
      <td> 0</td>
      <td> 54.51386</td>
      <td> 1.120081</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>  0.1238624</td>
      <td> 340.7358</td>
      <td>  0.6677646</td>
      <td>        0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 27725 columns</p>
</div><p>Next, we&#8217;re going to transform the data into a
<a class="reference external" href="http://vita.had.co.nz/papers/tidy-data.pdf">tidy</a> format, with
separate columns for sample ids, measurement types, the gene that was
measured, and its measurement value.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">barcoded</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span> <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;level_0&#39;</span><span class="p">,</span> <span class="s">&#39;level_1&#39;</span><span class="p">])</span>
<span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>level_0</th>
      <th>level_1</th>
      <th>variable</th>
      <th>value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> MB_S1</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>1</th>
      <td> MB_S1</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>2</th>
      <td> MB_S2</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>3</th>
      <td> MB_S2</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>4</th>
      <td> MB_S3</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td> 5.595054</td>
    </tr>
  </tbody>
</table>
</div><p>Now let&#8217;s rename these columns into something more useful, instead of
&#8220;level_0&#8221;</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;level_0&#39;</span><span class="p">:</span> <span class="s">&#39;sample_id&#39;</span><span class="p">,</span> <span class="s">&#39;level_1&#39;</span><span class="p">:</span> <span class="s">&#39;measurement&#39;</span><span class="p">,</span> <span class="s">&#39;variable&#39;</span><span class="p">:</span> <span class="s">&#39;gene_name&#39;</span><span class="p">})</span>
<span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>measurement</th>
      <th>gene_name</th>
      <th>value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> MB_S1</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>1</th>
      <td> MB_S1</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>2</th>
      <td> MB_S2</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>3</th>
      <td> MB_S2</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
    </tr>
    <tr>
      <th>4</th>
      <td> MB_S3</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td> 5.595054</td>
    </tr>
  </tbody>
</table>
</div><p>Next, we&#8217;re going to take some seemingly-duplicating steps, but trust
me, it&#8217;ll make the data easier.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;TPM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">value</span><span class="p">[</span><span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">measurement</span> <span class="o">==</span> <span class="s">&#39;TPM&#39;</span><span class="p">]</span>
<span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;Unique Barcodes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">value</span><span class="p">[</span><span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">measurement</span> <span class="o">==</span> <span class="s">&#39;Unique Barcodes&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Fill the values of the &#8220;<strong>TPM</strong>&#8220;&#8216;s forwards, since they appear first,
and fill the values of the &#8220;<strong>Unique Barcodes</strong>&#8221; backwards, since
they&#8217;re second</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">TPM</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">TPM</span><span class="o">.</span><span class="n">ffill</span><span class="p">()</span>
<span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;Unique Barcodes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;Unique Barcodes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">bfill</span><span class="p">()</span>
<span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>measurement</th>
      <th>gene_name</th>
      <th>value</th>
      <th>TPM</th>
      <th>Unique Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> MB_S1</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td> 0.000000</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>1</th>
      <td> MB_S1</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td> 0.000000</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>2</th>
      <td> MB_S2</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td> 0.000000</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>3</th>
      <td> MB_S2</td>
      <td> Unique Barcodes</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td> 0.000000</td>
      <td> 0</td>
    </tr>
    <tr>
      <th>4</th>
      <td> MB_S3</td>
      <td>             TPM</td>
      <td> 0610007L01RIK</td>
      <td> 5.595054</td>
      <td> 5.595054</td>
      <td> 0</td>
    </tr>
  </tbody>
</table>
</div><p>Drop the &#8220;<strong>measurement</strong>&#8221; column and drop duplicate rows.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s">&#39;measurement&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">barcoded_tidy</span> <span class="o">=</span> <span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>gene_name</th>
      <th>value</th>
      <th>TPM</th>
      <th>Unique Barcodes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> MB_S1</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td>  0.000000</td>
      <td>  0</td>
    </tr>
    <tr>
      <th>2</th>
      <td> MB_S2</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td>  0.000000</td>
      <td>  0</td>
    </tr>
    <tr>
      <th>4</th>
      <td> MB_S3</td>
      <td> 0610007L01RIK</td>
      <td> 5.595054</td>
      <td>  5.595054</td>
      <td>  0</td>
    </tr>
    <tr>
      <th>5</th>
      <td> MB_S3</td>
      <td> 0610007L01RIK</td>
      <td>        0</td>
      <td>  5.595054</td>
      <td>  0</td>
    </tr>
    <tr>
      <th>6</th>
      <td> MB_S1</td>
      <td> 0610007P14RIK</td>
      <td> 76.25091</td>
      <td> 76.250913</td>
      <td> 23</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-python"><div class="highlight"><pre><span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;log TPM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">barcoded_tidy</span><span class="o">.</span><span class="n">TPM</span><span class="p">)</span>
<span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;log Unique Barcodes&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">barcoded_tidy</span><span class="p">[</span><span class="s">&#39;Unique Barcodes&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>Now we can use the convenient linear model plot (<tt class="docutils literal"><span class="pre">lmplot</span></tt>) in
<tt class="docutils literal"><span class="pre">seaborn</span></tt> to plot these three samples together!</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">sns</span><span class="o">.</span><span class="n">lmplot</span><span class="p">(</span><span class="s">&#39;log TPM&#39;</span><span class="p">,</span> <span class="s">&#39;log Unique Barcodes&#39;</span><span class="p">,</span> <span class="n">barcoded_tidy</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&#39;sample_id&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>&lt;seaborn.axisgrid.FacetGrid at 0x123293390&gt;
</pre></div>
</div>
<img alt="../_images/shalek2013_189_1.png" src="../_images/shalek2013_189_1.png" />
</div>
</div>
<div class="section" id="supplementary-figures-6-20">
<h3>Supplementary Figures 6-20<a class="headerlink" href="#supplementary-figures-6-20" title="Permalink to this headline">¶</a></h3>
<p>Supplementary Figures
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig6.png">6</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig7.png">7</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig8.png">8</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig9.png">9</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig10.png">10</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig11.png">11</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig12.png">12</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig13.png">13</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig14.png">14</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig15.png">15</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig16.png">16</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig17.png">17</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig18.png">18</a>,
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig19.png">19</a>,
and
<a class="reference external" href="https://raw.githubusercontent.com/olgabot/olgabot.github.io-source/master/content/images/shalek2013_sfig20.png">20</a>,
deal with splicing data from the molecular barcodes, RNA-FISH,
flow-sorted cells, and single-cell RT-PCR and are out of the scope of
this reproduction.</p>
</div>
</div>
<div class="section" id="conclusions">
<h2>Conclusions<a class="headerlink" href="#conclusions" title="Permalink to this headline">¶</a></h2>
<p>While there may be minor, undocumented, differences between the methods
presented in the manuscript and the figures, the application of
<tt class="docutils literal"><span class="pre">`flotilla</span></tt> &lt;<a class="reference external" href="https://github.com/YeoLab/flotilla">https://github.com/YeoLab/flotilla</a>&gt;`_ presents an
opportunity to avoid these types of inconsistencies by strictly
documenting every change to code and every transformation of the data.
The biology the authors found is clearly real, as they did the knockout
experiment of <em>Ifnr-/-</em> and saw that indeed the maturation process was
affected, and <em>Stat2</em> and <em>Irf7</em> had much lower expression, as with the
&#8220;maturing&#8221; cells in the data.</p>
</div>
</div>


    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

    
  </body>
</html>