<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>flotilla.data_model.study &mdash; flotilla 0.2.8dev documentation</title>
    
    <link rel="stylesheet" href="../../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/bootswatch-3.3.4/paper/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.2.8dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../../_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="flotilla 0.2.8dev documentation" href="../../../index.html" />
    <link rel="up" title="flotilla.data_model" href="../data_model.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../../index.html">
          Flotilla</a>
        <span class="navbar-text navbar-version pull-left"><b>0.2.8dev</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../../../tutorial.html">Tutorial</a></li>
                <li><a href="../../../gallery/index.html">Gallery</a></li>
                <li><a href="http://yeolab.ucsd.edu">YeoLab@UCSD</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">flotilla</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../flotilla.html">flotilla package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../whatsnew.html">What&#8217;s new in the package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-8">v0.2.8 (........)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-7-april-20th-2015">v0.2.7 (April 20th, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-6-april-10th-2015">v0.2.6 (April 10th, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-5-march-3rd-2015">v0.2.5 (March 3rd, 2015)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-4-november-23rd-2014">v0.2.4 (November 23rd, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-3-november-17th-2014">v0.2.3 (November 17th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-2-november-7th-2014">v0.2.2 (November 7th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-1-november-6th-2014">v0.2.1 (November 6th, 2014)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../whatsnew.html#v0-2-0-november-5th-2014">v0.2.0 (November 5th, 2014)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">OS X Installation instructions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#create-a-flotilla-sandbox">Create a flotilla sandbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#install-and-update-all-packages-in-your-environment">Install and update all packages in your environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html#old-osx-installation-instructions">Old OSX Installation instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html#docker-instructions">Docker Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../api.html#general-use-plots">General use plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../api.html#interactive-plots">Interactive plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../api.html#machine-learning-plots">Machine learning plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../api.html#splicing-specific-plots">Splicing-specific plots</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../create_datapackage.html">Create your very own datapackage!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Flotilla tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial.html#install-flotilla">Install flotilla</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial.html#create-a-datapackage">Create a datapackage</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial.html#computing-on-splicing-data">Computing on splicing data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial.html#day-to-day-helpful-tips">Day to day helpful tips</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial.html#recreated-papers">Recreated papers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"></ul>
</li>
              
            
            
            
            
            
              <li class="hidden-sm"></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12">
      
  <h1>Source code for flotilla.data_model.study</h1><div class="highlight"><pre>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Data models for &quot;studies&quot; studies include attributes about the data and are</span>
<span class="sd">heavier in terms of data load</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">inspect</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">semantic_version</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>

<span class="kn">from</span> <span class="nn">.metadata</span> <span class="kn">import</span> <span class="n">MetaData</span><span class="p">,</span> <span class="n">PHENOTYPE_COL</span><span class="p">,</span> <span class="n">POOLED_COL</span><span class="p">,</span> <span class="n">OUTLIER_COL</span>
<span class="kn">from</span> <span class="nn">.expression</span> <span class="kn">import</span> <span class="n">ExpressionData</span><span class="p">,</span> <span class="n">SpikeInData</span>
<span class="kn">from</span> <span class="nn">.gene_ontology</span> <span class="kn">import</span> <span class="n">GeneOntologyData</span>
<span class="kn">from</span> <span class="nn">.quality_control</span> <span class="kn">import</span> <span class="n">MappingStatsData</span><span class="p">,</span> <span class="n">MIN_READS</span>
<span class="kn">from</span> <span class="nn">.splicing</span> <span class="kn">import</span> <span class="n">SplicingData</span><span class="p">,</span> <span class="n">FRACTION_DIFF_THRESH</span>
<span class="kn">from</span> <span class="nn">.supplemental</span> <span class="kn">import</span> <span class="n">SupplementalData</span>
<span class="kn">from</span> <span class="nn">..compute.predict</span> <span class="kn">import</span> <span class="n">PredictorConfigManager</span>
<span class="kn">from</span> <span class="nn">..datapackage</span> <span class="kn">import</span> <span class="n">datapackage_url_to_dict</span><span class="p">,</span> \
    <span class="n">check_if_already_downloaded</span><span class="p">,</span> <span class="n">make_study_datapackage</span>
<span class="kn">from</span> <span class="nn">..visualize.color</span> <span class="kn">import</span> <span class="n">blue</span>
<span class="kn">from</span> <span class="nn">..visualize.ipython_interact</span> <span class="kn">import</span> <span class="n">Interactive</span>
<span class="kn">from</span> <span class="nn">..datapackage</span> <span class="kn">import</span> <span class="n">FLOTILLA_DOWNLOAD_DIR</span>
<span class="kn">from</span> <span class="nn">..util</span> <span class="kn">import</span> <span class="n">load_csv</span><span class="p">,</span> <span class="n">load_json</span><span class="p">,</span> <span class="n">load_tsv</span><span class="p">,</span> <span class="n">load_gzip_pickle_df</span><span class="p">,</span> \
    <span class="n">load_pickle_df</span><span class="p">,</span> <span class="n">timestamp</span><span class="p">,</span> <span class="n">cached_property</span>


<span class="n">SPECIES_DATA_PACKAGE_BASE_URL</span> <span class="o">=</span> <span class="s">&#39;https://s3-us-west-2.amazonaws.com/&#39;</span> \
                                <span class="s">&#39;flotilla-projects&#39;</span>
<span class="n">DATAPACKAGE_RESOURCE_COMMON_KWS</span> <span class="o">=</span> <span class="p">(</span><span class="s">&#39;url&#39;</span><span class="p">,</span> <span class="s">&#39;path&#39;</span><span class="p">,</span> <span class="s">&#39;format&#39;</span><span class="p">,</span> <span class="s">&#39;compression&#39;</span><span class="p">,</span>
                                   <span class="s">&#39;name&#39;</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_is_absolute_path</span><span class="p">(</span><span class="n">location</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">location</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;http&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">location</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span>


<div class="viewcode-block" id="Study"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study">[docs]</a><span class="k">class</span> <span class="nc">Study</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;A biological study, with associated metadata, expression, and splicing</span>
<span class="sd">    data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">default_feature_set_ids</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c"># Data types with enough data that we&#39;d probably reduce them, and even</span>
    <span class="c"># then we might want to take subsets. E.g. most variant genes for</span>
    <span class="c"># expression. But we don&#39;t expect to do this for spikein or mapping_stats</span>
    <span class="c"># data</span>
    <span class="n">_subsetable_data_types</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="s">&#39;splicing&#39;</span><span class="p">]</span>

    <span class="n">initializers</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;metadata_data&#39;</span><span class="p">:</span> <span class="n">MetaData</span><span class="p">,</span>
                    <span class="s">&#39;expression_data&#39;</span><span class="p">:</span> <span class="n">ExpressionData</span><span class="p">,</span>
                    <span class="s">&#39;splicing_data&#39;</span><span class="p">:</span> <span class="n">SplicingData</span><span class="p">,</span>
                    <span class="s">&#39;mapping_stats_data&#39;</span><span class="p">:</span> <span class="n">MappingStatsData</span><span class="p">,</span>
                    <span class="s">&#39;spikein_data&#39;</span><span class="p">:</span> <span class="n">SpikeInData</span><span class="p">}</span>

    <span class="n">readers</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;tsv&#39;</span><span class="p">:</span> <span class="n">load_tsv</span><span class="p">,</span>
               <span class="s">&#39;csv&#39;</span><span class="p">:</span> <span class="n">load_csv</span><span class="p">,</span>
               <span class="s">&#39;json&#39;</span><span class="p">:</span> <span class="n">load_json</span><span class="p">,</span>
               <span class="s">&#39;pickle_df&#39;</span><span class="p">:</span> <span class="n">load_pickle_df</span><span class="p">,</span>
               <span class="s">&#39;gzip_pickle_df&#39;</span><span class="p">:</span> <span class="n">load_gzip_pickle_df</span><span class="p">}</span>

    <span class="n">_default_reducer_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;whiten&#39;</span><span class="p">:</span> <span class="bp">False</span><span class="p">,</span>
                               <span class="s">&#39;show_point_labels&#39;</span><span class="p">:</span> <span class="bp">False</span><span class="p">,</span>
                               <span class="s">&#39;show_vectors&#39;</span><span class="p">:</span> <span class="bp">False</span><span class="p">}</span>
    <span class="n">_common_id</span> <span class="o">=</span> <span class="s">&#39;common_id&#39;</span>
    <span class="n">_sample_id</span> <span class="o">=</span> <span class="s">&#39;sample_id&#39;</span>
    <span class="n">_event_name</span> <span class="o">=</span> <span class="s">&#39;event_name&#39;</span>

    <span class="n">_default_plot_kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;marker&#39;</span><span class="p">:</span> <span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="s">&#39;color&#39;</span><span class="p">:</span> <span class="n">blue</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s">&#39;0.1.0&#39;</span><span class="p">,</span>
                 <span class="n">metadata_pooled_col</span><span class="o">=</span><span class="n">POOLED_COL</span><span class="p">,</span>
                 <span class="n">metadata_minimum_samples</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                 <span class="n">metadata_phenotype_col</span><span class="o">=</span><span class="n">PHENOTYPE_COL</span><span class="p">,</span>
                 <span class="n">metadata_phenotype_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">metadata_phenotype_to_color</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">metadata_phenotype_to_marker</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">metadata_outlier_col</span><span class="o">=</span><span class="n">OUTLIER_COL</span><span class="p">,</span>
                 <span class="n">metadata_ignore_subset_cols</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">metadata_batch_cols</span><span class="o">=</span><span class="s">&#39;batch&#39;</span><span class="p">,</span>
                 <span class="n">metadata_batch_min_samples</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                 <span class="n">expression_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">expression_feature_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">expression_feature_rename_col</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">expression_feature_ignore_subset_cols</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">expression_log_base</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span>
                 <span class="n">expression_plus_one</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">expression_correct_batch_effects</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">splicing_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">splicing_feature_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">splicing_feature_rename_col</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">splicing_feature_ignore_subset_cols</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">splicing_feature_expression_id_col</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">mapping_stats_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">mapping_stats_number_mapped_col</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">mapping_stats_min_reads</span><span class="o">=</span><span class="n">MIN_READS</span><span class="p">,</span>
                 <span class="n">spikein_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">spikein_feature_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">drop_outliers</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">gene_ontology_data</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">license</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">default_sample_subset</span><span class="o">=</span><span class="s">&quot;all_samples&quot;</span><span class="p">,</span>
                 <span class="n">default_feature_subset</span><span class="o">=</span><span class="s">&quot;variant&quot;</span><span class="p">,</span>
                 <span class="n">supplemental_data</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Construct a biological study</span>

<span class="sd">        This class only accepts data, no filenames. All data must already</span>
<span class="sd">        have been read in and exist as Python objects.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_metadata : pandas.DataFrame</span>
<span class="sd">            The only required parameter. Samples as the index, with features as</span>
<span class="sd">            columns. Required column: &quot;phenotype&quot;. If there is a boolean</span>
<span class="sd">            column &quot;pooled&quot;, this will be used to separate pooled from single</span>
<span class="sd">            cells. Similarly, the column &quot;outliers&quot; will also be used to</span>
<span class="sd">            separate outlier cells from the rest.</span>
<span class="sd">        version : str</span>
<span class="sd">            A string describing the semantic version of the data. Must be in:</span>
<span class="sd">            major.minor.patch format, as the &quot;patch&quot; number will be increased</span>
<span class="sd">            if you change something in the study and then study.save() it.</span>
<span class="sd">            (default &quot;0.1.0&quot;)</span>
<span class="sd">        expression_data : pandas.DataFrame</span>
<span class="sd">            Samples x feature dataframe of gene expression measurements,</span>
<span class="sd">            e.g. from an RNA-Seq or a microarray experiment. Assumed to be</span>
<span class="sd">            log-transformed, i.e. you took the log of it. (default None)</span>
<span class="sd">        expression_feature_data : pandas.DatFrame</span>
<span class="sd">            Features x annotations dataframe describing other parameters</span>
<span class="sd">            of the gene expression features, e.g. mapping Ensembl IDs to gene</span>
<span class="sd">            symbols or gene biotypes. (default None)</span>
<span class="sd">        expression_feature_rename_col : str</span>
<span class="sd">            A column name in the expression_feature_data dataframe that you&#39;d</span>
<span class="sd">            like to rename the expression features to, in the plots. For</span>
<span class="sd">            example, if your gene IDs are Ensembl IDs, but you want to plot</span>
<span class="sd">            UCSC IDs, make sure the column you want, e.g. &quot;ucsc_id&quot; is in your</span>
<span class="sd">            dataframe and specify that. (default &quot;gene_name&quot;)</span>
<span class="sd">        expression_log_base : float</span>
<span class="sd">            If you want to log-transform your expression data (and it&#39;s not</span>
<span class="sd">            already log-transformed), use this number as the base of the</span>
<span class="sd">            transform. E.g. expression_log_base=10 will take the log10 of</span>
<span class="sd">            your data. (default None)</span>
<span class="sd">        thresh : float</span>
<span class="sd">            Minimum (non log-transformed) expression value. (default -inf)</span>
<span class="sd">        expression_plus_one : bool</span>
<span class="sd">            Whether or not to add 1 to the expression data. (default False)</span>
<span class="sd">        splicing_data : pandas.DataFrame</span>
<span class="sd">            Samples x feature dataframe of percent spliced in scores, e.g. as</span>
<span class="sd">            measured by the program MISO. Assumed that these values only fall</span>
<span class="sd">            between 0 and 1.</span>
<span class="sd">        splicing_feature_data : pandas.DataFrame</span>
<span class="sd">            features x other_features dataframe describing other parameters</span>
<span class="sd">            of the splicing features, e.g. mapping MISO IDs to Ensembl IDs or</span>
<span class="sd">            gene symbols or transcript types</span>
<span class="sd">        splicing_feature_rename_col : str</span>
<span class="sd">            A column name in the splicing_feature_data dataframe that you&#39;d</span>
<span class="sd">            like to rename the splicing features to, in the plots. For</span>
<span class="sd">            example, if your splicing IDs are MISO IDs, but you want to plot</span>
<span class="sd">            Ensembl IDs, make sure the column you want, e.g. &quot;ensembl_id&quot; is</span>
<span class="sd">            in your dataframe and specify that. Default &quot;gene_name&quot;.</span>
<span class="sd">        splicing_feature_expression_id_col : str</span>
<span class="sd">            A column name in the splicing_feature_data dataframe that</span>
<span class="sd">            corresponds to the row names of the expression data</span>
<span class="sd">        mapping_stats_data : pandas.DataFrame</span>
<span class="sd">            Samples x feature dataframe of mapping stats measurements.</span>
<span class="sd">            Currently, this</span>
<span class="sd">        mapping_stats_number_mapped_col : str</span>
<span class="sd">            A column name in the mapping_stats_data which specifies the</span>
<span class="sd">            number of (uniquely or not) mapped reads. Default &quot;Uniquely</span>
<span class="sd">            mapped reads number&quot;</span>
<span class="sd">        spikein_data : pandas.DataFrame</span>
<span class="sd">            samples x features DataFrame of spike-in expression values</span>
<span class="sd">        spikein_feature_data : pandas.DataFrame</span>
<span class="sd">            Features x other_features dataframe, e.g. of the molecular</span>
<span class="sd">            concentration of particular spikein transcripts</span>
<span class="sd">        drop_outliers : bool</span>
<span class="sd">            Whether or not to drop samples indicated as outliers in the</span>
<span class="sd">            sample_metadata from the other data, i.e. with a column</span>
<span class="sd">            named &#39;outlier&#39; in sample_metadata, then remove those</span>
<span class="sd">            samples from expression_data for further analysis</span>
<span class="sd">        species : str</span>
<span class="sd">            Name of the species and genome version, e.g. &#39;hg19&#39; or &#39;mm10&#39;.</span>
<span class="sd">        gene_ontology_data : pandas.DataFrame</span>
<span class="sd">            Gene ids x ontology categories dataframe used for GO analysis.</span>
<span class="sd">        metadata_pooled_col : str</span>
<span class="sd">            Column in metadata_data which specifies as a boolean</span>
<span class="sd">            whether or not this sample was pooled.</span>
<span class="sd">        supplemental_data : dict</span>
<span class="sd">            str: dataframe mapping of the attribute name, and the pandas</span>
<span class="sd">            dataframe</span>

<span class="sd">        Note</span>
<span class="sd">        ----</span>
<span class="sd">        This function explicitly specifies ALL the instance variables (except</span>
<span class="sd">        those that are marked by the @property decorator), because,</span>
<span class="sd">        as described [1], &quot;If you write initialization functions separate from</span>
<span class="sd">        __init__ then experienced developers will certainly see your code as a</span>
<span class="sd">        kid&#39;s playground.&quot;</span>

<span class="sd">        [1] http://stackoverflow.com/q/12513185/1628971</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{}</span><span class="se">\t</span><span class="s">Initializing Study</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{}</span><span class="se">\t</span><span class="s">Initializing Predictor configuration manager &quot;</span>
                         <span class="s">&quot;for Study</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span> <span class="o">=</span> <span class="n">predictor_config_manager</span> \
            <span class="k">if</span> <span class="n">predictor_config_manager</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> \
            <span class="k">else</span> <span class="n">PredictorConfigManager</span><span class="p">()</span>
        <span class="c"># self.predictor_config_manager = None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">species</span> <span class="o">=</span> <span class="n">species</span>
        <span class="k">if</span> <span class="n">gene_ontology_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">gene_ontology</span> <span class="o">=</span> <span class="n">GeneOntologyData</span><span class="p">(</span><span class="n">gene_ontology_data</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">license</span> <span class="o">=</span> <span class="n">license</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="n">title</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sources</span> <span class="o">=</span> <span class="n">sources</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">version</span> <span class="o">=</span> <span class="n">version</span>

        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">Loading metadata</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span> <span class="o">=</span> <span class="n">MetaData</span><span class="p">(</span>
            <span class="n">sample_metadata</span><span class="p">,</span> <span class="n">metadata_phenotype_order</span><span class="p">,</span>
            <span class="n">metadata_phenotype_to_color</span><span class="p">,</span>
            <span class="n">metadata_phenotype_to_marker</span><span class="p">,</span> <span class="n">pooled_col</span><span class="o">=</span><span class="n">metadata_pooled_col</span><span class="p">,</span>
            <span class="n">ignore_subset_cols</span><span class="o">=</span><span class="n">metadata_ignore_subset_cols</span><span class="p">,</span>
            <span class="n">outlier_col</span><span class="o">=</span><span class="n">metadata_outlier_col</span><span class="p">,</span>
            <span class="n">phenotype_col</span><span class="o">=</span><span class="n">metadata_phenotype_col</span><span class="p">,</span>
            <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span><span class="p">,</span>
            <span class="n">minimum_sample_subset</span><span class="o">=</span><span class="n">metadata_minimum_samples</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_subset</span> <span class="o">=</span> <span class="n">default_feature_subset</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">default_sample_subset</span> <span class="o">=</span> <span class="n">default_sample_subset</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">outlier_col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span> <span class="ow">and</span> <span class="n">drop_outliers</span><span class="p">:</span>
            <span class="n">outliers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">outlier_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">outliers</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">outlier_col</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span>

        <span class="c"># Get pooled samples</span>
        <span class="n">pooled</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">pooled_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">pooled_col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">pooled</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">pooled_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)]</span>
                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                    <span class="n">pooled</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pooled</span> <span class="o">=</span> <span class="n">pooled</span>

        <span class="k">if</span> <span class="n">mapping_stats_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">MappingStatsData</span><span class="p">(</span>
                <span class="n">mapping_stats_data</span><span class="p">,</span>
                <span class="n">number_mapped_col</span><span class="o">=</span><span class="n">mapping_stats_number_mapped_col</span><span class="p">,</span>
                <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span><span class="p">,</span>
                <span class="n">min_reads</span><span class="o">=</span><span class="n">mapping_stats_min_reads</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span><span class="o">.</span><span class="n">too_few_mapped</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">outliers_ids</span> <span class="o">=</span> <span class="s">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="s">&#39;Samples had too few mapped reads (&lt;{:.1e} reads)&#39;</span>
                    <span class="s">&#39;:</span><span class="se">\n\t</span><span class="s">{}</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mapping_stats_min_reads</span><span class="p">,</span> <span class="n">outliers_ids</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">feature_data_none</span> <span class="o">=</span> <span class="n">expression_feature_data</span> <span class="ow">is</span> <span class="bp">None</span> <span class="ow">or</span> \
            <span class="n">splicing_feature_data</span> <span class="ow">is</span> <span class="bp">None</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">species</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">feature_data_none</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">Loading species metadata from &#39;</span>
                             <span class="s">&#39;~/flotilla_packages</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
            <span class="n">species_kws</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_species_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">readers</span><span class="p">)</span>
            <span class="n">expression_feature_data</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                <span class="s">&#39;expression_feature_data&#39;</span><span class="p">,</span>
                <span class="bp">None</span><span class="p">)</span>
            <span class="n">expression_feature_rename_col</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                <span class="s">&#39;expression_feature_rename_col&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
            <span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;splicing_feature_data&#39;</span><span class="p">,</span>
                                                    <span class="bp">None</span><span class="p">)</span>
            <span class="n">splicing_feature_rename_col</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                <span class="s">&#39;splicing_feature_rename_col&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">expression_feature_data</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">expression_feature_data</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                    <span class="s">&#39;expression_feature_data&#39;</span><span class="p">,</span>
                    <span class="bp">None</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">expression_feature_rename_col</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">expression_feature_rename_col</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                    <span class="s">&#39;expression_feature_rename_col&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">splicing_feature_data</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                    <span class="s">&#39;splicing_feature_data&#39;</span><span class="p">,</span>
                    <span class="bp">None</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">splicing_feature_rename_col</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">splicing_feature_rename_col</span> <span class="o">=</span> <span class="n">species_kws</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span>
                    <span class="s">&#39;splicing_feature_rename_col&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">expression_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s">&quot;{}</span><span class="se">\t</span><span class="s">Loading expression data</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
            <span class="n">feature_ignore_subset_cols</span> <span class="o">=</span> <span class="n">expression_feature_ignore_subset_cols</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression</span> <span class="o">=</span> <span class="n">ExpressionData</span><span class="p">(</span>
                <span class="n">expression_data</span><span class="p">,</span>
                <span class="n">feature_data</span><span class="o">=</span><span class="n">expression_feature_data</span><span class="p">,</span>
                <span class="n">thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">feature_rename_col</span><span class="o">=</span><span class="n">expression_feature_rename_col</span><span class="p">,</span>
                <span class="n">outliers</span><span class="o">=</span><span class="n">outliers</span><span class="p">,</span> <span class="n">plus_one</span><span class="o">=</span><span class="n">expression_plus_one</span><span class="p">,</span>
                <span class="n">log_base</span><span class="o">=</span><span class="n">expression_log_base</span><span class="p">,</span> <span class="n">pooled</span><span class="o">=</span><span class="n">pooled</span><span class="p">,</span>
                <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span><span class="p">,</span>
                <span class="n">technical_outliers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span><span class="p">,</span>
                <span class="n">minimum_samples</span><span class="o">=</span><span class="n">metadata_minimum_samples</span><span class="p">,</span>
                <span class="n">feature_ignore_subset_cols</span><span class="o">=</span><span class="n">feature_ignore_subset_cols</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_set_ids</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_subsets</span>
                                                <span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="n">splicing_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{}</span><span class="se">\t</span><span class="s">Loading splicing data</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">timestamp</span><span class="p">()))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span> <span class="o">=</span> <span class="n">SplicingData</span><span class="p">(</span>
                <span class="n">splicing_data</span><span class="p">,</span> <span class="n">feature_data</span><span class="o">=</span><span class="n">splicing_feature_data</span><span class="p">,</span>
                <span class="n">feature_rename_col</span><span class="o">=</span><span class="n">splicing_feature_rename_col</span><span class="p">,</span>
                <span class="n">outliers</span><span class="o">=</span><span class="n">outliers</span><span class="p">,</span> <span class="n">pooled</span><span class="o">=</span><span class="n">pooled</span><span class="p">,</span>
                <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span><span class="p">,</span>
                <span class="n">technical_outliers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span><span class="p">,</span>
                <span class="n">minimum_samples</span><span class="o">=</span><span class="n">metadata_minimum_samples</span><span class="p">,</span>
                <span class="n">feature_ignore_subset_cols</span><span class="o">=</span><span class="n">splicing_feature_ignore_subset_cols</span><span class="p">,</span>
                <span class="n">feature_expression_id_col</span><span class="o">=</span><span class="n">splicing_feature_expression_id_col</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">if</span> <span class="n">spikein_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">spikein</span> <span class="o">=</span> <span class="n">SpikeInData</span><span class="p">(</span>
                <span class="n">spikein_data</span><span class="p">,</span> <span class="n">feature_data</span><span class="o">=</span><span class="n">spikein_feature_data</span><span class="p">,</span>
                <span class="n">technical_outliers</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">technical_outliers</span><span class="p">,</span>
                <span class="n">predictor_config_manager</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">predictor_config_manager</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">spikein</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">supplemental</span> <span class="o">=</span> <span class="n">SupplementalData</span><span class="p">(</span><span class="n">supplemental_data</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{}</span><span class="se">\t</span><span class="s">Successfully initialized a Study &quot;</span>
                         <span class="s">&quot;object!</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>

    <span class="k">def</span> <span class="nf">__setattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Check if the attribute already exists and warns on overwrite.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&#39;Over-writing attribute {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">))</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">Study</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__setattr__</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>

    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_col"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_col">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_col</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_col</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_order"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_order">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_order</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_order</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_to_color"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_to_color">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_to_color</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_to_color</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_to_marker"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_to_marker">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_to_marker</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_to_marker</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.sample_id_to_phenotype"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.sample_id_to_phenotype">[docs]</a>    <span class="k">def</span> <span class="nf">sample_id_to_phenotype</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.sample_id_to_color"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.sample_id_to_color">[docs]</a>    <span class="k">def</span> <span class="nf">sample_id_to_color</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_id_to_color</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_transitions"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_transitions">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_transitions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_transitions</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.phenotype_color_ordered"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.phenotype_color_ordered">[docs]</a>    <span class="k">def</span> <span class="nf">phenotype_color_ordered</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_color_order</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.default_sample_subsets"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.default_sample_subsets">[docs]</a>    <span class="k">def</span> <span class="nf">default_sample_subsets</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c"># move default_sample_subset to the front of the list, sort the rest</span>
        <span class="n">sorted_sample_subsets</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_subsets</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">default_sample_subset</span><span class="p">)))))</span>
        <span class="n">sorted_sample_subsets</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_sample_subset</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">sorted_sample_subsets</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.default_feature_subsets"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.default_feature_subsets">[docs]</a>    <span class="k">def</span> <span class="nf">default_feature_subsets</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">feature_subsets</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_subsetable_data_types</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">data_type</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
                <span class="n">feature_subsets</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data_type</span><span class="o">.</span><span class="n">feature_subsets</span>
            <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="k">return</span> <span class="n">feature_subsets</span>
</div>
    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="Study.from_datapackage_url"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.from_datapackage_url">[docs]</a>    <span class="k">def</span> <span class="nf">from_datapackage_url</span><span class="p">(</span>
            <span class="n">cls</span><span class="p">,</span> <span class="n">datapackage_url</span><span class="p">,</span>
            <span class="n">load_species_data</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">SPECIES_DATA_PACKAGE_BASE_URL</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create a study from a url of a datapackage.json file</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        datapackage_url : str</span>
<span class="sd">            HTTP url of a datapackage.json file, following the specification</span>
<span class="sd">            described here: http://dataprotocols.org/data-packages/ and</span>
<span class="sd">            requiring the following data resources: metadata,</span>
<span class="sd">            expression, splicing</span>
<span class="sd">        species_data_pacakge_base_url : str</span>
<span class="sd">            Base URL to fetch species-specific _ and splicing event</span>
<span class="sd">            metadata frnm.</span>
<span class="sd">            Default &#39;https://s3-us-west-2.amazonaws.com/flotilla-projects/&#39;</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        study : Study</span>
<span class="sd">            A &quot;study&quot; object containing the data described in the</span>
<span class="sd">            datapackage_url file</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        AttributeError</span>
<span class="sd">            If the datapackage.json file does not contain the required</span>
<span class="sd">            resources of metadata, expression, and splicing.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">datapackage</span> <span class="o">=</span> <span class="n">datapackage_url_to_dict</span><span class="p">(</span><span class="n">datapackage_url</span><span class="p">)</span>
        <span class="n">datapackage_dir</span> <span class="o">=</span> <span class="s">&#39;{}/{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">FLOTILLA_DOWNLOAD_DIR</span><span class="p">,</span>
                                         <span class="n">datapackage</span><span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">cls</span><span class="o">.</span><span class="n">from_datapackage</span><span class="p">(</span>
            <span class="n">datapackage</span><span class="p">,</span> <span class="n">load_species_data</span><span class="o">=</span><span class="n">load_species_data</span><span class="p">,</span>
            <span class="n">datapackage_dir</span><span class="o">=</span><span class="n">datapackage_dir</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">species_datapackage_base_url</span><span class="p">)</span>
</div>
    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="Study.from_datapackage_file"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.from_datapackage_file">[docs]</a>    <span class="k">def</span> <span class="nf">from_datapackage_file</span><span class="p">(</span>
            <span class="n">cls</span><span class="p">,</span> <span class="n">datapackage_filename</span><span class="p">,</span>
            <span class="n">load_species_data</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">SPECIES_DATA_PACKAGE_BASE_URL</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">datapackage_filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">Reading datapackage from {}</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">timestamp</span><span class="p">(),</span> <span class="n">datapackage_filename</span><span class="p">))</span>
            <span class="n">datapackage</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="n">datapackage_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">datapackage_filename</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cls</span><span class="o">.</span><span class="n">from_datapackage</span><span class="p">(</span>
            <span class="n">datapackage</span><span class="p">,</span> <span class="n">datapackage_dir</span><span class="o">=</span><span class="n">datapackage_dir</span><span class="p">,</span>
            <span class="n">load_species_data</span><span class="o">=</span><span class="n">load_species_data</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">species_datapackage_base_url</span><span class="p">)</span>
</div>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_filename_from_resource</span><span class="p">(</span><span class="n">resource</span><span class="p">,</span> <span class="n">datapackage_dir</span><span class="p">,</span>
                                <span class="n">datapackage_name</span><span class="p">):</span>
        <span class="k">if</span> <span class="s">&#39;url&#39;</span> <span class="ow">in</span> <span class="n">resource</span><span class="p">:</span>
            <span class="n">resource_url</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;url&#39;</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">_is_absolute_path</span><span class="p">(</span><span class="n">resource_url</span><span class="p">):</span>
                <span class="n">resource_url</span> <span class="o">=</span> <span class="s">&#39;{}/{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">datapackage_dir</span><span class="p">,</span>
                                              <span class="n">resource_url</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">check_if_already_downloaded</span><span class="p">(</span><span class="n">resource_url</span><span class="p">,</span>
                                                   <span class="n">datapackage_name</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">filename</span>
        <span class="k">elif</span> <span class="s">&#39;path&#39;</span> <span class="ow">in</span> <span class="n">resource</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;http&#39;</span><span class="p">):</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">check_if_already_downloaded</span><span class="p">(</span><span class="n">resource</span><span class="p">[</span><span class="s">&#39;path&#39;</span><span class="p">],</span>
                                                       <span class="n">datapackage_name</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">filename</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;path&#39;</span><span class="p">]</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">_is_absolute_path</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                    <span class="n">filename</span> <span class="o">=</span> <span class="s">&#39;{}/{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">datapackage_dir</span><span class="p">,</span>
                                              <span class="n">filename</span><span class="p">)</span>

                <span class="c"># Test if the file exists, if not, then add the datapackage</span>
                <span class="c"># file</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
                    <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">datapackage_dir</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">filename</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">None</span>

    <span class="nd">@classmethod</span>
<div class="viewcode-block" id="Study.from_datapackage"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.from_datapackage">[docs]</a>    <span class="k">def</span> <span class="nf">from_datapackage</span><span class="p">(</span>
            <span class="n">cls</span><span class="p">,</span> <span class="n">datapackage</span><span class="p">,</span> <span class="n">datapackage_dir</span><span class="o">=</span><span class="s">&#39;./&#39;</span><span class="p">,</span>
            <span class="n">load_species_data</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">SPECIES_DATA_PACKAGE_BASE_URL</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Create a study object from a datapackage dictionary</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        datapackage : dict</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        study : flotilla.Study</span>
<span class="sd">            Study object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">Parsing datapackage to create a Study &#39;</span>
                         <span class="s">&#39;object</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">timestamp</span><span class="p">()))</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">supplemental_data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">datapackage_name</span> <span class="o">=</span> <span class="n">datapackage</span><span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">resource</span> <span class="ow">in</span> <span class="n">datapackage</span><span class="p">[</span><span class="s">&#39;resources&#39;</span><span class="p">]:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">cls</span><span class="o">.</span><span class="n">_filename_from_resource</span><span class="p">(</span><span class="n">resource</span><span class="p">,</span> <span class="n">datapackage_dir</span><span class="p">,</span>
                                                   <span class="n">datapackage_name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">filename</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="c"># This is supplemental data</span>
                <span class="k">for</span> <span class="n">supplemental</span> <span class="ow">in</span> <span class="n">resource</span><span class="p">[</span><span class="s">u&#39;resources&#39;</span><span class="p">]:</span>
                    <span class="n">filename</span> <span class="o">=</span> <span class="n">cls</span><span class="o">.</span><span class="n">_filename_from_resource</span><span class="p">(</span><span class="n">supplemental</span><span class="p">,</span>
                                                           <span class="n">datapackage_dir</span><span class="p">,</span>
                                                           <span class="n">datapackage_name</span><span class="p">)</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">supplemental</span><span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">]</span>

                    <span class="n">reader</span> <span class="o">=</span> <span class="n">cls</span><span class="o">.</span><span class="n">readers</span><span class="p">[</span><span class="n">supplemental</span><span class="p">[</span><span class="s">&#39;format&#39;</span><span class="p">]]</span>
                    <span class="n">compression</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;compression&#39;</span> <span class="ow">not</span> <span class="ow">in</span> \
                                          <span class="n">supplemental</span> <span class="k">else</span> \
                        <span class="n">supplemental</span><span class="p">[</span><span class="s">&#39;compression&#39;</span><span class="p">]</span>
                    <span class="n">header</span> <span class="o">=</span> <span class="n">supplemental</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;header&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                    <span class="n">index_col</span> <span class="o">=</span> <span class="n">supplemental</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;index_col&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">reader</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">,</span>
                                <span class="n">header</span><span class="o">=</span><span class="n">header</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="n">index_col</span><span class="p">)</span>
                    <span class="n">supplemental_data</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>
            <span class="k">else</span><span class="p">:</span>

                <span class="n">name</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">]</span>

                <span class="n">reader</span> <span class="o">=</span> <span class="n">cls</span><span class="o">.</span><span class="n">readers</span><span class="p">[</span><span class="n">resource</span><span class="p">[</span><span class="s">&#39;format&#39;</span><span class="p">]]</span>
                <span class="n">compression</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;compression&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">resource</span> <span class="k">else</span> \
                    <span class="n">resource</span><span class="p">[</span><span class="s">&#39;compression&#39;</span><span class="p">]</span>
                <span class="n">header</span> <span class="o">=</span> <span class="n">resource</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;header&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
                <span class="n">index_col</span> <span class="o">=</span> <span class="n">resource</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;index_col&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

                <span class="n">dfs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">reader</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">,</span>
                                   <span class="n">header</span><span class="o">=</span><span class="n">header</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="n">index_col</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">resource</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
                        <span class="n">DATAPACKAGE_RESOURCE_COMMON_KWS</span><span class="p">):</span>
                    <span class="n">kwargs</span><span class="p">[</span><span class="s">&#39;{}_{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">key</span><span class="p">)]</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>

        <span class="n">species_kws</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">species</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;species&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">datapackage</span> <span class="k">else</span> <span class="n">datapackage</span><span class="p">[</span>
            <span class="s">&#39;species&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">load_species_data</span> <span class="ow">and</span> <span class="n">species</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">species_kws</span> <span class="o">=</span> <span class="n">cls</span><span class="o">.</span><span class="n">load_species_data</span><span class="p">(</span><span class="n">species</span><span class="p">,</span> <span class="n">cls</span><span class="o">.</span><span class="n">readers</span><span class="p">,</span>
                                                <span class="n">species_datapackage_base_url</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">dfs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;metadata&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span><span class="s">&#39;The datapackage.json file is required to &#39;</span>
                                 <span class="s">&#39;have the &quot;metadata&quot; resource&#39;</span><span class="p">)</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="s">&#39;{}_data&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">),</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">dfs</span><span class="o">.</span><span class="n">iteritems</span><span class="p">())</span>

        <span class="n">nones</span> <span class="o">=</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span> <span class="k">if</span> <span class="n">v</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">nones</span><span class="p">:</span>
            <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">species_kws</span><span class="p">)</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">dfs</span><span class="p">)</span>

        <span class="n">license</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;license&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">datapackage</span> <span class="k">else</span> <span class="n">datapackage</span><span class="p">[</span>
            <span class="s">&#39;license&#39;</span><span class="p">]</span>
        <span class="n">title</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;title&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">datapackage</span> <span class="k">else</span> <span class="n">datapackage</span><span class="p">[</span>
            <span class="s">&#39;title&#39;</span><span class="p">]</span>
        <span class="n">sources</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;sources&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">datapackage</span> <span class="k">else</span> <span class="n">datapackage</span><span class="p">[</span>
            <span class="s">&#39;sources&#39;</span><span class="p">]</span>
        <span class="n">version</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;datapackage_version&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">datapackage</span> <span class="k">else</span> \
            <span class="n">datapackage</span><span class="p">[</span><span class="s">&#39;datapackage_version&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">semantic_version</span><span class="o">.</span><span class="n">validate</span><span class="p">(</span><span class="n">version</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s">&#39;{} is not a valid version string. Please use semantic &#39;</span>
                <span class="s">&#39;versioning, with major.minor.patch, e.g. 0.1.2 is a valid &#39;</span>
                <span class="s">&#39;version string&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">))</span>
        <span class="n">study</span> <span class="o">=</span> <span class="n">Study</span><span class="p">(</span><span class="n">sample_metadata</span><span class="o">=</span><span class="n">sample_metadata</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">,</span>
                      <span class="n">license</span><span class="o">=</span><span class="n">license</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="n">sources</span><span class="p">,</span>
                      <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">,</span> <span class="n">supplemental_data</span><span class="o">=</span><span class="n">supplemental_data</span><span class="p">,</span>
                      <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">study</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Study.load_species_data"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.load_species_data">[docs]</a>    <span class="k">def</span> <span class="nf">load_species_data</span><span class="p">(</span>
            <span class="n">species</span><span class="p">,</span> <span class="n">readers</span><span class="p">,</span>
            <span class="n">species_datapackage_base_url</span><span class="o">=</span><span class="n">SPECIES_DATA_PACKAGE_BASE_URL</span><span class="p">):</span>
        <span class="n">dfs</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">species_data_url</span> <span class="o">=</span> <span class="s">&#39;{}/{}/datapackage.json&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">species_datapackage_base_url</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
            <span class="n">species_datapackage</span> <span class="o">=</span> <span class="n">datapackage_url_to_dict</span><span class="p">(</span>
                <span class="n">species_data_url</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">resource</span> <span class="ow">in</span> <span class="n">species_datapackage</span><span class="p">[</span><span class="s">&#39;resources&#39;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="s">&#39;url&#39;</span> <span class="ow">in</span> <span class="n">resource</span><span class="p">:</span>
                    <span class="n">resource_url</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;url&#39;</span><span class="p">]</span>
                    <span class="n">filename</span> <span class="o">=</span> <span class="n">check_if_already_downloaded</span><span class="p">(</span><span class="n">resource_url</span><span class="p">,</span>
                                                           <span class="n">species</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">filename</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;path&#39;</span><span class="p">]</span>

                <span class="n">reader</span> <span class="o">=</span> <span class="n">readers</span><span class="p">[</span><span class="n">resource</span><span class="p">[</span><span class="s">&#39;format&#39;</span><span class="p">]]</span>

                <span class="n">compression</span> <span class="o">=</span> <span class="bp">None</span> <span class="k">if</span> <span class="s">&#39;compression&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">resource</span> <span class="k">else</span> \
                    <span class="n">resource</span><span class="p">[</span><span class="s">&#39;compression&#39;</span><span class="p">]</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="s">&#39;name&#39;</span><span class="p">]</span>
                <span class="n">dfs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">reader</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                   <span class="n">compression</span><span class="o">=</span><span class="n">compression</span><span class="p">)</span>
                <span class="n">other_keys</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">resource</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span>
                    <span class="n">DATAPACKAGE_RESOURCE_COMMON_KWS</span><span class="p">)</span>
                <span class="n">name_no_data</span> <span class="o">=</span> <span class="n">name</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&#39;_data&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">other_keys</span><span class="p">:</span>
                    <span class="n">new_key</span> <span class="o">=</span> <span class="s">&#39;{}_{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">name_no_data</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span>
                    <span class="n">dfs</span><span class="p">[</span><span class="n">new_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">resource</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">IOError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">)</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Error loading species {} data:&#39;</span>
                             <span class="s">&#39; {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">species</span><span class="p">,</span> <span class="n">e</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">dfs</span>
</div>
<div class="viewcode-block" id="Study.detect_outliers"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.detect_outliers">[docs]</a>    <span class="k">def</span> <span class="nf">detect_outliers</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span>
                        <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">featurewise</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                        <span class="n">reducer</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">standardize</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">reducer_kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">bins</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">outlier_detection_method</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">outlier_detection_method_kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">sample_subset</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">sample_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_sample_subset</span>

        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">feature_subset</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">feature_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_subset</span>

        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;expression&quot;</span><span class="p">:</span>
            <span class="n">datamodel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;splicing&quot;</span><span class="p">:</span>
            <span class="n">datamodel</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s">&#39;{} not a supported data type&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">data_type</span><span class="p">))</span>

        <span class="n">reducer</span><span class="p">,</span> <span class="n">outlier_detector</span> <span class="o">=</span> <span class="n">datamodel</span><span class="o">.</span><span class="n">detect_outliers</span><span class="p">(</span>
            <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
            <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span> <span class="n">reducer</span><span class="o">=</span><span class="n">reducer</span><span class="p">,</span> <span class="n">standardize</span><span class="o">=</span><span class="n">standardize</span><span class="p">,</span>
            <span class="n">reducer_kwargs</span><span class="o">=</span><span class="n">reducer_kwargs</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span>
            <span class="n">outlier_detection_method</span><span class="o">=</span><span class="n">outlier_detection_method</span><span class="p">,</span>
            <span class="n">outlier_detection_method_kwargs</span><span class="o">=</span><span class="n">outlier_detection_method_kwargs</span><span class="p">)</span>

        <span class="n">outlier_detector</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">reducer</span><span class="o">.</span><span class="n">reduced_space</span><span class="p">)</span>
        <span class="n">outlier_detector</span><span class="o">.</span><span class="n">title</span> <span class="o">=</span> <span class="s">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="s">&#39;outlier&#39;</span><span class="p">,</span> <span class="n">data_type</span><span class="p">,</span> <span class="n">sample_subset</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">])</span>
        <span class="k">print</span> <span class="s">&quot;setting outlier type:</span><span class="se">\&quot;</span><span class="s">{}</span><span class="se">\&quot;</span><span class="s"> in metadata&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">outlier_detector</span><span class="o">.</span><span class="n">title</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">outlier_detector</span><span class="o">.</span><span class="n">title</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">outlier_detector</span><span class="o">.</span><span class="n">title</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">outlier_detector</span><span class="o">.</span><span class="n">title</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">outlier_detector</span><span class="o">.</span><span class="n">outliers</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">reducer</span><span class="p">,</span> <span class="n">outlier_detector</span>
</div>
<div class="viewcode-block" id="Study.drop_outliers"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.drop_outliers">[docs]</a>    <span class="k">def</span> <span class="nf">drop_outliers</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Assign samples marked as &quot;outlier&quot; in metadata, to other datas&quot;&quot;&quot;</span>
        <span class="n">outliers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;outlier&#39;</span><span class="p">][</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s">&#39;outlier&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">index</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">outlier_samples</span> <span class="o">=</span> <span class="n">outliers</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">outlier_samples</span> <span class="o">=</span> <span class="n">outliers</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Study.maybe_make_directory"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.maybe_make_directory">[docs]</a>    <span class="k">def</span> <span class="nf">maybe_make_directory</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="c"># Make the directory if it&#39;s not already there</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">directory</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
            <span class="k">pass</span>
</div>
<div class="viewcode-block" id="Study.feature_subset_to_feature_ids"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.feature_subset_to_feature_ids">[docs]</a>    <span class="k">def</span> <span class="nf">feature_subset_to_feature_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                      <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Given a name of a feature subset, get the associated feature ids</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data_type : str</span>
<span class="sd">            A string describing the data type, e.g. &quot;expression&quot;</span>
<span class="sd">        feature_subset : str</span>
<span class="sd">            A string describing the subset of data type (must be already</span>
<span class="sd">            calculated)</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        feature_ids : list of strings</span>
<span class="sd">            List of features ids from the specified datatype</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s">&#39;expression&#39;</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">data_type</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="p">)</span>
        <span class="k">elif</span> <span class="s">&#39;splicing&#39;</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">data_type</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.sample_subset_to_sample_ids"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.sample_subset_to_sample_ids">[docs]</a>    <span class="k">def</span> <span class="nf">sample_subset_to_sample_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotype_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Convert a string naming a subset of phenotypes in the data into</span>
<span class="sd">        sample ids</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        phenotype_subset : str</span>
<span class="sd">            A valid string describing a boolean phenotype described in the</span>
<span class="sd">            metadata data</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        sample_ids : list of strings</span>
<span class="sd">            List of sample ids in the data</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c"># IF this is a list of IDs</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_subsets</span><span class="p">[</span><span class="n">phenotype_subset</span><span class="p">]</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">KeyError</span><span class="p">,</span> <span class="ne">TypeError</span><span class="p">):</span>
            <span class="k">pass</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span> <span class="o">==</span> <span class="n">phenotype_subset</span>
            <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">phenotype_subset</span> <span class="ow">is</span> <span class="bp">None</span> <span class="ow">or</span> <span class="s">&#39;all_samples&#39;</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span>
                    <span class="n">phenotype_subset</span><span class="p">):</span>
                <span class="n">sample_ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                     <span class="n">dtype</span><span class="o">=</span><span class="nb">bool</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">phenotype_subset</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&quot;~&quot;</span><span class="p">):</span>
                <span class="n">sample_ind</span> <span class="o">=</span> <span class="o">~</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">phenotype_subset</span><span class="o">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">&quot;~&quot;</span><span class="p">)],</span>
                    <span class="n">dtype</span><span class="o">=</span><span class="s">&#39;bool&#39;</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">sample_ind</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">phenotype_subset</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s">&#39;bool&#39;</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">sample_ind</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">sample_ids</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">phenotype_subset</span>
</div>
<div class="viewcode-block" id="Study.plot_pca"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_pca">[docs]</a>    <span class="k">def</span> <span class="nf">plot_pca</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">x_pc</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">y_pc</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                 <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">title</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">,</span> <span class="n">featurewise</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">plot_violins</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">reduce_kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                 <span class="n">color_samples_by</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">bokeh</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">most_variant_features</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">std_multiplier</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                 <span class="n">scale_by_variance</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                 <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Performs DataFramePCA on both expression and splicing study_data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data_type : str</span>
<span class="sd">            One of the names of the data types, e.g. &quot;expression&quot; or</span>
<span class="sd">            &quot;splicing&quot; (default &quot;expression&quot;)</span>
<span class="sd">        x_pc : int, optional</span>
<span class="sd">            Which principal component to plot on the x-axis (default 1)</span>
<span class="sd">        y_pc : int, optional</span>
<span class="sd">            Which principal component to plot on the y-axis (default 2)</span>
<span class="sd">        sample_subset : str or None</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used. (default None)</span>
<span class="sd">        feature_subset : str or None</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used. (default None)</span>
<span class="sd">        title : str, optional</span>
<span class="sd">            Title of the reduced space plot (default &#39;&#39;)</span>
<span class="sd">        featurewise : bool, optional</span>
<span class="sd">            If True, the features are reduced on the samples, and the plotted</span>
<span class="sd">            points are features, not samples. (default False)</span>
<span class="sd">        plot_violins : bool</span>
<span class="sd">            Whether or not to make the violinplots of the top features. This</span>
<span class="sd">            can take a long time, so to save time you can turn it off if you</span>
<span class="sd">            just want a quick look at the PCA. (default False)</span>
<span class="sd">        show_point_labels : bool, optional</span>
<span class="sd">            Whether or not to show the labels of the points. If this is</span>
<span class="sd">            samplewise (default), then this labels the samples. If this is</span>
<span class="sd">            featurewise, then this labels the features. (default False)</span>
<span class="sd">        reduce_kwargs : dict, optional</span>
<span class="sd">            Keyword arguments to the reducer (default None)</span>
<span class="sd">        color_samples_by : str, optional</span>
<span class="sd">            Instead of coloring the samples by their phenotype, color them by</span>
<span class="sd">            this column in the metadata. (default None)</span>
<span class="sd">        bokeh : bool, optional</span>
<span class="sd">            If True, plot a javascripty/interactive bokeh plot instead of a</span>
<span class="sd">            static printable figure (default False)</span>
<span class="sd">        most_variant_features : bool, optional</span>
<span class="sd">            If True, then only take the most variant of the provided features.</span>
<span class="sd">            The most variant are determined by taking the features whose</span>
<span class="sd">            variance is ``std_multiplier``standard deviations away from the</span>
<span class="sd">            mean feature variance (default False)</span>
<span class="sd">        std_multiplier : float, optional</span>
<span class="sd">            If ``most_variant_features`` is True, then use this as a cutoff</span>
<span class="sd">            for the minimum variance of a feature to be included (default 2)</span>
<span class="sd">        scale_by_variance : bool, optional</span>
<span class="sd">            If True, then scale the x- and y-axes by the explained variance</span>
<span class="sd">            ratio of the principal component dimensions. Only valid for PCA</span>
<span class="sd">            and its variations, not for NMF or tSNE. (default True)</span>
<span class="sd">        kwargs : other keyword arguments</span>
<span class="sd">            All other keyword arguments are passed to</span>
<span class="sd">            :py:meth:`DecomopsitionViz.plot`</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sample_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_sample_subset</span> \
            <span class="k">if</span> <span class="n">sample_subset</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">sample_subset</span>
        <span class="n">feature_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_subset</span> \
            <span class="k">if</span> <span class="n">feature_subset</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">feature_subset</span>

        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="n">label_to_color</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">label_to_marker</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">groupby</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">order</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">color_samples_by_phenotype</span> <span class="o">=</span> <span class="n">color_samples_by</span> \
            <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_col</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">featurewise</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">color_samples_by</span> <span class="ow">is</span> <span class="bp">None</span> <span class="ow">or</span> <span class="n">color_samples_by_phenotype</span><span class="p">:</span>
                <span class="n">label_to_color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span>
                <span class="n">label_to_marker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span>
                <span class="n">groupby</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span>
                <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">groupby</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">color_samples_by</span><span class="p">]</span>

        <span class="k">if</span> <span class="s">&quot;expression&quot;</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">data_type</span><span class="p">):</span>
            <span class="n">reducer</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span>
                <span class="n">x_pc</span><span class="o">=</span><span class="n">x_pc</span><span class="p">,</span> <span class="n">y_pc</span><span class="o">=</span><span class="n">y_pc</span><span class="p">,</span> <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span>
                <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span> <span class="n">std_multiplier</span><span class="o">=</span><span class="n">std_multiplier</span><span class="p">,</span>
                <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span> <span class="n">show_point_labels</span><span class="o">=</span><span class="n">show_point_labels</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">reduce_kwargs</span><span class="o">=</span><span class="n">reduce_kwargs</span><span class="p">,</span>
                <span class="n">plot_violins</span><span class="o">=</span><span class="n">plot_violins</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">,</span>
                <span class="n">bokeh</span><span class="o">=</span><span class="n">bokeh</span><span class="p">,</span> <span class="n">most_variant_features</span><span class="o">=</span><span class="n">most_variant_features</span><span class="p">,</span>
                <span class="n">scale_by_variance</span><span class="o">=</span><span class="n">scale_by_variance</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">elif</span> <span class="s">&quot;splicing&quot;</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">data_type</span><span class="p">):</span>
            <span class="n">reducer</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span>
                <span class="n">x_pc</span><span class="o">=</span><span class="n">x_pc</span><span class="p">,</span> <span class="n">y_pc</span><span class="o">=</span><span class="n">y_pc</span><span class="p">,</span> <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span>
                <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span> <span class="n">std_multiplier</span><span class="o">=</span><span class="n">std_multiplier</span><span class="p">,</span>
                <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span> <span class="n">show_point_labels</span><span class="o">=</span><span class="n">show_point_labels</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span> <span class="n">reduce_kwargs</span><span class="o">=</span><span class="n">reduce_kwargs</span><span class="p">,</span>
                <span class="n">plot_violins</span><span class="o">=</span><span class="n">plot_violins</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">,</span>
                <span class="n">bokeh</span><span class="o">=</span><span class="n">bokeh</span><span class="p">,</span> <span class="n">most_variant_features</span><span class="o">=</span><span class="n">most_variant_features</span><span class="p">,</span>
                <span class="n">scale_by_variance</span><span class="o">=</span><span class="n">scale_by_variance</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;The data type {} does not exist in this study&#39;</span>
                             <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">data_type</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">reducer</span>
</div>
<div class="viewcode-block" id="Study.plot_graph"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_graph">[docs]</a>    <span class="k">def</span> <span class="nf">plot_graph</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                   <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">featurewise</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                   <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot the graph (network) of these data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data_type : str</span>
<span class="sd">            One of the names of the data types, e.g. &quot;expression&quot; or &quot;splicing&quot;</span>
<span class="sd">        sample_subset : str or None</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">featurewise</span><span class="p">:</span>
            <span class="n">label_to_color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span>
            <span class="n">label_to_marker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span>
            <span class="n">groupby</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">label_to_color</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">label_to_marker</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">groupby</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;expression&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">networks</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;splicing&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">networks</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_classifier"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_classifier">[docs]</a>    <span class="k">def</span> <span class="nf">plot_classifier</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">trait</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">feature_subset</span><span class="o">=</span><span class="s">&#39;all_genes&#39;</span><span class="p">,</span>
                        <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">,</span>
                        <span class="n">show_point_labels</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                        <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot a predictor for the specified data type and trait(s)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        data_type : str</span>
<span class="sd">            One of the names of the data types, e.g. &quot;expression&quot; or &quot;splicing&quot;</span>
<span class="sd">        trait : str</span>
<span class="sd">            Column name in the metadata data that you would like</span>
<span class="sd">            to classify on</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">trait_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="n">trait</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="n">trait_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">trait</span><span class="p">)</span>
            <span class="n">trait_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">trait_ids</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">trait_data</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
            <span class="n">all_true</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">trait_data</span><span class="p">)</span>
            <span class="n">all_false</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="o">~</span><span class="n">trait_data</span><span class="p">)</span>
            <span class="n">too_few_categories</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">all_false</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="n">all_true</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="n">too_few_categories</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">trait_data</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="mi">1</span>
        <span class="n">nothing_to_classify</span> <span class="o">=</span> <span class="n">all_true</span> <span class="ow">or</span> <span class="n">all_false</span> <span class="ow">or</span> <span class="n">too_few_categories</span>

        <span class="k">if</span> <span class="n">nothing_to_classify</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;All samples are True (or all samples are &quot;</span>
                             <span class="s">&quot;False) or all are the same, cannot classify&quot;</span>
                             <span class="s">&quot;when all samples are the same&quot;</span><span class="p">)</span>
        <span class="n">trait_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">trait_data</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">trait</span><span class="p">,</span>
                               <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">feature_subset</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span> <span class="k">if</span> <span class="n">feature_subset</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">feature_subset</span>
        <span class="n">sample_subset</span> <span class="o">=</span> <span class="s">&#39;none&#39;</span> <span class="k">if</span> <span class="n">sample_subset</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">sample_subset</span>
        <span class="n">data_name</span> <span class="o">=</span> <span class="s">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">sample_subset</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">])</span>

        <span class="n">label_to_color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span>
        <span class="n">label_to_marker</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span>
        <span class="n">groupby</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span>

        <span class="n">order</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span>
        <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_color_ordered</span>

        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;expression&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_classifier</span><span class="p">(</span>
                <span class="n">data_name</span><span class="o">=</span><span class="n">data_name</span><span class="p">,</span> <span class="n">trait</span><span class="o">=</span><span class="n">trait_data</span><span class="p">,</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">show_point_labels</span><span class="o">=</span><span class="n">show_point_labels</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
                <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;splicing&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_classifier</span><span class="p">(</span>
                <span class="n">data_name</span><span class="o">=</span><span class="n">data_name</span><span class="p">,</span> <span class="n">trait</span><span class="o">=</span><span class="n">trait_data</span><span class="p">,</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="n">label_to_color</span><span class="p">,</span>
                <span class="n">label_to_marker</span><span class="o">=</span><span class="n">label_to_marker</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">groupby</span><span class="p">,</span>
                <span class="n">show_point_labels</span><span class="o">=</span><span class="n">show_point_labels</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">title</span><span class="p">,</span>
                <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.modality_assignments"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.modality_assignments">[docs]</a>    <span class="k">def</span> <span class="nf">modality_assignments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                             <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get modality assignments of splicing data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_subset : str or None, optional</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None, optional</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        expression_thresh : float, optional</span>
<span class="sd">            Minimum expression value, of the original input. E.g. if the</span>
<span class="sd">            original input is already log-transformed, then this threshold is</span>
<span class="sd">            on the log values.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        modalities : pandas.DataFrame</span>
<span class="sd">            A (n_phenotypes, n_events) shaped DataFrame of the assigned</span>
<span class="sd">            modality</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">min_expression</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="ow">and</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="n">min_expression</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">modality_assignments</span><span class="p">(</span>
            <span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
            <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="n">min_samples</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.modality_counts"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.modality_counts">[docs]</a>    <span class="k">def</span> <span class="nf">modality_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get number of splicing events in modality categories</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_subset : str or None, optional</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None, optional</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        expression_thresh : float, optional</span>
<span class="sd">            Minimum expression value, of the original input. E.g. if the</span>
<span class="sd">            original input is already log-transformed, then this threshold is</span>
<span class="sd">            on the log values.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        modalities : pandas.DataFrame</span>
<span class="sd">            A (n_phenotypes, n_modalities) shaped DataFrame of the number of</span>
<span class="sd">            events assigned to each modality</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">min_expression</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span> <span class="ow">and</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="n">min_expression</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">modality_assignments</span><span class="p">(</span>
            <span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
            <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="n">min_samples</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_modalities_bars"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_modalities_bars">[docs]</a>    <span class="k">def</span> <span class="nf">plot_modalities_bars</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                             <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">percentages</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Make grouped barplots of the number of modalities per phenotype</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_subset : str or None</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        expression_thresh : float</span>
<span class="sd">            If greater than -inf, then filter on splicing events in genes</span>
<span class="sd">            with expression at least this value</span>
<span class="sd">        percentages : bool</span>
<span class="sd">            If True, plot percentages instead of counts</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_modalities_bars</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                                           <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
                                           <span class="n">percentages</span><span class="o">=</span><span class="n">percentages</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_modalities_reduced"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_modalities_reduced">[docs]</a>    <span class="k">def</span> <span class="nf">plot_modalities_reduced</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot splicing events with modality assignments in NMF space</span>

<span class="sd">        This will plot a separate NMF space for each celltype in the data, as</span>
<span class="sd">        well as one for all samples.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_subset : str or None</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        expression_thresh : float</span>
<span class="sd">            If greater than -inf, then filter on splicing events in genes</span>
<span class="sd">            with expression at least this value</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="n">grouped</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">)</span>

        <span class="c"># Account for bar plot and plot of the reduced space of ALL samples</span>
        <span class="n">n</span> <span class="o">=</span> <span class="n">grouped</span><span class="o">.</span><span class="n">ngroups</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;all&#39;</span><span class="p">]</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="n">n</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
        <span class="n">all_ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_modalities_reduced</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span>
                                              <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                                              <span class="n">ax</span><span class="o">=</span><span class="n">all_ax</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">&#39;all samples&#39;</span><span class="p">)</span>
        <span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">((</span><span class="n">celltype</span><span class="p">,</span> <span class="n">series</span><span class="p">),</span> <span class="n">ax</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">grouped</span><span class="p">,</span> <span class="n">axes</span><span class="p">)):</span>
            <span class="n">groups</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">celltype</span><span class="p">)</span>
            <span class="c"># sys.stdout.write(&#39;\n---- {} ----\n&#39;.format(celltype))</span>
            <span class="n">samples</span> <span class="o">=</span> <span class="n">series</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
            <span class="c"># legend = i == 0</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_modalities_reduced</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span>
                                                  <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                                                  <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">celltype</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.celltype_sizes"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.celltype_sizes">[docs]</a>    <span class="k">def</span> <span class="nf">celltype_sizes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;splicing&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;expression&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                                         <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                                       <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</div>
    <span class="nd">@property</span>
<div class="viewcode-block" id="Study.celltype_event_counts"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.celltype_event_counts">[docs]</a>    <span class="k">def</span> <span class="nf">celltype_event_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Number of cells that detected each event, per celltype</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">count</span><span class="p">())</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.unique_celltype_event_counts"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.unique_celltype_event_counts">[docs]</a>    <span class="k">def</span> <span class="nf">unique_celltype_event_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">celltype_event_counts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">celltype_event_counts</span>
        <span class="k">return</span> <span class="n">celltype_event_counts</span><span class="p">[</span><span class="n">celltype_event_counts</span> <span class="o">&lt;=</span> <span class="n">n</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="Study.percent_unique_celltype_events"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.percent_unique_celltype_events">[docs]</a>    <span class="k">def</span> <span class="nf">percent_unique_celltype_events</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">n_unique</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unique_celltype_event_counts</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">n_total</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">celltype_event_counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">n_unique</span> <span class="o">/</span> <span class="n">n_total</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="c"># @property</span>
    <span class="c"># def celltype_modalities(self):</span>
    <span class="c"># &quot;&quot;&quot;Return modality assignments of each celltype</span>
    <span class="c">#     &quot;&quot;&quot;</span>
    <span class="c">#     return self.splicing.data.groupby(</span>
    <span class="c">#         self.sample_id_to_phenotype, axis=0).apply(</span>
    <span class="c">#         lambda x: self.splicing.modalities(x.index))</span>
</div>
<div class="viewcode-block" id="Study.plot_modalities_lavalamps"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_modalities_lavalamps">[docs]</a>    <span class="k">def</span> <span class="nf">plot_modalities_lavalamps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                  <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                  <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot each modality in each celltype on a separate axes</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample_subset : str or None</span>
<span class="sd">            Which subset of the samples to use, based on some phenotype</span>
<span class="sd">            column in the experiment design data. If None, all samples are</span>
<span class="sd">            used.</span>
<span class="sd">        feature_subset : str or None</span>
<span class="sd">            Which subset of the features to used, based on some feature type</span>
<span class="sd">            in the expression data (e.g. &quot;variant&quot;). If None, all features</span>
<span class="sd">            are used.</span>
<span class="sd">        expression_thresh : float</span>
<span class="sd">            If greater than -inf, then filter on splicing events in genes</span>
<span class="sd">            with expression at least this value</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_modalities_lavalamps</span><span class="p">(</span>
            <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
            <span class="n">phenotype_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
            <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_event_modality_estimation"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_event_modality_estimation">[docs]</a>    <span class="k">def</span> <span class="nf">plot_event_modality_estimation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">event_id</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                                       <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_event_modality_estimation</span><span class="p">(</span>
            <span class="n">event_id</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
            <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_event"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_event">[docs]</a>    <span class="k">def</span> <span class="nf">plot_event</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">nmf_space</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot the violinplot and NMF transitions of a splicing event</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_feature</span><span class="p">(</span>
            <span class="n">feature_id</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span>
            <span class="n">phenotype_groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
            <span class="n">phenotype_order</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_color_ordered</span><span class="p">,</span>
            <span class="n">phenotype_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
            <span class="n">phenotype_to_marker</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span><span class="p">,</span> <span class="n">nmf_space</span><span class="o">=</span><span class="n">nmf_space</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_gene"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_gene">[docs]</a>    <span class="k">def</span> <span class="nf">plot_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">nmf_space</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_feature</span><span class="p">(</span>
            <span class="n">feature_id</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span>
            <span class="n">phenotype_groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
            <span class="n">phenotype_order</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_color_ordered</span><span class="p">,</span>
            <span class="n">phenotype_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
            <span class="n">phenotype_to_marker</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span><span class="p">,</span> <span class="n">nmf_space</span><span class="o">=</span><span class="n">nmf_space</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_lavalamp_pooled_inconsistent"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_lavalamp_pooled_inconsistent">[docs]</a>    <span class="k">def</span> <span class="nf">plot_lavalamp_pooled_inconsistent</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
            <span class="n">fraction_diff_thresh</span><span class="o">=</span><span class="n">FRACTION_DIFF_THRESH</span><span class="p">,</span>
            <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="c"># grouped_ids = self.splicing.data.groupby(self.sample_id_to_color,</span>
        <span class="c"># axis=0)</span>
        <span class="n">celltype_groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sample_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="c"># Only plotting one sample_subset</span>
            <span class="n">celltype_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">celltype_groups</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">sample_subset</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># Plotting all the celltypes</span>
            <span class="n">celltype_samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>

        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
            <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="n">feature_subset</span><span class="p">)</span>

        <span class="n">celltype_and_sample_ids</span> <span class="o">=</span> <span class="n">celltype_groups</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">celltype_and_sample_ids</span><span class="p">):</span>
            <span class="c"># import pdb; pdb.set_trace()</span>

            <span class="c"># Assumes all samples of a sample_subset have the same color...</span>
            <span class="c"># probably wrong</span>
            <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">[</span><span class="n">phenotype</span><span class="p">]</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="n">celltype_samples</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span><span class="n">expression_thresh</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">sample_ids</span><span class="p">,</span> <span class="p">:]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_lavalamp_pooled_inconsistent</span><span class="p">(</span>
                <span class="n">data</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span> <span class="n">fraction_diff_thresh</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.percent_pooled_inconsistent"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.percent_pooled_inconsistent">[docs]</a>    <span class="k">def</span> <span class="nf">percent_pooled_inconsistent</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
            <span class="n">fraction_diff_thresh</span><span class="o">=</span><span class="n">FRACTION_DIFF_THRESH</span><span class="p">,</span>
            <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="n">celltype_groups</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">sample_subset</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="c"># Only plotting one sample_subset</span>
            <span class="n">celltype_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">celltype_groups</span><span class="o">.</span><span class="n">groups</span><span class="p">[</span><span class="n">sample_subset</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># Plotting all the celltypes</span>
            <span class="n">celltype_samples</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>

        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
            <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="n">feature_subset</span><span class="p">)</span>

        <span class="n">celltype_and_sample_ids</span> <span class="o">=</span> <span class="n">celltype_groups</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">([</span><span class="n">celltype_groups</span><span class="o">.</span><span class="n">groups</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>
                                            <span class="p">[</span><span class="s">&#39;n_events&#39;</span><span class="p">,</span> <span class="s">&#39;percent&#39;</span><span class="p">]])</span>
        <span class="n">percents</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">celltype_and_sample_ids</span><span class="p">):</span>
            <span class="c"># import pdb; pdb.set_trace()</span>

            <span class="c"># Assumes all samples of a sample_subset have the same color...</span>
            <span class="c"># probably wrong</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="n">celltype_samples</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span><span class="n">expression_thresh</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">sample_ids</span><span class="p">,</span> <span class="p">:]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">data</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">singles</span><span class="p">,</span> <span class="n">pooled</span><span class="p">,</span> <span class="n">not_measured_in_pooled</span><span class="p">,</span> <span class="n">pooled_inconsistent</span> \
                    <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">pooled_inconsistent</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span>
                                                        <span class="n">fraction_diff_thresh</span><span class="p">)</span>
                <span class="n">percent</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">_divide_inconsistent_and_pooled</span><span class="p">(</span>
                    <span class="n">pooled</span><span class="p">,</span> <span class="n">pooled_inconsistent</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">percent</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">percents</span><span class="p">[</span><span class="n">phenotype</span><span class="p">,</span> <span class="s">&#39;percent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">percent</span>
            <span class="n">percents</span><span class="p">[</span><span class="n">phenotype</span><span class="p">,</span> <span class="s">&#39;n_events&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">percents</span>
</div>
<div class="viewcode-block" id="Study.expression_vs_inconsistent_splicing"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.expression_vs_inconsistent_splicing">[docs]</a>    <span class="k">def</span> <span class="nf">expression_vs_inconsistent_splicing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Percentage of events inconsistent with pooled at expression threshs</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        bins : list-like</span>
<span class="sd">            List of expression cutoffs</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        expression_vs_inconsistent : pd.DataFrame</span>
<span class="sd">            A (len(bins), n_phenotypes) dataframe of the percentage of events</span>
<span class="sd">            in single cells that are inconsistent with pooled</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">bins</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">emin</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data_original</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()))</span>
            <span class="n">emax</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data_original</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">()))</span>
            <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">emin</span><span class="p">,</span> <span class="n">emax</span><span class="p">)</span>

        <span class="n">expression_vs_inconsistent</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">bins</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">percent_pooled_inconsistent</span><span class="p">(</span><span class="n">expression_thresh</span><span class="o">=</span><span class="n">x</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">expression_vs_inconsistent</span>
</div>
<div class="viewcode-block" id="Study.plot_expression_vs_inconsistent_splicing"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_expression_vs_inconsistent_splicing">[docs]</a>    <span class="k">def</span> <span class="nf">plot_expression_vs_inconsistent_splicing</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

        <span class="n">expression_vs_inconsistent</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression_vs_inconsistent_splicing</span><span class="p">(</span>
            <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

        <span class="c"># Plot the percent inconsistent</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">phenotype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">expression_vs_inconsistent</span><span class="p">[(</span><span class="n">phenotype</span><span class="p">,</span> <span class="s">&#39;percent&#39;</span><span class="p">)]</span>
            <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">[</span><span class="n">phenotype</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="s">&#39;o-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;Expression threshold&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Percent events inconsistent with pooled&#39;</span><span class="p">)</span>
        <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>

        <span class="c"># Plot number of events at each cutoff</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">phenotype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">:</span>
            <span class="n">s</span> <span class="o">=</span> <span class="n">expression_vs_inconsistent</span><span class="p">[(</span><span class="n">phenotype</span><span class="p">,</span> <span class="s">&#39;n_events&#39;</span><span class="p">)]</span>
            <span class="n">color</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">[</span><span class="n">phenotype</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="s">&#39;o-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">&#39;Expression threshold&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;Number of events&#39;</span><span class="p">)</span>
        <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">despine</span><span class="p">()</span>
</div>
<div class="viewcode-block" id="Study.plot_clustermap"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_clustermap">[docs]</a>    <span class="k">def</span> <span class="nf">plot_clustermap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s">&#39;euclidean&#39;</span><span class="p">,</span>
                        <span class="n">method</span><span class="o">=</span><span class="s">&#39;average&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                        <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Visualize hierarchical relationships within samples and features</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">feature_subset</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">feature_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_subset</span>

        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;expression&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_clustermap</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
                <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="n">scale_fig_by_data</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;splicing&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_clustermap</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
                <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="n">scale_fig_by_data</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_correlations"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_correlations">[docs]</a>    <span class="k">def</span> <span class="nf">plot_correlations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                          <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s">&#39;euclidean&#39;</span><span class="p">,</span>
                          <span class="n">method</span><span class="o">=</span><span class="s">&#39;average&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">featurewise</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                          <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Visualize clustered correlations of samples across features</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>


<span class="sd">        Returns</span>
<span class="sd">        -------</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">feature_subset</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">feature_subset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">default_feature_subset</span>

        <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
        <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span>
                                                         <span class="n">feature_subset</span><span class="p">,</span>
                                                         <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">figsize</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">scale_fig_by_data</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;If &quot;scale_fig_by_data&quot; is true, then cannot &#39;</span>
                             <span class="s">&#39;also specify &quot;figsize&quot;&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;expression&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_correlations</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span>
                <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="n">scale_fig_by_data</span><span class="p">,</span>
                <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&quot;splicing&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_correlations</span><span class="p">(</span>
                <span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
                <span class="n">metric</span><span class="o">=</span><span class="n">metric</span><span class="p">,</span> <span class="n">sample_id_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_color</span><span class="p">,</span>
                <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">scale_fig_by_data</span><span class="o">=</span><span class="n">scale_fig_by_data</span><span class="p">,</span>
                <span class="n">featurewise</span><span class="o">=</span><span class="n">featurewise</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_lavalamps"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_lavalamps">[docs]</a>    <span class="k">def</span> <span class="nf">plot_lavalamps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">feature_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                       <span class="n">expression_thresh</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_splicing_on_expression</span><span class="p">(</span>
                <span class="n">expression_thresh</span><span class="o">=</span><span class="n">expression_thresh</span><span class="p">,</span>
                <span class="n">sample_subset</span><span class="o">=</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">feature_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">feature_subset_to_feature_ids</span><span class="p">(</span>
                <span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">feature_subset</span><span class="p">,</span> <span class="n">rename</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">data</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_lavalamp</span><span class="p">(</span><span class="n">sample_ids</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span>
                                    <span class="n">feature_ids</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                                    <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                                    <span class="n">phenotype_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
                                    <span class="n">order</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_big_nmf_space_transitions"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_big_nmf_space_transitions">[docs]</a>    <span class="k">def</span> <span class="nf">plot_big_nmf_space_transitions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;expression&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_big_nmf_space_transitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_transitions</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_color_ordered</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_big_nmf_space_transitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_transitions</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_color_ordered</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_marker</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_two_samples"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_two_samples">[docs]</a>    <span class="k">def</span> <span class="nf">plot_two_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span>
                         <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Plot a scatterplot of two samples&#39; data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sample1 : str</span>
<span class="sd">            Name of the sample to plot on the x-axis</span>
<span class="sd">        sample2 : str</span>
<span class="sd">            Name of the sample to plot on the y-axis</span>
<span class="sd">        data_type : &quot;expression&quot; | &quot;splicing&quot;</span>
<span class="sd">            Type of data to plot. Default &quot;expression&quot;</span>
<span class="sd">        Any other keyword arguments valid for seaborn.jointplot</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        jointgrid : seaborn.axisgrid.JointGrid</span>
<span class="sd">            Returns a JointGrid instance</span>

<span class="sd">        See Also</span>
<span class="sd">        -------</span>
<span class="sd">        seaborn.jointplot</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;expression&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_two_samples</span><span class="p">(</span><span class="n">sample1</span><span class="p">,</span> <span class="n">sample2</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.plot_two_features"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.plot_two_features">[docs]</a>    <span class="k">def</span> <span class="nf">plot_two_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature1</span><span class="p">,</span> <span class="n">feature2</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span>
                          <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Make a scatterplot of two features&#39; data</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        feature1 : str</span>
<span class="sd">            Name of the feature to plot on the x-axis. If you have a</span>
<span class="sd">            feature_data dataframe for this data type, will attempt to map</span>
<span class="sd">            the common name, e.g. &quot;RBFOX2&quot; back to the crazy name,</span>
<span class="sd">            e.g. &quot;ENSG00000100320&quot;</span>
<span class="sd">        feature2 : str</span>
<span class="sd">            Name of the feature to plot on the y-axis. If you have a</span>
<span class="sd">            feature_data dataframe for this data type, will attempt to map</span>
<span class="sd">            the common name, e.g. &quot;RBFOX2&quot; back to the crazy name,</span>
<span class="sd">            e.g. &quot;ENSG00000100320&quot;</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>


<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;expression&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plot_two_features</span><span class="p">(</span>
                <span class="n">feature1</span><span class="p">,</span> <span class="n">feature2</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">plot_two_features</span><span class="p">(</span>
                <span class="n">feature1</span><span class="p">,</span> <span class="n">feature2</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span>
                <span class="n">label_to_color</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.nmf_space_positions"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.nmf_space_positions">[docs]</a>    <span class="k">def</span> <span class="nf">nmf_space_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;splicing&#39;</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">nmf_space_positions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.nmf_space_transitions"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.nmf_space_transitions">[docs]</a>    <span class="k">def</span> <span class="nf">nmf_space_transitions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotype_transitions</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span>
                              <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The change in NMF space of splicing events across phenotypes</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        phenotype_transitions : list of length-2 tuples of str</span>
<span class="sd">            List of (&#39;phenotype1&#39;, &#39;phenotype2&#39;) transitions whose change in</span>
<span class="sd">            distribution you are interested in</span>
<span class="sd">        data_type : &#39;splicing&#39; | &#39;expression&#39;</span>
<span class="sd">            Which data type to calculate this on. (default=&#39;splicing&#39;)</span>
<span class="sd">        n : int</span>
<span class="sd">            Minimum number of samples per phenotype, per event</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        big_transitions : pandas.DataFrame</span>
<span class="sd">            A (n_events, n_transitions) dataframe of the NMF distances between</span>
<span class="sd">            splicing events</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">phenotype_transitions</span> <span class="o">==</span> <span class="s">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">phenotype_transitions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_transitions</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">nmf_space_transitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">phenotype_transitions</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.big_nmf_space_transitions"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.big_nmf_space_transitions">[docs]</a>    <span class="k">def</span> <span class="nf">big_nmf_space_transitions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phenotype_transitions</span><span class="o">=</span><span class="s">&#39;all&#39;</span><span class="p">,</span>
                                  <span class="n">data_type</span><span class="o">=</span><span class="s">&#39;splicing&#39;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Splicing events whose change in NMF space is large</span>

<span class="sd">        By large, we mean that difference is 2 standard deviations away from</span>
<span class="sd">        the mean</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        phenotype_transitions : list of length-2 tuples of str</span>
<span class="sd">            List of (&#39;phenotype1&#39;, &#39;phenotype2&#39;) transitions whose change in</span>
<span class="sd">            distribution you are interested in</span>
<span class="sd">        data_type : &#39;splicing&#39; | &#39;expression&#39;</span>
<span class="sd">            Which data type to calculate this on. (default=&#39;splicing&#39;)</span>
<span class="sd">        n : int</span>
<span class="sd">            Minimum number of samples per phenotype, per event</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        big_transitions : pandas.DataFrame</span>
<span class="sd">            A (n_events, n_transitions) dataframe of the NMF distances between</span>
<span class="sd">            splicing events</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">phenotype_transitions</span> <span class="o">==</span> <span class="s">&#39;all&#39;</span><span class="p">:</span>
            <span class="n">phenotype_transitions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_transitions</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s">&#39;splicing&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">big_nmf_space_transitions</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_id_to_phenotype</span><span class="p">,</span> <span class="n">phenotype_transitions</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Study.save"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.save">[docs]</a>    <span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_name</span><span class="p">,</span> <span class="n">flotilla_dir</span><span class="o">=</span><span class="n">FLOTILLA_DOWNLOAD_DIR</span><span class="p">):</span>

        <span class="n">metadata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">data_original</span>

        <span class="n">metadata_kws</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;pooled_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">pooled_col</span><span class="p">,</span>
                        <span class="s">&#39;phenotype_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_col</span><span class="p">,</span>
                        <span class="s">&#39;phenotype_order&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_order</span><span class="p">,</span>
                        <span class="s">&#39;phenotype_to_color&#39;</span><span class="p">:</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_to_color</span><span class="p">,</span>
                        <span class="s">&#39;phenotype_to_marker&#39;</span><span class="p">:</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">phenotype_to_marker</span><span class="p">,</span>
                        <span class="s">&#39;minimum_samples&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">minimum_sample_subset</span><span class="p">,</span>
                        <span class="s">&#39;outlier_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">outlier_col</span><span class="p">}</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">expression</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data_original</span>
            <span class="n">expression_kws</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s">&#39;log_base&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">log_base</span><span class="p">,</span>
                <span class="s">&#39;thresh&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">thresh_original</span><span class="p">,</span>
                <span class="s">&#39;plus_one&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">plus_one</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">expression</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">expression_kws</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">expression_feature_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_data</span>
            <span class="n">expression_feature_kws</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s">&#39;rename_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_rename_col</span><span class="p">,</span>
                <span class="s">&#39;ignore_subset_cols&#39;</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_ignore_subset_cols</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">expression_feature_data</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">expression_feature_kws</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">splicing</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data_original</span>
            <span class="n">splicing_kws</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">splicing</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">splicing_kws</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_data</span>
            <span class="n">splicing_feature_kws</span> <span class="o">=</span> \
                <span class="p">{</span><span class="s">&#39;rename_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_rename_col</span><span class="p">,</span>
                 <span class="s">&#39;ignore_subset_cols&#39;</span><span class="p">:</span>
                     <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_ignore_subset_cols</span><span class="p">,</span>
                 <span class="s">&#39;expression_id_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_expression_id_col</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">splicing_feature_data</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">splicing_feature_kws</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">spikein</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">spikein</span><span class="o">.</span><span class="n">data_original</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">spikein</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">gene_ontology</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_ontology</span><span class="o">.</span><span class="n">data</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">gene_ontology</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">mapping_stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span><span class="o">.</span><span class="n">data_original</span>
            <span class="n">mapping_stats_kws</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s">&#39;number_mapped_col&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span><span class="o">.</span><span class="n">number_mapped_col</span><span class="p">,</span>
                <span class="s">&#39;min_reads&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">mapping_stats</span><span class="o">.</span><span class="n">min_reads</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">mapping_stats</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">mapping_stats_kws</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="n">supplemental_attributes</span> <span class="o">=</span> <span class="n">inspect</span><span class="o">.</span><span class="n">getmembers</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">supplemental</span><span class="p">,</span>
                                                     <span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="ow">not</span>
                                                     <span class="p">(</span><span class="n">inspect</span><span class="o">.</span><span class="n">isroutine</span><span class="p">(</span><span class="n">a</span><span class="p">)))</span>
        <span class="n">supplemental_attributes</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">supplemental_attributes</span>
                                   <span class="k">if</span> <span class="ow">not</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;__&#39;</span><span class="p">)</span> <span class="ow">and</span>
                                          <span class="n">a</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;__&#39;</span><span class="p">))]</span>
        <span class="n">supplemental_kws</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">supplemental_name</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">supplemental_attributes</span><span class="p">:</span>
            <span class="n">supplemental_kws</span><span class="p">[</span><span class="n">supplemental_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span>

        <span class="c"># Increase the version number</span>
        <span class="n">version</span> <span class="o">=</span> <span class="n">semantic_version</span><span class="o">.</span><span class="n">Version</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">version</span><span class="p">)</span>
        <span class="n">version</span><span class="o">.</span><span class="n">patch</span> <span class="o">=</span> <span class="n">version</span><span class="o">.</span><span class="n">patch</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">version</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">version</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">make_study_datapackage</span><span class="p">(</span>
            <span class="n">study_name</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="n">expression</span><span class="p">,</span> <span class="n">splicing</span><span class="p">,</span>
            <span class="n">spikein</span><span class="p">,</span> <span class="n">mapping_stats</span><span class="p">,</span> <span class="n">metadata_kws</span><span class="o">=</span><span class="n">metadata_kws</span><span class="p">,</span>
            <span class="n">expression_kws</span><span class="o">=</span><span class="n">expression_kws</span><span class="p">,</span> <span class="n">splicing_kws</span><span class="o">=</span><span class="n">splicing_kws</span><span class="p">,</span>
            <span class="n">mapping_stats_kws</span><span class="o">=</span><span class="n">mapping_stats_kws</span><span class="p">,</span>
            <span class="n">expression_feature_kws</span><span class="o">=</span><span class="n">expression_feature_kws</span><span class="p">,</span>
            <span class="n">expression_feature_data</span><span class="o">=</span><span class="n">expression_feature_data</span><span class="p">,</span>
            <span class="n">splicing_feature_data</span><span class="o">=</span><span class="n">splicing_feature_data</span><span class="p">,</span>
            <span class="n">splicing_feature_kws</span><span class="o">=</span><span class="n">splicing_feature_kws</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">species</span><span class="p">,</span>
            <span class="n">license</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">license</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">title</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">sources</span><span class="p">,</span>
            <span class="n">version</span><span class="o">=</span><span class="n">version</span><span class="p">,</span> <span class="n">flotilla_dir</span><span class="o">=</span><span class="n">flotilla_dir</span><span class="p">,</span>
            <span class="n">gene_ontology</span><span class="o">=</span><span class="n">gene_ontology</span><span class="p">,</span> <span class="n">supplemental_kws</span><span class="o">=</span><span class="n">supplemental_kws</span><span class="p">)</span>
</div>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_maybe_get_axis_name</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">alt_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">alt_name</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">alt_name</span> <span class="o">=</span> <span class="s">&#39;columns&#39;</span> <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="s">&#39;index&#39;</span>
        <span class="n">axis</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">axis</span> <span class="o">==</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">axis</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="p">):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">axis</span><span class="o">.</span><span class="n">names</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">axis</span><span class="o">.</span><span class="n">name</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">alt_name</span> <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="bp">None</span> <span class="k">else</span> <span class="n">name</span>
        <span class="k">return</span> <span class="n">name</span>

    <span class="nd">@cached_property</span><span class="p">()</span>
    <span class="k">def</span> <span class="nf">tidy_splicing_with_expression</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;A tall &#39;tidy&#39; dataframe of samples with expression and splicing</span>

<span class="sd">        :return:</span>
<span class="sd">        :rtype:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># Establish common strings</span>

        <span class="n">splicing_common_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_data</span><span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">feature_expression_id_col</span><span class="p">]</span>

        <span class="c"># Tidify splicing</span>
        <span class="n">splicing</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span>
        <span class="n">splicing_index_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_maybe_get_axis_name</span><span class="p">(</span><span class="n">splicing</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">splicing_columns_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_maybe_get_axis_name</span><span class="p">(</span><span class="n">splicing</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">splicing</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
                                <span class="n">id_vars</span><span class="o">=</span><span class="n">splicing_index_name</span><span class="p">,</span>
                                <span class="n">value_name</span><span class="o">=</span><span class="s">&#39;psi&#39;</span><span class="p">,</span>
                                <span class="n">var_name</span><span class="o">=</span><span class="n">splicing_columns_name</span><span class="p">)</span>

        <span class="n">s</span> <span class="o">=</span> <span class="n">splicing_common_id</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

        <span class="n">event_name_to_ensembl_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span><span class="nb">zip</span><span class="p">([</span><span class="n">k</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)),</span> <span class="n">v</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span>
              <span class="n">s</span><span class="o">.</span><span class="n">iteritems</span><span class="p">()]))</span>
        <span class="n">index</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">event_name_to_ensembl_ids</span><span class="p">)</span>
        <span class="n">event_name_to_ensembl_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span>
                                              <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_common_id</span><span class="p">)</span>

        <span class="n">rename_columns</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">splicing_index_name</span> <span class="o">==</span> <span class="s">&#39;index&#39;</span><span class="p">:</span>
            <span class="n">rename_columns</span><span class="p">[</span><span class="n">splicing_index_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_id</span>
        <span class="k">if</span> <span class="n">splicing_columns_name</span> <span class="o">==</span> <span class="s">&#39;columns&#39;</span><span class="p">:</span>
            <span class="n">rename_columns</span><span class="p">[</span><span class="n">splicing_columns_name</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_event_name</span>
            <span class="n">splicing_columns_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_event_name</span>
        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">rename_columns</span><span class="p">)</span>

        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">splicing_columns_name</span><span class="p">)</span>
        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">event_name_to_ensembl_ids</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">event_name_to_ensembl_ids</span><span class="p">)</span>

        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="n">splicing_tidy</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;index&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_event_name</span><span class="p">})</span>

        <span class="c"># Tidify expression</span>
        <span class="n">expression</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data_original</span>
        <span class="n">expression_index_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_maybe_get_axis_name</span><span class="p">(</span><span class="n">expression</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">expression_tidy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">expression</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span>
                                  <span class="n">id_vars</span><span class="o">=</span><span class="n">expression_index_name</span><span class="p">,</span>
                                  <span class="n">value_name</span><span class="o">=</span><span class="s">&#39;expression&#39;</span><span class="p">,</span>
                                  <span class="n">var_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_common_id</span><span class="p">)</span>
        <span class="c"># This will only do anything if there is a column named &quot;index&quot; so</span>
        <span class="c"># no need to check anything</span>
        <span class="n">expression_tidy</span> <span class="o">=</span> <span class="n">expression_tidy</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;index&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_id</span><span class="p">})</span>
        <span class="n">expression_tidy</span> <span class="o">=</span> <span class="n">expression_tidy</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

        <span class="n">splicing_tidy_with_expression</span> <span class="o">=</span> <span class="n">splicing_tidy</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">expression_tidy</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_common_id</span><span class="p">],</span>
            <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_id</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_common_id</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">splicing_tidy_with_expression</span>

<div class="viewcode-block" id="Study.filter_splicing_on_expression"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.filter_splicing_on_expression">[docs]</a>    <span class="k">def</span> <span class="nf">filter_splicing_on_expression</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">expression_thresh</span><span class="p">,</span>
                                      <span class="n">sample_subset</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filter splicing events on expression values</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        expression_thresh : float</span>
<span class="sd">            Minimum expression value, of the original input. E.g. if the</span>
<span class="sd">            original input is already log-transformed, then this threshold is</span>
<span class="sd">            on the log values.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        psi : pandas.DataFrame</span>
<span class="sd">            A (n_samples, n_features)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">min_expression</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data_original</span><span class="o">.</span><span class="n">min</span><span class="p">()</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span> \
                <span class="ow">and</span> <span class="n">expression_thresh</span> <span class="o">&gt;</span> <span class="n">min_expression</span><span class="p">:</span>
            <span class="n">columns</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_maybe_get_axis_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                                <span class="n">alt_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_event_name</span><span class="p">)</span>
            <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_maybe_get_axis_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                              <span class="n">alt_name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_id</span><span class="p">)</span>

            <span class="n">sample_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample_subset_to_sample_ids</span><span class="p">(</span><span class="n">sample_subset</span><span class="p">)</span>
            <span class="n">splicing_with_expression</span> <span class="o">=</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">tidy_splicing_with_expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">tidy_splicing_with_expression</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                        <span class="n">sample_ids</span><span class="p">)]</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">splicing_with_expression</span><span class="o">.</span><span class="n">expression</span> <span class="o">&gt;=</span> <span class="n">expression_thresh</span>
            <span class="n">splicing_high_expression</span> <span class="o">=</span> <span class="n">splicing_with_expression</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">ind</span><span class="p">]</span>
            <span class="n">splicing_high_expression</span> <span class="o">=</span> \
                <span class="n">splicing_high_expression</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">columns</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                <span class="n">filtered_psi</span> <span class="o">=</span> <span class="n">splicing_high_expression</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">&#39;psi&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">filtered_psi</span> <span class="o">=</span> <span class="n">splicing_high_expression</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span>
                    <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">&#39;psi&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">filtered_psi</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">splicing</span><span class="o">.</span><span class="n">data</span>
</div>
<div class="viewcode-block" id="Study.go_enrichment"><a class="viewcode-back" href="../../../flotilla.data_model.study.html#flotilla.data_model.study.Study.go_enrichment">[docs]</a>    <span class="k">def</span> <span class="nf">go_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_ids</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">domain</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                      <span class="n">p_value_cutoff</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">min_feature_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                      <span class="n">min_background_size</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Calculate gene ontology enrichment of provided features</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        feature_ids : list-like</span>
<span class="sd">            Features to calculate gene ontology enrichment on</span>
<span class="sd">        background : list-like, optional</span>
<span class="sd">            Features to use as the background</span>
<span class="sd">        domain : str or list, optional</span>
<span class="sd">            Only calculate GO enrichment for a particular GO category or</span>
<span class="sd">            subset of categories. Valid domains:</span>
<span class="sd">            &#39;biological_process&#39;, &#39;molecular_function&#39;, &#39;cellular_component&#39;</span>
<span class="sd">        p_value_cutoff : float, optional</span>
<span class="sd">            Maximum accepted Bonferroni-corrected p-value</span>
<span class="sd">        min_feature_size : int, optional</span>
<span class="sd">            Minimum number of features of interest overlapping in a GO Term,</span>
<span class="sd">            to calculate enrichment</span>
<span class="sd">        min_background_size : int, optional</span>
<span class="sd">            Minimum number of features in the background overlapping a GO Term</span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        enrichment : pandas.DataFrame</span>
<span class="sd">            A (go_categories, columns) dataframe showing the GO</span>
<span class="sd">            enrichment categories that were enriched in the features</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">background</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s">&#39;No background provided, defaulting to all &#39;</span>
                          <span class="s">&#39;expressed genes&#39;</span><span class="p">)</span>
            <span class="n">background</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_ontology</span><span class="o">.</span><span class="n">enrichment</span><span class="p">(</span>
            <span class="n">feature_ids</span><span class="p">,</span> <span class="n">background</span><span class="o">=</span><span class="n">background</span><span class="p">,</span>
            <span class="n">cross_reference</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">expression</span><span class="o">.</span><span class="n">feature_renamer_series</span><span class="p">,</span>
            <span class="n">domain</span><span class="o">=</span><span class="n">domain</span><span class="p">,</span> <span class="n">p_value_cutoff</span><span class="o">=</span><span class="n">p_value_cutoff</span><span class="p">,</span>
            <span class="n">min_feature_size</span><span class="o">=</span><span class="n">min_feature_size</span><span class="p">,</span>
            <span class="n">min_background_size</span><span class="o">=</span><span class="n">min_background_size</span><span class="p">)</span>

<span class="c"># Add interactive visualizations</span></div></div>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_classifier</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_classifier</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_graph</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_graph</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_pca</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_pca</span>
<span class="c"># Study.interactive_localZ = Interactive.interactive_localZ</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_lavalamp_pooled_inconsistent</span> <span class="o">=</span> \
    <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_lavalamp_pooled_inconsistent</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_choose_outliers</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_choose_outliers</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_reset_outliers</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_reset_outliers</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_clustermap</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_clustermap</span>
<span class="n">Study</span><span class="o">.</span><span class="n">interactive_correlations</span> <span class="o">=</span> <span class="n">Interactive</span><span class="o">.</span><span class="n">interactive_correlations</span>
</pre></div>

    </div>
      
  </div>
</div>
<div class="container">
    
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2014, Olga B. Botvinnik, Michael T. Lovci, Patrick Liu, Yan Song, and Gene Yeo.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.<br/>
    </p>
  </div>
</footer>
    <a href="http://olgabotvinnik.com/">Olga B. Botvinnik</a> is funded by the
    <a href="https://ndseg.asee.org/">NDSEG</a> fellowship and is a
    <a href="http://numfocus.org/">NumFOCUS</a>
    <a href="http://numfocus.org/pages/john_hunter_fellowship_2014.html">John
    Hunter Technology Fellow</a>.<br/>
    Michael T. Lovci was partially funded by a fellowship from
    <a href="http://www.gene.com/">Genentech</a>.</br>
    Partially funded by <a href="http://www.nih.gov/">NIH</a> grants NS075449
    and HG004659 and <a href="http://www.cirm.ca.gov/">CIRM</a> grants RB4-06045
    and TR3-05676 to <a href="http://yeolab.ucsd.edu/yeolab/Home.html">Gene
    Yeo</a>. <br/>

    
  </body>
</html>